{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "First, we need to initialize the object (a Pertpy package internal dataset), and if the object isn't already preprocessed and/or clustered, perform those steps.\n",
    "\n",
    "## Imports & Options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "\n",
      "Analysis Functions\n",
      "\tanalyze_causal_network\n",
      "\tanalyze_composition\n",
      "\tanalyze_receptor_ligand\n",
      "\tcalculate_dea_deseq2\n",
      "\tcluster\n",
      "\tclustering\n",
      "\tcommunication\n",
      "\tcomposition\n",
      "\tcompute_distance\n",
      "\tfind_marker_genes\n",
      "\tperform_augur\n",
      "\tperform_celltypist\n",
      "\tperform_dea\n",
      "\tperform_differential_prioritization\n",
      "\tperform_gsea\n",
      "\tperform_mixscape\n",
      "\tperform_pathway_interference\n",
      "\tperturbations\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import crispr as cr \n",
    "from crispr import Crispr\n",
    "import pertpy as pt\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Options\n",
    "print(\"\\nAnalysis Functions\\n\\t\" + \"\\n\\t\".join(list(pd.Series(\n",
    "    [np.nan if \"__\" in x else x for x in dir(cr.ax)]).dropna())))\n",
    "file = \"perturb-seq\"\n",
    "pd.options.display.max_columns = 100\n",
    "\n",
    "#  Set Arguments\n",
    "kwargs_init = dict(assay=\"rna\", assay_protein=\"adt\",\n",
    "                   col_sample_id=\"replicate\",\n",
    "                   col_gene_symbols=\"gene_symbol\",  \n",
    "                   col_cell_type=\"leiden\", \n",
    "                   col_perturbed=\"perturbation\", \n",
    "                   col_guide_rna=\"NT\", \n",
    "                   col_num_umis=None,\n",
    "                   kws_process_guide_rna=False,\n",
    "                   col_condition=\"gene_target\", \n",
    "                   key_control=\"NT\", \n",
    "                   key_treatment=\"Perturbed\")\n",
    "target_gene_idents = [\"NT\",\"JAK2\",\"STAT1\",\"IFNGR1\",\"IFNGR2\", \"IRF1\"]\n",
    "file_path = pt.dt.papalexi_2021()\n",
    "file_path[kwargs_init[\"assay\"]].layers[\"counts\"] = file_path[\n",
    "    kwargs_init[\"assay\"]].X.copy()  # X contains counts layer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Object\n",
    "\n",
    "This code instantiates the CRISPR object, which is the main way of interacting with this package as an end-user.\n",
    "\n",
    "This is more code than you would need in real life; it just ensures that certain public datasets are loaded from the source for various reasons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "<<< INITIALIZING OMICS CLASS OBJECT >>>\n",
      "\n",
      "col_gene_symbols=\"gene_symbol\"\n",
      "col_cell_type=\"leiden\"\n",
      "col_sample_id=\"replicate\"\n",
      "col_batch=\"replicate\"\n",
      "col_subject=None\n",
      "col_condition=\"gene_target\"\n",
      "col_num_umis=None\n",
      "key_control=\"NT\"\n",
      "key_treatment=\"Perturbed\"\n",
      "\n",
      "\n",
      "<<< LOADING OBJECT >>>\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/elizabeth/elizabeth/crispr/crispr/processing/preprocessing.py\", line 137, in create_object\n",
      "    adata.obs_names_make_unique()\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/py-bio/lib/python3.10/site-packages/mudata/_core/mudata.py\", line 974, in obs_names_make_unique\n",
      "    self._obs.index = obs_names\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/py-bio/lib/python3.10/site-packages/pandas/core/generic.py\", line 6307, in __setattr__\n",
      "    return object.__setattr__(self, name, value)\n",
      "  File \"properties.pyx\", line 69, in pandas._libs.properties.AxisProperty.__set__\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/py-bio/lib/python3.10/site-packages/pandas/core/generic.py\", line 813, in _set_axis\n",
      "    self._mgr.set_axis(axis, labels)\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/py-bio/lib/python3.10/site-packages/pandas/core/internals/managers.py\", line 239, in set_axis\n",
      "    self._validate_set_axis(axis, new_labels)\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/py-bio/lib/python3.10/site-packages/pandas/core/internals/base.py\", line 98, in _validate_set_axis\n",
      "    raise ValueError(\n",
      "ValueError: Length mismatch: Expected axis has 20729 elements, new values have 82916 elements\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Cell Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "20729\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Gene Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "\n",
      " MuData object with n_obs × n_vars = 20729 × 18776\n",
      "  4 modalities\n",
      "    rna:\t20729 x 18649\n",
      "      obs:\t'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'nCount_HTO', 'nFeature_HTO', 'nCount_GDO', 'nCount_ADT', 'nFeature_ADT', 'percent.mito', 'MULTI_ID', 'HTO_classification', 'guide_ID', 'gene_target', 'NT', 'perturbation', 'replicate', 'S.Score', 'G2M.Score', 'Phase'\n",
      "      var:\t'name'\n",
      "      layers:\t'counts'\n",
      "    adt:\t20729 x 4\n",
      "      obs:\t'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'nCount_HTO', 'nFeature_HTO', 'nCount_GDO', 'nCount_ADT', 'nFeature_ADT', 'percent.mito', 'MULTI_ID', 'HTO_classification', 'guide_ID', 'gene_target', 'NT', 'perturbation', 'replicate', 'S.Score', 'G2M.Score', 'Phase'\n",
      "      var:\t'name'\n",
      "    hto:\t20729 x 12\n",
      "      obs:\t'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'nCount_HTO', 'nFeature_HTO', 'nCount_GDO', 'nCount_ADT', 'nFeature_ADT', 'percent.mito', 'MULTI_ID', 'HTO_classification', 'guide_ID', 'gene_target', 'NT', 'perturbation', 'replicate', 'S.Score', 'G2M.Score', 'Phase'\n",
      "      var:\t'name'\n",
      "    gdo:\t20729 x 111\n",
      "      obs:\t'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'nCount_HTO', 'nFeature_HTO', 'nCount_GDO', 'nCount_ADT', 'nFeature_ADT', 'percent.mito', 'MULTI_ID', 'HTO_classification', 'guide_ID', 'gene_target', 'NT', 'perturbation', 'replicate', 'S.Score', 'G2M.Score', 'Phase'\n",
      "      var:\t'name'\n",
      "                    rna:orig.ident  rna:nCount_RNA  rna:nFeature_RNA  \\\n",
      "index                                                                  \n",
      "l1_AAACCTGAGCCAGAAC          Lane1           17207              3942   \n",
      "l1_AAACCTGAGTGGACGT          Lane1            9506              2948   \n",
      "l1_AAACCTGCATGAGCGA          Lane1           15256              4258   \n",
      "l1_AAACCTGTCTTGTCAT          Lane1            5135              1780   \n",
      "l1_AAACGGGAGAACAACT          Lane1            9673              2671   \n",
      "...                            ...             ...               ...   \n",
      "l8_TTTGTCAGTCACTTCC          Lane8           20863              4884   \n",
      "l8_TTTGTCAGTGACGGTA          Lane8           17553              3787   \n",
      "l8_TTTGTCAGTTCCACAA          Lane8           15106              4185   \n",
      "l8_TTTGTCATCACGCATA          Lane8           11209              3204   \n",
      "l8_TTTGTCATCTGTACGA          Lane8           23088              5073   \n",
      "\n",
      "                     rna:nCount_HTO  rna:nFeature_HTO  rna:nCount_GDO  \\\n",
      "index                                                                   \n",
      "l1_AAACCTGAGCCAGAAC            99.0                 4           576.0   \n",
      "l1_AAACCTGAGTGGACGT            35.0                 5           190.0   \n",
      "l1_AAACCTGCATGAGCGA            66.0                 4           212.0   \n",
      "l1_AAACCTGTCTTGTCAT            22.0                 3           243.0   \n",
      "l1_AAACGGGAGAACAACT            99.0                 5           198.0   \n",
      "...                             ...               ...             ...   \n",
      "l8_TTTGTCAGTCACTTCC           294.0                 4           290.0   \n",
      "l8_TTTGTCAGTGACGGTA           528.0                 3           870.0   \n",
      "l8_TTTGTCAGTTCCACAA           154.0                 6           267.0   \n",
      "l8_TTTGTCATCACGCATA           132.0                 3           202.0   \n",
      "l8_TTTGTCATCTGTACGA            95.0                 4           234.0   \n",
      "\n",
      "                     rna:nCount_ADT  rna:nFeature_ADT  rna:percent.mito  \\\n",
      "index                                                                     \n",
      "l1_AAACCTGAGCCAGAAC           801.0                 4          2.295577   \n",
      "l1_AAACCTGAGTGGACGT           545.0                 4          4.512939   \n",
      "l1_AAACCTGCATGAGCGA           344.0                 4          4.116413   \n",
      "l1_AAACCTGTCTTGTCAT           539.0                 4          5.491723   \n",
      "l1_AAACGGGAGAACAACT          1053.0                 4          3.359868   \n",
      "...                             ...               ...               ...   \n",
      "l8_TTTGTCAGTCACTTCC           398.0                 4          5.469012   \n",
      "l8_TTTGTCAGTGACGGTA          3042.0                 4          2.159175   \n",
      "l8_TTTGTCAGTTCCACAA           212.0                 4          2.661194   \n",
      "l8_TTTGTCATCACGCATA           258.0                 4          7.369078   \n",
      "l8_TTTGTCATCTGTACGA           584.0                 4          2.191615   \n",
      "\n",
      "                    rna:MULTI_ID rna:HTO_classification rna:guide_ID  \\\n",
      "index                                                                  \n",
      "l1_AAACCTGAGCCAGAAC      rep1-tx                rep1-tx      STAT2g2   \n",
      "l1_AAACCTGAGTGGACGT      rep1-tx                rep1-tx       CAV1g4   \n",
      "l1_AAACCTGCATGAGCGA      rep1-tx                rep1-tx      STAT1g2   \n",
      "l1_AAACCTGTCTTGTCAT      rep1-tx                rep1-tx       CD86g1   \n",
      "l1_AAACGGGAGAACAACT      rep1-tx                rep1-tx       IRF7g2   \n",
      "...                          ...                    ...          ...   \n",
      "l8_TTTGTCAGTCACTTCC      rep3-tx                rep3-tx      CMTM6g1   \n",
      "l8_TTTGTCAGTGACGGTA      rep4-tx                rep2-tx         NTg4   \n",
      "l8_TTTGTCAGTTCCACAA      rep3-tx                rep3-tx       ATF2g1   \n",
      "l8_TTTGTCATCACGCATA      rep3-tx                rep3-tx       CAV1g2   \n",
      "l8_TTTGTCATCTGTACGA      rep3-tx                rep3-tx      STAT3g1   \n",
      "\n",
      "                    rna:gene_target   rna:NT rna:perturbation rna:replicate  \\\n",
      "index                                                                         \n",
      "l1_AAACCTGAGCCAGAAC           STAT2  STAT2g2        Perturbed          rep1   \n",
      "l1_AAACCTGAGTGGACGT            CAV1   CAV1g4        Perturbed          rep1   \n",
      "l1_AAACCTGCATGAGCGA           STAT1  STAT1g2        Perturbed          rep1   \n",
      "l1_AAACCTGTCTTGTCAT            CD86   CD86g1        Perturbed          rep1   \n",
      "l1_AAACGGGAGAACAACT            IRF7   IRF7g2        Perturbed          rep1   \n",
      "...                             ...      ...              ...           ...   \n",
      "l8_TTTGTCAGTCACTTCC           CMTM6  CMTM6g1        Perturbed          rep3   \n",
      "l8_TTTGTCAGTGACGGTA              NT       NT               NT          rep2   \n",
      "l8_TTTGTCAGTTCCACAA            ATF2   ATF2g1        Perturbed          rep3   \n",
      "l8_TTTGTCATCACGCATA            CAV1   CAV1g2        Perturbed          rep3   \n",
      "l8_TTTGTCATCTGTACGA           STAT3  STAT3g1        Perturbed          rep3   \n",
      "\n",
      "                     rna:S.Score  rna:G2M.Score rna:Phase adt:orig.ident  \\\n",
      "index                                                                      \n",
      "l1_AAACCTGAGCCAGAAC    -0.252716      -0.771309        G1          Lane1   \n",
      "l1_AAACCTGAGTGGACGT    -0.123802      -0.332603        G1          Lane1   \n",
      "l1_AAACCTGCATGAGCGA    -0.154633      -0.694418        G1          Lane1   \n",
      "l1_AAACCTGTCTTGTCAT    -0.061262      -0.037820        G1          Lane1   \n",
      "l1_AAACGGGAGAACAACT    -0.132188      -0.353156        G1          Lane1   \n",
      "...                          ...            ...       ...            ...   \n",
      "l8_TTTGTCAGTCACTTCC    -0.323562      -0.794679        G1          Lane8   \n",
      "l8_TTTGTCAGTGACGGTA    -0.153514      -0.632655        G1          Lane8   \n",
      "l8_TTTGTCAGTTCCACAA    -0.191933      -0.574283        G1          Lane8   \n",
      "l8_TTTGTCATCACGCATA    -0.134585      -0.501513        G1          Lane8   \n",
      "l8_TTTGTCATCTGTACGA    -0.461981      -1.071414        G1          Lane8   \n",
      "\n",
      "                     adt:nCount_RNA  adt:nFeature_RNA  adt:nCount_HTO  \\\n",
      "index                                                                   \n",
      "l1_AAACCTGAGCCAGAAC           17207              3942            99.0   \n",
      "l1_AAACCTGAGTGGACGT            9506              2948            35.0   \n",
      "l1_AAACCTGCATGAGCGA           15256              4258            66.0   \n",
      "l1_AAACCTGTCTTGTCAT            5135              1780            22.0   \n",
      "l1_AAACGGGAGAACAACT            9673              2671            99.0   \n",
      "...                             ...               ...             ...   \n",
      "l8_TTTGTCAGTCACTTCC           20863              4884           294.0   \n",
      "l8_TTTGTCAGTGACGGTA           17553              3787           528.0   \n",
      "l8_TTTGTCAGTTCCACAA           15106              4185           154.0   \n",
      "l8_TTTGTCATCACGCATA           11209              3204           132.0   \n",
      "l8_TTTGTCATCTGTACGA           23088              5073            95.0   \n",
      "\n",
      "                     adt:nFeature_HTO  adt:nCount_GDO  adt:nCount_ADT  \\\n",
      "index                                                                   \n",
      "l1_AAACCTGAGCCAGAAC                 4           576.0           801.0   \n",
      "l1_AAACCTGAGTGGACGT                 5           190.0           545.0   \n",
      "l1_AAACCTGCATGAGCGA                 4           212.0           344.0   \n",
      "l1_AAACCTGTCTTGTCAT                 3           243.0           539.0   \n",
      "l1_AAACGGGAGAACAACT                 5           198.0          1053.0   \n",
      "...                               ...             ...             ...   \n",
      "l8_TTTGTCAGTCACTTCC                 4           290.0           398.0   \n",
      "l8_TTTGTCAGTGACGGTA                 3           870.0          3042.0   \n",
      "l8_TTTGTCAGTTCCACAA                 6           267.0           212.0   \n",
      "l8_TTTGTCATCACGCATA                 3           202.0           258.0   \n",
      "l8_TTTGTCATCTGTACGA                 4           234.0           584.0   \n",
      "\n",
      "                     adt:nFeature_ADT  adt:percent.mito adt:MULTI_ID  \\\n",
      "index                                                                  \n",
      "l1_AAACCTGAGCCAGAAC                 4          2.295577      rep1-tx   \n",
      "l1_AAACCTGAGTGGACGT                 4          4.512939      rep1-tx   \n",
      "l1_AAACCTGCATGAGCGA                 4          4.116413      rep1-tx   \n",
      "l1_AAACCTGTCTTGTCAT                 4          5.491723      rep1-tx   \n",
      "l1_AAACGGGAGAACAACT                 4          3.359868      rep1-tx   \n",
      "...                               ...               ...          ...   \n",
      "l8_TTTGTCAGTCACTTCC                 4          5.469012      rep3-tx   \n",
      "l8_TTTGTCAGTGACGGTA                 4          2.159175      rep4-tx   \n",
      "l8_TTTGTCAGTTCCACAA                 4          2.661194      rep3-tx   \n",
      "l8_TTTGTCATCACGCATA                 4          7.369078      rep3-tx   \n",
      "l8_TTTGTCATCTGTACGA                 4          2.191615      rep3-tx   \n",
      "\n",
      "                    adt:HTO_classification adt:guide_ID adt:gene_target  \\\n",
      "index                                                                     \n",
      "l1_AAACCTGAGCCAGAAC                rep1-tx      STAT2g2           STAT2   \n",
      "l1_AAACCTGAGTGGACGT                rep1-tx       CAV1g4            CAV1   \n",
      "l1_AAACCTGCATGAGCGA                rep1-tx      STAT1g2           STAT1   \n",
      "l1_AAACCTGTCTTGTCAT                rep1-tx       CD86g1            CD86   \n",
      "l1_AAACGGGAGAACAACT                rep1-tx       IRF7g2            IRF7   \n",
      "...                                    ...          ...             ...   \n",
      "l8_TTTGTCAGTCACTTCC                rep3-tx      CMTM6g1           CMTM6   \n",
      "l8_TTTGTCAGTGACGGTA                rep2-tx         NTg4              NT   \n",
      "l8_TTTGTCAGTTCCACAA                rep3-tx       ATF2g1            ATF2   \n",
      "l8_TTTGTCATCACGCATA                rep3-tx       CAV1g2            CAV1   \n",
      "l8_TTTGTCATCTGTACGA                rep3-tx      STAT3g1           STAT3   \n",
      "\n",
      "                      adt:NT adt:perturbation adt:replicate  adt:S.Score  \\\n",
      "index                                                                      \n",
      "l1_AAACCTGAGCCAGAAC  STAT2g2        Perturbed          rep1    -0.252716   \n",
      "l1_AAACCTGAGTGGACGT   CAV1g4        Perturbed          rep1    -0.123802   \n",
      "l1_AAACCTGCATGAGCGA  STAT1g2        Perturbed          rep1    -0.154633   \n",
      "l1_AAACCTGTCTTGTCAT   CD86g1        Perturbed          rep1    -0.061262   \n",
      "l1_AAACGGGAGAACAACT   IRF7g2        Perturbed          rep1    -0.132188   \n",
      "...                      ...              ...           ...          ...   \n",
      "l8_TTTGTCAGTCACTTCC  CMTM6g1        Perturbed          rep3    -0.323562   \n",
      "l8_TTTGTCAGTGACGGTA       NT               NT          rep2    -0.153514   \n",
      "l8_TTTGTCAGTTCCACAA   ATF2g1        Perturbed          rep3    -0.191933   \n",
      "l8_TTTGTCATCACGCATA   CAV1g2        Perturbed          rep3    -0.134585   \n",
      "l8_TTTGTCATCTGTACGA  STAT3g1        Perturbed          rep3    -0.461981   \n",
      "\n",
      "                     adt:G2M.Score adt:Phase hto:orig.ident  hto:nCount_RNA  \\\n",
      "index                                                                         \n",
      "l1_AAACCTGAGCCAGAAC      -0.771309        G1          Lane1           17207   \n",
      "l1_AAACCTGAGTGGACGT      -0.332603        G1          Lane1            9506   \n",
      "l1_AAACCTGCATGAGCGA      -0.694418        G1          Lane1           15256   \n",
      "l1_AAACCTGTCTTGTCAT      -0.037820        G1          Lane1            5135   \n",
      "l1_AAACGGGAGAACAACT      -0.353156        G1          Lane1            9673   \n",
      "...                            ...       ...            ...             ...   \n",
      "l8_TTTGTCAGTCACTTCC      -0.794679        G1          Lane8           20863   \n",
      "l8_TTTGTCAGTGACGGTA      -0.632655        G1          Lane8           17553   \n",
      "l8_TTTGTCAGTTCCACAA      -0.574283        G1          Lane8           15106   \n",
      "l8_TTTGTCATCACGCATA      -0.501513        G1          Lane8           11209   \n",
      "l8_TTTGTCATCTGTACGA      -1.071414        G1          Lane8           23088   \n",
      "\n",
      "                     hto:nFeature_RNA  hto:nCount_HTO  hto:nFeature_HTO  \\\n",
      "index                                                                     \n",
      "l1_AAACCTGAGCCAGAAC              3942            99.0                 4   \n",
      "l1_AAACCTGAGTGGACGT              2948            35.0                 5   \n",
      "l1_AAACCTGCATGAGCGA              4258            66.0                 4   \n",
      "l1_AAACCTGTCTTGTCAT              1780            22.0                 3   \n",
      "l1_AAACGGGAGAACAACT              2671            99.0                 5   \n",
      "...                               ...             ...               ...   \n",
      "l8_TTTGTCAGTCACTTCC              4884           294.0                 4   \n",
      "l8_TTTGTCAGTGACGGTA              3787           528.0                 3   \n",
      "l8_TTTGTCAGTTCCACAA              4185           154.0                 6   \n",
      "l8_TTTGTCATCACGCATA              3204           132.0                 3   \n",
      "l8_TTTGTCATCTGTACGA              5073            95.0                 4   \n",
      "\n",
      "                     hto:nCount_GDO  hto:nCount_ADT  hto:nFeature_ADT  \\\n",
      "index                                                                   \n",
      "l1_AAACCTGAGCCAGAAC           576.0           801.0                 4   \n",
      "l1_AAACCTGAGTGGACGT           190.0           545.0                 4   \n",
      "l1_AAACCTGCATGAGCGA           212.0           344.0                 4   \n",
      "l1_AAACCTGTCTTGTCAT           243.0           539.0                 4   \n",
      "l1_AAACGGGAGAACAACT           198.0          1053.0                 4   \n",
      "...                             ...             ...               ...   \n",
      "l8_TTTGTCAGTCACTTCC           290.0           398.0                 4   \n",
      "l8_TTTGTCAGTGACGGTA           870.0          3042.0                 4   \n",
      "l8_TTTGTCAGTTCCACAA           267.0           212.0                 4   \n",
      "l8_TTTGTCATCACGCATA           202.0           258.0                 4   \n",
      "l8_TTTGTCATCTGTACGA           234.0           584.0                 4   \n",
      "\n",
      "                     hto:percent.mito hto:MULTI_ID hto:HTO_classification  \\\n",
      "index                                                                       \n",
      "l1_AAACCTGAGCCAGAAC          2.295577      rep1-tx                rep1-tx   \n",
      "l1_AAACCTGAGTGGACGT          4.512939      rep1-tx                rep1-tx   \n",
      "l1_AAACCTGCATGAGCGA          4.116413      rep1-tx                rep1-tx   \n",
      "l1_AAACCTGTCTTGTCAT          5.491723      rep1-tx                rep1-tx   \n",
      "l1_AAACGGGAGAACAACT          3.359868      rep1-tx                rep1-tx   \n",
      "...                               ...          ...                    ...   \n",
      "l8_TTTGTCAGTCACTTCC          5.469012      rep3-tx                rep3-tx   \n",
      "l8_TTTGTCAGTGACGGTA          2.159175      rep4-tx                rep2-tx   \n",
      "l8_TTTGTCAGTTCCACAA          2.661194      rep3-tx                rep3-tx   \n",
      "l8_TTTGTCATCACGCATA          7.369078      rep3-tx                rep3-tx   \n",
      "l8_TTTGTCATCTGTACGA          2.191615      rep3-tx                rep3-tx   \n",
      "\n",
      "                    hto:guide_ID hto:gene_target   hto:NT hto:perturbation  \\\n",
      "index                                                                        \n",
      "l1_AAACCTGAGCCAGAAC      STAT2g2           STAT2  STAT2g2        Perturbed   \n",
      "l1_AAACCTGAGTGGACGT       CAV1g4            CAV1   CAV1g4        Perturbed   \n",
      "l1_AAACCTGCATGAGCGA      STAT1g2           STAT1  STAT1g2        Perturbed   \n",
      "l1_AAACCTGTCTTGTCAT       CD86g1            CD86   CD86g1        Perturbed   \n",
      "l1_AAACGGGAGAACAACT       IRF7g2            IRF7   IRF7g2        Perturbed   \n",
      "...                          ...             ...      ...              ...   \n",
      "l8_TTTGTCAGTCACTTCC      CMTM6g1           CMTM6  CMTM6g1        Perturbed   \n",
      "l8_TTTGTCAGTGACGGTA         NTg4              NT       NT               NT   \n",
      "l8_TTTGTCAGTTCCACAA       ATF2g1            ATF2   ATF2g1        Perturbed   \n",
      "l8_TTTGTCATCACGCATA       CAV1g2            CAV1   CAV1g2        Perturbed   \n",
      "l8_TTTGTCATCTGTACGA      STAT3g1           STAT3  STAT3g1        Perturbed   \n",
      "\n",
      "                    hto:replicate  hto:S.Score  hto:G2M.Score hto:Phase  \\\n",
      "index                                                                     \n",
      "l1_AAACCTGAGCCAGAAC          rep1    -0.252716      -0.771309        G1   \n",
      "l1_AAACCTGAGTGGACGT          rep1    -0.123802      -0.332603        G1   \n",
      "l1_AAACCTGCATGAGCGA          rep1    -0.154633      -0.694418        G1   \n",
      "l1_AAACCTGTCTTGTCAT          rep1    -0.061262      -0.037820        G1   \n",
      "l1_AAACGGGAGAACAACT          rep1    -0.132188      -0.353156        G1   \n",
      "...                           ...          ...            ...       ...   \n",
      "l8_TTTGTCAGTCACTTCC          rep3    -0.323562      -0.794679        G1   \n",
      "l8_TTTGTCAGTGACGGTA          rep2    -0.153514      -0.632655        G1   \n",
      "l8_TTTGTCAGTTCCACAA          rep3    -0.191933      -0.574283        G1   \n",
      "l8_TTTGTCATCACGCATA          rep3    -0.134585      -0.501513        G1   \n",
      "l8_TTTGTCATCTGTACGA          rep3    -0.461981      -1.071414        G1   \n",
      "\n",
      "                    gdo:orig.ident  gdo:nCount_RNA  gdo:nFeature_RNA  \\\n",
      "index                                                                  \n",
      "l1_AAACCTGAGCCAGAAC          Lane1           17207              3942   \n",
      "l1_AAACCTGAGTGGACGT          Lane1            9506              2948   \n",
      "l1_AAACCTGCATGAGCGA          Lane1           15256              4258   \n",
      "l1_AAACCTGTCTTGTCAT          Lane1            5135              1780   \n",
      "l1_AAACGGGAGAACAACT          Lane1            9673              2671   \n",
      "...                            ...             ...               ...   \n",
      "l8_TTTGTCAGTCACTTCC          Lane8           20863              4884   \n",
      "l8_TTTGTCAGTGACGGTA          Lane8           17553              3787   \n",
      "l8_TTTGTCAGTTCCACAA          Lane8           15106              4185   \n",
      "l8_TTTGTCATCACGCATA          Lane8           11209              3204   \n",
      "l8_TTTGTCATCTGTACGA          Lane8           23088              5073   \n",
      "\n",
      "                     gdo:nCount_HTO  gdo:nFeature_HTO  gdo:nCount_GDO  \\\n",
      "index                                                                   \n",
      "l1_AAACCTGAGCCAGAAC            99.0                 4           576.0   \n",
      "l1_AAACCTGAGTGGACGT            35.0                 5           190.0   \n",
      "l1_AAACCTGCATGAGCGA            66.0                 4           212.0   \n",
      "l1_AAACCTGTCTTGTCAT            22.0                 3           243.0   \n",
      "l1_AAACGGGAGAACAACT            99.0                 5           198.0   \n",
      "...                             ...               ...             ...   \n",
      "l8_TTTGTCAGTCACTTCC           294.0                 4           290.0   \n",
      "l8_TTTGTCAGTGACGGTA           528.0                 3           870.0   \n",
      "l8_TTTGTCAGTTCCACAA           154.0                 6           267.0   \n",
      "l8_TTTGTCATCACGCATA           132.0                 3           202.0   \n",
      "l8_TTTGTCATCTGTACGA            95.0                 4           234.0   \n",
      "\n",
      "                     gdo:nCount_ADT  gdo:nFeature_ADT  gdo:percent.mito  \\\n",
      "index                                                                     \n",
      "l1_AAACCTGAGCCAGAAC           801.0                 4          2.295577   \n",
      "l1_AAACCTGAGTGGACGT           545.0                 4          4.512939   \n",
      "l1_AAACCTGCATGAGCGA           344.0                 4          4.116413   \n",
      "l1_AAACCTGTCTTGTCAT           539.0                 4          5.491723   \n",
      "l1_AAACGGGAGAACAACT          1053.0                 4          3.359868   \n",
      "...                             ...               ...               ...   \n",
      "l8_TTTGTCAGTCACTTCC           398.0                 4          5.469012   \n",
      "l8_TTTGTCAGTGACGGTA          3042.0                 4          2.159175   \n",
      "l8_TTTGTCAGTTCCACAA           212.0                 4          2.661194   \n",
      "l8_TTTGTCATCACGCATA           258.0                 4          7.369078   \n",
      "l8_TTTGTCATCTGTACGA           584.0                 4          2.191615   \n",
      "\n",
      "                    gdo:MULTI_ID gdo:HTO_classification gdo:guide_ID  \\\n",
      "index                                                                  \n",
      "l1_AAACCTGAGCCAGAAC      rep1-tx                rep1-tx      STAT2g2   \n",
      "l1_AAACCTGAGTGGACGT      rep1-tx                rep1-tx       CAV1g4   \n",
      "l1_AAACCTGCATGAGCGA      rep1-tx                rep1-tx      STAT1g2   \n",
      "l1_AAACCTGTCTTGTCAT      rep1-tx                rep1-tx       CD86g1   \n",
      "l1_AAACGGGAGAACAACT      rep1-tx                rep1-tx       IRF7g2   \n",
      "...                          ...                    ...          ...   \n",
      "l8_TTTGTCAGTCACTTCC      rep3-tx                rep3-tx      CMTM6g1   \n",
      "l8_TTTGTCAGTGACGGTA      rep4-tx                rep2-tx         NTg4   \n",
      "l8_TTTGTCAGTTCCACAA      rep3-tx                rep3-tx       ATF2g1   \n",
      "l8_TTTGTCATCACGCATA      rep3-tx                rep3-tx       CAV1g2   \n",
      "l8_TTTGTCATCTGTACGA      rep3-tx                rep3-tx      STAT3g1   \n",
      "\n",
      "                    gdo:gene_target   gdo:NT gdo:perturbation gdo:replicate  \\\n",
      "index                                                                         \n",
      "l1_AAACCTGAGCCAGAAC           STAT2  STAT2g2        Perturbed          rep1   \n",
      "l1_AAACCTGAGTGGACGT            CAV1   CAV1g4        Perturbed          rep1   \n",
      "l1_AAACCTGCATGAGCGA           STAT1  STAT1g2        Perturbed          rep1   \n",
      "l1_AAACCTGTCTTGTCAT            CD86   CD86g1        Perturbed          rep1   \n",
      "l1_AAACGGGAGAACAACT            IRF7   IRF7g2        Perturbed          rep1   \n",
      "...                             ...      ...              ...           ...   \n",
      "l8_TTTGTCAGTCACTTCC           CMTM6  CMTM6g1        Perturbed          rep3   \n",
      "l8_TTTGTCAGTGACGGTA              NT       NT               NT          rep2   \n",
      "l8_TTTGTCAGTTCCACAA            ATF2   ATF2g1        Perturbed          rep3   \n",
      "l8_TTTGTCATCACGCATA            CAV1   CAV1g2        Perturbed          rep3   \n",
      "l8_TTTGTCATCTGTACGA           STAT3  STAT3g1        Perturbed          rep3   \n",
      "\n",
      "                     gdo:S.Score  gdo:G2M.Score gdo:Phase  \n",
      "index                                                      \n",
      "l1_AAACCTGAGCCAGAAC    -0.252716      -0.771309        G1  \n",
      "l1_AAACCTGAGTGGACGT    -0.123802      -0.332603        G1  \n",
      "l1_AAACCTGCATGAGCGA    -0.154633      -0.694418        G1  \n",
      "l1_AAACCTGTCTTGTCAT    -0.061262      -0.037820        G1  \n",
      "l1_AAACGGGAGAACAACT    -0.132188      -0.353156        G1  \n",
      "...                          ...            ...       ...  \n",
      "l8_TTTGTCAGTCACTTCC    -0.323562      -0.794679        G1  \n",
      "l8_TTTGTCAGTGACGGTA    -0.153514      -0.632655        G1  \n",
      "l8_TTTGTCAGTTCCACAA    -0.191933      -0.574283        G1  \n",
      "l8_TTTGTCATCACGCATA    -0.134585      -0.501513        G1  \n",
      "l8_TTTGTCATCTGTACGA    -0.461981      -1.071414        G1  \n",
      "\n",
      "[20729 rows x 76 columns] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 20729 × 18649\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'nCount_HTO', 'nFeature_HTO', 'nCount_GDO', 'nCount_ADT', 'nFeature_ADT', 'percent.mito', 'MULTI_ID', 'HTO_classification', 'guide_ID', 'gene_target', 'NT', 'perturbation', 'replicate', 'S.Score', 'G2M.Score', 'Phase'\n",
      "    var: 'name'\n",
      "    layers: 'counts'\n",
      "                            name\n",
      "gene_symbol                     \n",
      "rna:AL627309.1        AL627309.1\n",
      "rna:AP006222.2        AP006222.2\n",
      "rna:RP4-669L17.10  RP4-669L17.10\n",
      "rna:RP11-206L10.3  RP11-206L10.3\n",
      "rna:RP11-206L10.2  RP11-206L10.2\n",
      "                        orig.ident  nCount_RNA  nFeature_RNA  nCount_HTO  \\\n",
      "index                                                                      \n",
      "rna:l1_AAACCTGAGCCAGAAC      Lane1       17207          3942        99.0   \n",
      "rna:l1_AAACCTGAGTGGACGT      Lane1        9506          2948        35.0   \n",
      "rna:l1_AAACCTGCATGAGCGA      Lane1       15256          4258        66.0   \n",
      "rna:l1_AAACCTGTCTTGTCAT      Lane1        5135          1780        22.0   \n",
      "rna:l1_AAACGGGAGAACAACT      Lane1        9673          2671        99.0   \n",
      "\n",
      "                         nFeature_HTO  nCount_GDO  nCount_ADT  nFeature_ADT  \\\n",
      "index                                                                         \n",
      "rna:l1_AAACCTGAGCCAGAAC             4       576.0       801.0             4   \n",
      "rna:l1_AAACCTGAGTGGACGT             5       190.0       545.0             4   \n",
      "rna:l1_AAACCTGCATGAGCGA             4       212.0       344.0             4   \n",
      "rna:l1_AAACCTGTCTTGTCAT             3       243.0       539.0             4   \n",
      "rna:l1_AAACGGGAGAACAACT             5       198.0      1053.0             4   \n",
      "\n",
      "                         percent.mito MULTI_ID HTO_classification guide_ID  \\\n",
      "index                                                                        \n",
      "rna:l1_AAACCTGAGCCAGAAC      2.295577  rep1-tx            rep1-tx  STAT2g2   \n",
      "rna:l1_AAACCTGAGTGGACGT      4.512939  rep1-tx            rep1-tx   CAV1g4   \n",
      "rna:l1_AAACCTGCATGAGCGA      4.116413  rep1-tx            rep1-tx  STAT1g2   \n",
      "rna:l1_AAACCTGTCTTGTCAT      5.491723  rep1-tx            rep1-tx   CD86g1   \n",
      "rna:l1_AAACGGGAGAACAACT      3.359868  rep1-tx            rep1-tx   IRF7g2   \n",
      "\n",
      "                        gene_target       NT perturbation replicate   S.Score  \\\n",
      "index                                                                           \n",
      "rna:l1_AAACCTGAGCCAGAAC       STAT2  STAT2g2    Perturbed      rep1 -0.252716   \n",
      "rna:l1_AAACCTGAGTGGACGT        CAV1   CAV1g4    Perturbed      rep1 -0.123802   \n",
      "rna:l1_AAACCTGCATGAGCGA       STAT1  STAT1g2    Perturbed      rep1 -0.154633   \n",
      "rna:l1_AAACCTGTCTTGTCAT        CD86   CD86g1    Perturbed      rep1 -0.061262   \n",
      "rna:l1_AAACGGGAGAACAACT        IRF7   IRF7g2    Perturbed      rep1 -0.132188   \n",
      "\n",
      "                         G2M.Score Phase  \n",
      "index                                     \n",
      "rna:l1_AAACCTGAGCCAGAAC  -0.771309    G1  \n",
      "rna:l1_AAACCTGAGTGGACGT  -0.332603    G1  \n",
      "rna:l1_AAACCTGCATGAGCGA  -0.694418    G1  \n",
      "rna:l1_AAACCTGTCTTGTCAT  -0.037820    G1  \n",
      "rna:l1_AAACGGGAGAACAACT  -0.353156    G1  \n",
      "                        rna:orig.ident  rna:nCount_RNA  rna:nFeature_RNA  \\\n",
      "index                                                                      \n",
      "rna:l1_AAACCTGAGCCAGAAC          Lane1         17207.0            3942.0   \n",
      "rna:l1_AAACCTGAGTGGACGT          Lane1          9506.0            2948.0   \n",
      "rna:l1_AAACCTGCATGAGCGA          Lane1         15256.0            4258.0   \n",
      "rna:l1_AAACCTGTCTTGTCAT          Lane1          5135.0            1780.0   \n",
      "rna:l1_AAACGGGAGAACAACT          Lane1          9673.0            2671.0   \n",
      "...                                ...             ...               ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC            NaN             NaN               NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA            NaN             NaN               NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA            NaN             NaN               NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA            NaN             NaN               NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA            NaN             NaN               NaN   \n",
      "\n",
      "                         rna:nCount_HTO  rna:nFeature_HTO  rna:nCount_GDO  \\\n",
      "index                                                                       \n",
      "rna:l1_AAACCTGAGCCAGAAC            99.0               4.0           576.0   \n",
      "rna:l1_AAACCTGAGTGGACGT            35.0               5.0           190.0   \n",
      "rna:l1_AAACCTGCATGAGCGA            66.0               4.0           212.0   \n",
      "rna:l1_AAACCTGTCTTGTCAT            22.0               3.0           243.0   \n",
      "rna:l1_AAACGGGAGAACAACT            99.0               5.0           198.0   \n",
      "...                                 ...               ...             ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA             NaN               NaN             NaN   \n",
      "\n",
      "                         rna:nCount_ADT  rna:nFeature_ADT  rna:percent.mito  \\\n",
      "index                                                                         \n",
      "rna:l1_AAACCTGAGCCAGAAC           801.0               4.0          2.295577   \n",
      "rna:l1_AAACCTGAGTGGACGT           545.0               4.0          4.512939   \n",
      "rna:l1_AAACCTGCATGAGCGA           344.0               4.0          4.116413   \n",
      "rna:l1_AAACCTGTCTTGTCAT           539.0               4.0          5.491723   \n",
      "rna:l1_AAACGGGAGAACAACT          1053.0               4.0          3.359868   \n",
      "...                                 ...               ...               ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC             NaN               NaN               NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA             NaN               NaN               NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA             NaN               NaN               NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA             NaN               NaN               NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA             NaN               NaN               NaN   \n",
      "\n",
      "                        rna:MULTI_ID rna:HTO_classification rna:guide_ID  \\\n",
      "index                                                                      \n",
      "rna:l1_AAACCTGAGCCAGAAC      rep1-tx                rep1-tx      STAT2g2   \n",
      "rna:l1_AAACCTGAGTGGACGT      rep1-tx                rep1-tx       CAV1g4   \n",
      "rna:l1_AAACCTGCATGAGCGA      rep1-tx                rep1-tx      STAT1g2   \n",
      "rna:l1_AAACCTGTCTTGTCAT      rep1-tx                rep1-tx       CD86g1   \n",
      "rna:l1_AAACGGGAGAACAACT      rep1-tx                rep1-tx       IRF7g2   \n",
      "...                              ...                    ...          ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC          NaN                    NaN          NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA          NaN                    NaN          NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA          NaN                    NaN          NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA          NaN                    NaN          NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA          NaN                    NaN          NaN   \n",
      "\n",
      "                        rna:gene_target   rna:NT rna:perturbation  \\\n",
      "index                                                               \n",
      "rna:l1_AAACCTGAGCCAGAAC           STAT2  STAT2g2        Perturbed   \n",
      "rna:l1_AAACCTGAGTGGACGT            CAV1   CAV1g4        Perturbed   \n",
      "rna:l1_AAACCTGCATGAGCGA           STAT1  STAT1g2        Perturbed   \n",
      "rna:l1_AAACCTGTCTTGTCAT            CD86   CD86g1        Perturbed   \n",
      "rna:l1_AAACGGGAGAACAACT            IRF7   IRF7g2        Perturbed   \n",
      "...                                 ...      ...              ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC             NaN      NaN              NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA             NaN      NaN              NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA             NaN      NaN              NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA             NaN      NaN              NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA             NaN      NaN              NaN   \n",
      "\n",
      "                        rna:replicate  rna:S.Score  rna:G2M.Score rna:Phase  \\\n",
      "index                                                                         \n",
      "rna:l1_AAACCTGAGCCAGAAC          rep1    -0.252716      -0.771309        G1   \n",
      "rna:l1_AAACCTGAGTGGACGT          rep1    -0.123802      -0.332603        G1   \n",
      "rna:l1_AAACCTGCATGAGCGA          rep1    -0.154633      -0.694418        G1   \n",
      "rna:l1_AAACCTGTCTTGTCAT          rep1    -0.061262      -0.037820        G1   \n",
      "rna:l1_AAACGGGAGAACAACT          rep1    -0.132188      -0.353156        G1   \n",
      "...                               ...          ...            ...       ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC           NaN          NaN            NaN       NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA           NaN          NaN            NaN       NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA           NaN          NaN            NaN       NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA           NaN          NaN            NaN       NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA           NaN          NaN            NaN       NaN   \n",
      "\n",
      "                        adt:orig.ident  adt:nCount_RNA  adt:nFeature_RNA  \\\n",
      "index                                                                      \n",
      "rna:l1_AAACCTGAGCCAGAAC            NaN             NaN               NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT            NaN             NaN               NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA            NaN             NaN               NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT            NaN             NaN               NaN   \n",
      "rna:l1_AAACGGGAGAACAACT            NaN             NaN               NaN   \n",
      "...                                ...             ...               ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC            NaN             NaN               NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA            NaN             NaN               NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA            NaN             NaN               NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA            NaN             NaN               NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA            NaN             NaN               NaN   \n",
      "\n",
      "                         adt:nCount_HTO  adt:nFeature_HTO  adt:nCount_GDO  \\\n",
      "index                                                                       \n",
      "rna:l1_AAACCTGAGCCAGAAC             NaN               NaN             NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT             NaN               NaN             NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA             NaN               NaN             NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT             NaN               NaN             NaN   \n",
      "rna:l1_AAACGGGAGAACAACT             NaN               NaN             NaN   \n",
      "...                                 ...               ...             ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA             NaN               NaN             NaN   \n",
      "\n",
      "                         adt:nCount_ADT  adt:nFeature_ADT  adt:percent.mito  \\\n",
      "index                                                                         \n",
      "rna:l1_AAACCTGAGCCAGAAC             NaN               NaN               NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT             NaN               NaN               NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA             NaN               NaN               NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT             NaN               NaN               NaN   \n",
      "rna:l1_AAACGGGAGAACAACT             NaN               NaN               NaN   \n",
      "...                                 ...               ...               ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC             NaN               NaN               NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA             NaN               NaN               NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA             NaN               NaN               NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA             NaN               NaN               NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA             NaN               NaN               NaN   \n",
      "\n",
      "                        adt:MULTI_ID adt:HTO_classification adt:guide_ID  \\\n",
      "index                                                                      \n",
      "rna:l1_AAACCTGAGCCAGAAC          NaN                    NaN          NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT          NaN                    NaN          NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA          NaN                    NaN          NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT          NaN                    NaN          NaN   \n",
      "rna:l1_AAACGGGAGAACAACT          NaN                    NaN          NaN   \n",
      "...                              ...                    ...          ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC          NaN                    NaN          NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA          NaN                    NaN          NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA          NaN                    NaN          NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA          NaN                    NaN          NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA          NaN                    NaN          NaN   \n",
      "\n",
      "                        adt:gene_target adt:NT adt:perturbation adt:replicate  \\\n",
      "index                                                                           \n",
      "rna:l1_AAACCTGAGCCAGAAC             NaN    NaN              NaN           NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT             NaN    NaN              NaN           NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA             NaN    NaN              NaN           NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT             NaN    NaN              NaN           NaN   \n",
      "rna:l1_AAACGGGAGAACAACT             NaN    NaN              NaN           NaN   \n",
      "...                                 ...    ...              ...           ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC             NaN    NaN              NaN           NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA             NaN    NaN              NaN           NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA             NaN    NaN              NaN           NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA             NaN    NaN              NaN           NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA             NaN    NaN              NaN           NaN   \n",
      "\n",
      "                         adt:S.Score  adt:G2M.Score adt:Phase hto:orig.ident  \\\n",
      "index                                                                          \n",
      "rna:l1_AAACCTGAGCCAGAAC          NaN            NaN       NaN            NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT          NaN            NaN       NaN            NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA          NaN            NaN       NaN            NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT          NaN            NaN       NaN            NaN   \n",
      "rna:l1_AAACGGGAGAACAACT          NaN            NaN       NaN            NaN   \n",
      "...                              ...            ...       ...            ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC          NaN            NaN       NaN            NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA          NaN            NaN       NaN            NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA          NaN            NaN       NaN            NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA          NaN            NaN       NaN            NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA          NaN            NaN       NaN            NaN   \n",
      "\n",
      "                         hto:nCount_RNA  hto:nFeature_RNA  hto:nCount_HTO  \\\n",
      "index                                                                       \n",
      "rna:l1_AAACCTGAGCCAGAAC             NaN               NaN             NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT             NaN               NaN             NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA             NaN               NaN             NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT             NaN               NaN             NaN   \n",
      "rna:l1_AAACGGGAGAACAACT             NaN               NaN             NaN   \n",
      "...                                 ...               ...             ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA             NaN               NaN             NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA             NaN               NaN             NaN   \n",
      "\n",
      "                         hto:nFeature_HTO  hto:nCount_GDO  hto:nCount_ADT  \\\n",
      "index                                                                       \n",
      "rna:l1_AAACCTGAGCCAGAAC               NaN             NaN             NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT               NaN             NaN             NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA               NaN             NaN             NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT               NaN             NaN             NaN   \n",
      "rna:l1_AAACGGGAGAACAACT               NaN             NaN             NaN   \n",
      "...                                   ...             ...             ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC               NaN             NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA               NaN             NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA               NaN             NaN             NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA               NaN             NaN             NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA               NaN             NaN             NaN   \n",
      "\n",
      "                         hto:nFeature_ADT  hto:percent.mito hto:MULTI_ID  \\\n",
      "index                                                                      \n",
      "rna:l1_AAACCTGAGCCAGAAC               NaN               NaN          NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT               NaN               NaN          NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA               NaN               NaN          NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT               NaN               NaN          NaN   \n",
      "rna:l1_AAACGGGAGAACAACT               NaN               NaN          NaN   \n",
      "...                                   ...               ...          ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC               NaN               NaN          NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA               NaN               NaN          NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA               NaN               NaN          NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA               NaN               NaN          NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA               NaN               NaN          NaN   \n",
      "\n",
      "                        hto:HTO_classification hto:guide_ID hto:gene_target  \\\n",
      "index                                                                         \n",
      "rna:l1_AAACCTGAGCCAGAAC                    NaN          NaN             NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT                    NaN          NaN             NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA                    NaN          NaN             NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT                    NaN          NaN             NaN   \n",
      "rna:l1_AAACGGGAGAACAACT                    NaN          NaN             NaN   \n",
      "...                                        ...          ...             ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC                    NaN          NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA                    NaN          NaN             NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA                    NaN          NaN             NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA                    NaN          NaN             NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA                    NaN          NaN             NaN   \n",
      "\n",
      "                        hto:NT hto:perturbation hto:replicate  hto:S.Score  \\\n",
      "index                                                                        \n",
      "rna:l1_AAACCTGAGCCAGAAC    NaN              NaN           NaN          NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT    NaN              NaN           NaN          NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA    NaN              NaN           NaN          NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT    NaN              NaN           NaN          NaN   \n",
      "rna:l1_AAACGGGAGAACAACT    NaN              NaN           NaN          NaN   \n",
      "...                        ...              ...           ...          ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC    NaN              NaN           NaN          NaN   \n",
      "gdo:l8_TTTGTCAGTGACGGTA    NaN              NaN           NaN          NaN   \n",
      "gdo:l8_TTTGTCAGTTCCACAA    NaN              NaN           NaN          NaN   \n",
      "gdo:l8_TTTGTCATCACGCATA    NaN              NaN           NaN          NaN   \n",
      "gdo:l8_TTTGTCATCTGTACGA    NaN              NaN           NaN          NaN   \n",
      "\n",
      "                         hto:G2M.Score hto:Phase gdo:orig.ident  \\\n",
      "index                                                             \n",
      "rna:l1_AAACCTGAGCCAGAAC            NaN       NaN            NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT            NaN       NaN            NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA            NaN       NaN            NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT            NaN       NaN            NaN   \n",
      "rna:l1_AAACGGGAGAACAACT            NaN       NaN            NaN   \n",
      "...                                ...       ...            ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC            NaN       NaN          Lane8   \n",
      "gdo:l8_TTTGTCAGTGACGGTA            NaN       NaN          Lane8   \n",
      "gdo:l8_TTTGTCAGTTCCACAA            NaN       NaN          Lane8   \n",
      "gdo:l8_TTTGTCATCACGCATA            NaN       NaN          Lane8   \n",
      "gdo:l8_TTTGTCATCTGTACGA            NaN       NaN          Lane8   \n",
      "\n",
      "                         gdo:nCount_RNA  gdo:nFeature_RNA  gdo:nCount_HTO  \\\n",
      "index                                                                       \n",
      "rna:l1_AAACCTGAGCCAGAAC             NaN               NaN             NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT             NaN               NaN             NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA             NaN               NaN             NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT             NaN               NaN             NaN   \n",
      "rna:l1_AAACGGGAGAACAACT             NaN               NaN             NaN   \n",
      "...                                 ...               ...             ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC         20863.0            4884.0           294.0   \n",
      "gdo:l8_TTTGTCAGTGACGGTA         17553.0            3787.0           528.0   \n",
      "gdo:l8_TTTGTCAGTTCCACAA         15106.0            4185.0           154.0   \n",
      "gdo:l8_TTTGTCATCACGCATA         11209.0            3204.0           132.0   \n",
      "gdo:l8_TTTGTCATCTGTACGA         23088.0            5073.0            95.0   \n",
      "\n",
      "                         gdo:nFeature_HTO  gdo:nCount_GDO  gdo:nCount_ADT  \\\n",
      "index                                                                       \n",
      "rna:l1_AAACCTGAGCCAGAAC               NaN             NaN             NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT               NaN             NaN             NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA               NaN             NaN             NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT               NaN             NaN             NaN   \n",
      "rna:l1_AAACGGGAGAACAACT               NaN             NaN             NaN   \n",
      "...                                   ...             ...             ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC               4.0           290.0           398.0   \n",
      "gdo:l8_TTTGTCAGTGACGGTA               3.0           870.0          3042.0   \n",
      "gdo:l8_TTTGTCAGTTCCACAA               6.0           267.0           212.0   \n",
      "gdo:l8_TTTGTCATCACGCATA               3.0           202.0           258.0   \n",
      "gdo:l8_TTTGTCATCTGTACGA               4.0           234.0           584.0   \n",
      "\n",
      "                         gdo:nFeature_ADT  gdo:percent.mito gdo:MULTI_ID  \\\n",
      "index                                                                      \n",
      "rna:l1_AAACCTGAGCCAGAAC               NaN               NaN          NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT               NaN               NaN          NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA               NaN               NaN          NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT               NaN               NaN          NaN   \n",
      "rna:l1_AAACGGGAGAACAACT               NaN               NaN          NaN   \n",
      "...                                   ...               ...          ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC               4.0          5.469012      rep3-tx   \n",
      "gdo:l8_TTTGTCAGTGACGGTA               4.0          2.159175      rep4-tx   \n",
      "gdo:l8_TTTGTCAGTTCCACAA               4.0          2.661194      rep3-tx   \n",
      "gdo:l8_TTTGTCATCACGCATA               4.0          7.369078      rep3-tx   \n",
      "gdo:l8_TTTGTCATCTGTACGA               4.0          2.191615      rep3-tx   \n",
      "\n",
      "                        gdo:HTO_classification gdo:guide_ID gdo:gene_target  \\\n",
      "index                                                                         \n",
      "rna:l1_AAACCTGAGCCAGAAC                    NaN          NaN             NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT                    NaN          NaN             NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA                    NaN          NaN             NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT                    NaN          NaN             NaN   \n",
      "rna:l1_AAACGGGAGAACAACT                    NaN          NaN             NaN   \n",
      "...                                        ...          ...             ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC                rep3-tx      CMTM6g1           CMTM6   \n",
      "gdo:l8_TTTGTCAGTGACGGTA                rep2-tx         NTg4              NT   \n",
      "gdo:l8_TTTGTCAGTTCCACAA                rep3-tx       ATF2g1            ATF2   \n",
      "gdo:l8_TTTGTCATCACGCATA                rep3-tx       CAV1g2            CAV1   \n",
      "gdo:l8_TTTGTCATCTGTACGA                rep3-tx      STAT3g1           STAT3   \n",
      "\n",
      "                          gdo:NT gdo:perturbation gdo:replicate  gdo:S.Score  \\\n",
      "index                                                                          \n",
      "rna:l1_AAACCTGAGCCAGAAC      NaN              NaN           NaN          NaN   \n",
      "rna:l1_AAACCTGAGTGGACGT      NaN              NaN           NaN          NaN   \n",
      "rna:l1_AAACCTGCATGAGCGA      NaN              NaN           NaN          NaN   \n",
      "rna:l1_AAACCTGTCTTGTCAT      NaN              NaN           NaN          NaN   \n",
      "rna:l1_AAACGGGAGAACAACT      NaN              NaN           NaN          NaN   \n",
      "...                          ...              ...           ...          ...   \n",
      "gdo:l8_TTTGTCAGTCACTTCC  CMTM6g1        Perturbed          rep3    -0.323562   \n",
      "gdo:l8_TTTGTCAGTGACGGTA       NT               NT          rep2    -0.153514   \n",
      "gdo:l8_TTTGTCAGTTCCACAA   ATF2g1        Perturbed          rep3    -0.191933   \n",
      "gdo:l8_TTTGTCATCACGCATA   CAV1g2        Perturbed          rep3    -0.134585   \n",
      "gdo:l8_TTTGTCATCTGTACGA  STAT3g1        Perturbed          rep3    -0.461981   \n",
      "\n",
      "                         gdo:G2M.Score gdo:Phase  \n",
      "index                                             \n",
      "rna:l1_AAACCTGAGCCAGAAC            NaN       NaN  \n",
      "rna:l1_AAACCTGAGTGGACGT            NaN       NaN  \n",
      "rna:l1_AAACCTGCATGAGCGA            NaN       NaN  \n",
      "rna:l1_AAACCTGTCTTGTCAT            NaN       NaN  \n",
      "rna:l1_AAACGGGAGAACAACT            NaN       NaN  \n",
      "...                                ...       ...  \n",
      "gdo:l8_TTTGTCAGTCACTTCC      -0.794679        G1  \n",
      "gdo:l8_TTTGTCAGTGACGGTA      -0.632655        G1  \n",
      "gdo:l8_TTTGTCAGTTCCACAA      -0.574283        G1  \n",
      "gdo:l8_TTTGTCATCACGCATA      -0.501513        G1  \n",
      "gdo:l8_TTTGTCATCTGTACGA      -1.071414        G1  \n",
      "\n",
      "[82916 rows x 76 columns] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbol\"\n",
      "col_cell_type=\"leiden\"\n",
      "col_sample_id=\"replicate\"\n",
      "col_batch=\"replicate\"\n",
      "col_subject=None\n",
      "col_condition=\"gene_target\"\n",
      "col_num_umis=None\n",
      "col_target_genes=\"gene_target\"\n",
      "col_perturbed=\"perturbation\"\n",
      "col_guide_rna=\"NT\"\n",
      "col_guide_split=\"guide_split\"\n",
      "key_control=\"NT\"\n",
      "key_treatment=\"Perturbed\"\n",
      "key_nonperturbed=\"NP\"\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 20729 × 18649\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'nCount_HTO', 'nFeature_HTO', 'nCount_GDO', 'nCount_ADT', 'nFeature_ADT', 'percent.mito', 'MULTI_ID', 'HTO_classification', 'guide_ID', 'gene_target', 'NT', 'perturbation', 'replicate', 'S.Score', 'G2M.Score', 'Phase'\n",
      "    var: 'name'\n",
      "    layers: 'counts'\n"
     ]
    }
   ],
   "source": [
    "self = Crispr(file_path, **kwargs_init)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing\n",
    "\n",
    "We will just use defaults for the preprocessing arguments (very little filtering, consistent with the Pertpy tutorial for this object)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "self.preprocess(kws_hvg=True, kws_scale=True, regress_out=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = self.cluster()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyses\n",
    "\n",
    "The following examples concern CRISPR or other perturbation design-specific analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mixscape: Cell-Level Perturbation Classification & Scoring\n",
    "\n",
    "**Is a perturbed cell detectibly perturbed, and to what extent?** Mixscape first calculates the \"**perturbation signature**\" by determining which control condition cells most closely resemble each perturbed cell in terms of mRNA expression and then subtracts the control expression from that of the perturbed cells' (i.e., centers perturbed cells' gene expression on their control neighbors).\n",
    "\n",
    "Then, it **identifies** and removes perturbed **cells with no detectible perturbation** (i.e., assigns them to predicted classes of perturbed versus not perturbed). You can then create visuals based on whether the cell is detectibly perturbed, \"non-perturbed\" (not detectibly perturbed), or control (no treatment). Optionally, you can visualize protein expression by this predicted class in certain multi-modal data.\n",
    "\n",
    "**Are there perturbation-specific clusters?** Mixscape uses linear discriminant analysis (LDA) to cluster cells that resemble each other in terms of gene expression and perturbation condition. _(LDA reduces dimensionality and attempts to maximize the separability of classes. Unperturbed cells are removed from analysis.)_ \n",
    "\n",
    "<u> __Features__ </u>  \n",
    "\n",
    "- Plot targeting efficiency.\n",
    "- Remove confounds (e.g., cell cycle, batch effects)\n",
    "- Classify cells as affected or unaffected (i.e., \"escapees\") by the perturbation\n",
    "- Quantify and visualize degree of perturbation response\n",
    "\n",
    "<u> __Input__ </u> \n",
    "\n",
    "See documentation, but the key arguments are listed here.\n",
    "\n",
    "* **col_cell_type**: If you want to run using a different cell classification column, (e.g., CellTypist annotations that weren't used for the original `self._columns[\"col_cell_type\"]``), you can specify a different column by passing `col_cell_type=<column name>` if you'd like.\n",
    "* **target_gene_idents**: A list of gene symbols to focus on in plots/analyses. Specify as True to include all.\n",
    "* **target_gene_idents**: The default layer of data used is \"log1p.\" Remember that Mixscape centers cells on their control neighbors when considering whether to use centered and/or scaled data.\n",
    "* **protein_of_interest**: If you have protein expression data (e.g., CITE-seq), you can specify a protein whose expression to plot against the perturbation conditions.\n",
    "\n",
    "<u> __Output__ </u>  \n",
    "\n",
    "Assuming your `Crispr` object is named \"self\":\n",
    "- Targeting Efficiency: `self.figures[\"mixscape\"][\"targeting_efficiency\"]`\n",
    "- Differential Expression Ordered by Posterior Probabilities: `self.figures[\"mixscape\"][\"DEX_ordered_by_ppp_heat\"]`\n",
    "- Posterior Probabilities Violin Plot: `self.figures[\"mixscape\"][\"ppp_violin\"]`\n",
    "- Perturbation Scores: `self.figures[\"mixscape\"][\"perturbation_score\"]`\n",
    "- Perturbation Clusters (from LDA): `self.figures[\"mixscape\"][\"perturbation_clusters\"]`\n",
    "\n",
    "The above instructions are for accessing output via the object attributes. Assuming output is assigned to a variable `figs` (i.e., `figs = ` would replace the `_ = ` in the code below), replace `self.figures[\"mixscape\"]` in the above code with `figs`.\n",
    "\n",
    "<u> __Notes__ </u>  \n",
    "\n",
    "- If `._columns[\"col_sample_id\"]` is not None, perturbation scores will by default be calculated and/or plotted taking that into consideration (e.g., biological replicates) unless `col_split_by=False`. That argument can also be set to a different column name explicitly, in which case that specification will be used as the `col_split_by` argument in Pertpy Mixscape functions in place of sample ID."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run Mixscape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "target_gene_idents = [\"NT\",\"JAK2\",\"STAT1\",\"IFNGR1\",\"IFNGR2\", \"IRF1\"]\n",
    "_ = self.run_mixscape(target_gene_idents=target_gene_idents,\n",
    "                      protein_of_interest=\"PDL1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create Different Mixscape Plots\n",
    "\n",
    "If you want to re-create mixscape **plots with <u> different target genes and/or proteins of interest**</u> later, you can use `self.plot_mixscape(<ONE OR MORE TARGET GENES>)`. If you want a different color for the perturbation score curves, specify `color=` in that method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tgis = pd.Series(self.rna.uns[\"mixscape\"].keys()).sample(1)\n",
    "_ = self.plot_mixscape(tgis, color=\"red\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py-bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
