{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports and Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #808000; text-decoration-color: #808000; font-weight: bold\">ryp2 is not installed. Install with </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">pip install rpy2 </span><span style=\"color: #808000; text-decoration-color: #808000; font-weight: bold\">to run tools with R support.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;33mryp2 is not installed. Install with \u001b[0m\u001b[1;32mpip install rpy2 \u001b[0m\u001b[1;33mto run tools with R support.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading data from `https://omnipathdb.org/queries/enzsub?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/interactions?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/complexes?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/annotations?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/intercell?format=json`\n",
      "Downloading data from `https://omnipathdb.org/about?format=text`\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import re\n",
    "import json\n",
    "from datetime import datetime\n",
    "import functools\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sb\n",
    "from anndata import AnnData\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import corescpy as cr\n",
    "\n",
    "pd.options.display.max_columns = 100\n",
    "pd.options.display.max_rows = 200\n",
    "sc.settings.set_figure_params(dpi=100, frameon=False, figsize=(20, 20))\n",
    "\n",
    "palette = \"tab20\"\n",
    "\n",
    "\n",
    "def construct_file(sample, slide, date=None, timestamp=None,\n",
    "                   panel_id=\"TUQ97N\", prefix=\"output-XETG00189\",\n",
    "                   project_owner=\"EA\", run=\"CHO-001\", directory=None):\n",
    "    \"\"\"Construct file path from information.\"\"\"\n",
    "    if isinstance(sample, str):\n",
    "        sample = [sample]\n",
    "    if \"outputs\" not in directory and os.path.exists(\n",
    "            os.path.join(directory, \"outputs\")):\n",
    "        directory = os.path.join(directory, \"outputs\")\n",
    "    print(directory)\n",
    "    panel_id, prefix, project_owner, slide, date, timestamp = [\n",
    "        [x] * len(sample) if isinstance(x, str) else list(x) if x else x\n",
    "        for x in [panel_id, prefix, project_owner, slide, date, timestamp]]\n",
    "    run = [run] * len(sample) if isinstance(run, (str, int, float)) else run\n",
    "    block = [\"-\".join(i) for i in zip(sample, panel_id, project_owner)]\n",
    "    fff = [f\"{prefix[i]}__{slide[i]}__{block[i]}\" for i in range(len(sample))]\n",
    "    if date is None or timestamp is None:\n",
    "        for i, x in enumerate(fff):  # iterate current file stems\n",
    "            ddd = os.path.join(directory, panel_id[i], run[i])\n",
    "            print(ddd)\n",
    "            matches = sum([x in d for d in os.listdir(ddd)])\n",
    "            if  matches != 1:\n",
    "                raise ValueError(f\"{x} found in 0 or multiple file paths\",\n",
    "                                 f\"\\n\\n{os.listdir(ddd)}\")\n",
    "            fff[i] = os.path.join(ddd, np.array(os.listdir(ddd))[np.where([\n",
    "                x in d for d in os.listdir(ddd)])[0][0]])  # find match\n",
    "    else:\n",
    "        fff = [os.path.join(directory, panel_id[i], run[i],\n",
    "                            f\"{x}__{date[i]}__{timestamp[i]}\")\n",
    "               for i, x in enumerate(fff)]\n",
    "    return fff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs\n",
      "/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/CHO-001\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Status</th>\n",
       "      <th>Project</th>\n",
       "      <th>Sample Location</th>\n",
       "      <th>Inflamed</th>\n",
       "      <th>Slide</th>\n",
       "      <th>GRID ID</th>\n",
       "      <th>Procedure Date</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Hispanic</th>\n",
       "      <th>Diagnosis</th>\n",
       "      <th>Project.1</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Disease Status</th>\n",
       "      <th>Date Collected</th>\n",
       "      <th>Date Sectioned</th>\n",
       "      <th>Date of Hybridization</th>\n",
       "      <th>Storage in 4C</th>\n",
       "      <th>Created By</th>\n",
       "      <th>Created</th>\n",
       "      <th>Storage Status</th>\n",
       "      <th>Sample Location.1</th>\n",
       "      <th>Storage Row</th>\n",
       "      <th>Storage Col</th>\n",
       "      <th>Checked Out By</th>\n",
       "      <th>Stricture</th>\n",
       "      <th>Condition</th>\n",
       "      <th>file_path</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Uninflamed-50452A</th>\n",
       "      <td>50452A</td>\n",
       "      <td>50452</td>\n",
       "      <td>Available</td>\n",
       "      <td>scRNA &amp; Spatial</td>\n",
       "      <td>Terminal Ileum</td>\n",
       "      <td>uninflamed</td>\n",
       "      <td>0010700</td>\n",
       "      <td>1011335</td>\n",
       "      <td>2023-12-04</td>\n",
       "      <td>69</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>NO</td>\n",
       "      <td>Crohn's Disease</td>\n",
       "      <td>Helmsley-Senescence</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2024-01-22 15:25:38</td>\n",
       "      <td>2024-01-23 15:25:58</td>\n",
       "      <td>YES</td>\n",
       "      <td>ksenija sabic</td>\n",
       "      <td>2024-02-06 15:26:47.887</td>\n",
       "      <td>Not in storage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>Uninflamed</td>\n",
       "      <td>/mnt/cho_lab/disk2/elizabeth/data/shared-xeniu...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Sample ID  Patient     Status          Project  \\\n",
       "Sample                                                             \n",
       "Uninflamed-50452A    50452A    50452  Available  scRNA & Spatial   \n",
       "\n",
       "                  Sample Location    Inflamed    Slide  GRID ID  \\\n",
       "Sample                                                            \n",
       "Uninflamed-50452A  Terminal Ileum  uninflamed  0010700  1011335   \n",
       "\n",
       "                  Procedure Date  Age     Sex   Race Hispanic  \\\n",
       "Sample                                                          \n",
       "Uninflamed-50452A     2023-12-04   69  female  white       NO   \n",
       "\n",
       "                         Diagnosis            Project.1  Procedure  \\\n",
       "Sample                                                               \n",
       "Uninflamed-50452A  Crohn's Disease  Helmsley-Senescence  Resection   \n",
       "\n",
       "                  Disease Status  Date Collected      Date Sectioned  \\\n",
       "Sample                                                                 \n",
       "Uninflamed-50452A         Active             NaN 2024-01-22 15:25:38   \n",
       "\n",
       "                  Date of Hybridization Storage in 4C     Created By  \\\n",
       "Sample                                                                 \n",
       "Uninflamed-50452A   2024-01-23 15:25:58           YES  ksenija sabic   \n",
       "\n",
       "                                  Created  Storage Status  Sample Location.1  \\\n",
       "Sample                                                                         \n",
       "Uninflamed-50452A 2024-02-06 15:26:47.887  Not in storage                NaN   \n",
       "\n",
       "                   Storage Row  Storage Col  Checked Out By Stricture  \\\n",
       "Sample                                                                  \n",
       "Uninflamed-50452A          NaN          NaN             NaN      None   \n",
       "\n",
       "                    Condition  \\\n",
       "Sample                          \n",
       "Uninflamed-50452A  Uninflamed   \n",
       "\n",
       "                                                           file_path  \n",
       "Sample                                                                \n",
       "Uninflamed-50452A  /mnt/cho_lab/disk2/elizabeth/data/shared-xeniu...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Column Names\n",
    "col_sample_id_original = \"Sample ID\"\n",
    "col_sample_id = \"Sample\"\n",
    "col_subject = \"Patient\"\n",
    "col_batch = \"Slide\"\n",
    "col_path = \"file_path\"\n",
    "# col_date_time, col_date = \"Date Sectioned\", \"Date\"\n",
    "col_date_time, col_date = None, None\n",
    "col_inflamed, col_stricture = \"Inflamed\", \"Stricture\"\n",
    "col_condition = \"Condition\"\n",
    "meta_rn = {\"Name\": col_subject, \"Slide ID\": col_batch,\n",
    "           \"Inflammation Status\": col_inflamed}\n",
    "\n",
    "# Directories & Metadata\n",
    "# Replace manually or mirror my file/directory tree in your home\n",
    "include_stricture = True\n",
    "run = \"CHO-001\"\n",
    "samples = [\"50452A\", \"50452B\", \"50452C\"]\n",
    "# samples = [\"50452A\", \"50452B\"]\n",
    "# samples = [\"50452A\", \"50452B\", \"50452C\", \"50564A4\", \"50618B5\"]\n",
    "# run = [\"CHO-001\"] * 3 + [\"CHO-002\"] * 3\n",
    "panel_id = \"TUQ97N\"\n",
    "prefix = \"output-XETG00189\"\n",
    "project_owner = \"EA\"\n",
    "ddu = os.path.expanduser(\"~\")\n",
    "ddl = \"/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library\"\n",
    "ddd = os.path.join(ddl, \"outputs\", panel_id)\n",
    "panel = os.path.join(ddu, \"projects/senescence/ProposedGenePanel.xlsx\")\n",
    "file_ann = os.path.join(ddu, \"corescpy/examples/annotation_guide.xlsx\")\n",
    "\n",
    "# Input/Output Options\n",
    "reload = True\n",
    "out_dir = os.path.join(ddd, \"nebraska\")  # set to None to avoid saving\n",
    "if not os.path.exists(out_dir):\n",
    "    os.makedirs(out_dir)\n",
    "\n",
    "# Computing Resources\n",
    "gpu = False\n",
    "sc.settings.n_jobs = 4\n",
    "sc.settings.max_memory = 150\n",
    "\n",
    "# Read Metadata & Other Information\n",
    "annot_df = pd.read_excel(file_ann)\n",
    "metadata = pd.read_excel(os.path.join(ddl, \"Xenium_Samples_02092024.xlsx\"),\n",
    "                         dtype={\"Slide ID\": str})\n",
    "metadata = metadata.rename(meta_rn, axis=1)\n",
    "if samples not in [\"all\", None]:  # subset by sample ID?\n",
    "    metadata = metadata.set_index(col_sample_id_original).loc[\n",
    "        samples].reset_index()\n",
    "\n",
    "# Processing & Clustering Options\n",
    "resolution = 0.5\n",
    "min_dist = 1\n",
    "n_comps = 20\n",
    "# col_qscore = ?\n",
    "# custom_thresholds = {col_qscore: [, None]}\n",
    "custom_thresholds = None\n",
    "genes_subset = list(annot_df.iloc[:, 0])\n",
    "kws_pp = dict(cell_filter_pmt=None, cell_filter_ncounts=[50, None],\n",
    "              cell_filter_ngene=[30, None], gene_filter_ncell=[3, None],\n",
    "              gene_filter_ncounts=[3, None],\n",
    "              custom_thresholds=custom_thresholds,\n",
    "              kws_scale=dict(max_value=10, zero_center=True))\n",
    "kws_umap = dict(min_dist=min_dist, method=\"rapids\") if gpu is True else dict(\n",
    "    min_dist=min_dist)\n",
    "kws_cluster = dict(use_gpu=gpu, kws_umap=kws_umap, kws_neighbors=None,\n",
    "                   use_highly_variable=False, n_comps=n_comps,\n",
    "                   genes_subset=genes_subset, resolution=resolution)\n",
    "# kws_subcluster = dict(use_gpu=gpu, kws_umap=dict(min_dist=0.3),\n",
    "#                       kws_neighbors=None, use_highly_variable=False,\n",
    "#                       n_comps=n_comps, genes_subset=genes_subset,\n",
    "#                       method_cluster=\"leiden\", resolution=0.5)\n",
    "kws_subcluster = dict(method_cluster=\"leiden\", resolution=0.5)\n",
    "\n",
    "# Revise Metadata & Construct Variables from Options\n",
    "if col_stricture not in metadata.columns:\n",
    "    metadata.loc[:, col_stricture] = metadata[\"Sample Location\"].apply(\n",
    "        lambda x: \"Stricture\" if \"stricture\" in x.lower() else \"None\")\n",
    "metadata.loc[:, col_condition] = metadata.apply(\n",
    "    lambda x: \"Stricture\" if \"stricture\" in x[col_stricture].lower() else x[\n",
    "        col_inflamed].capitalize() , axis=1)\n",
    "if col_date_time:\n",
    "    metadata.loc[:, col_date] = metadata[col_date_time].apply(\n",
    "        lambda x: datetime.strftime(x, \"%Y%m%d\"))\n",
    "    dates = list(metadata[col_date])\n",
    "else:\n",
    "    dates = None\n",
    "metadata.loc[:, col_sample_id] = metadata.apply(\n",
    "    lambda x: f\"{x[col_condition]}-{x[col_sample_id_original]}\" , axis=1)\n",
    "metadata = metadata.set_index(col_sample_id)\n",
    "metadata.loc[:, col_path] = construct_file(\n",
    "    list(metadata[col_sample_id_original]), list(metadata[col_batch]),\n",
    "    dates, panel_id=panel_id, prefix=prefix,\n",
    "    project_owner=project_owner, run=run, directory=ddl)\n",
    "col_cell_type = \"Annotation\"\n",
    "file_path_dict = dict(zip(metadata.index.values, metadata[\"file_path\"]))\n",
    "kws_init = dict(col_batch=col_batch, col_subject=col_subject,\n",
    "                col_sample_id=col_sample_id, col_cell_type=col_cell_type)\n",
    "\n",
    "# Annotation File\n",
    "col_assignment = \"Bin\"\n",
    "assign = pd.read_excel(file_ann, index_col=0).dropna(subset=col_assignment)\n",
    "sources = assign[col_assignment].unique()\n",
    "rename = dict(zip(sources, [\" \".join([i.capitalize() if i and i[\n",
    "    0] != \"(\" and not i.isupper() and i not in [\n",
    "        \"IgG\", \"IgA\"] else i for i in x.split(\" \")]) if len(x.split(\n",
    "            \" \")) > 1 else x for x in [re.sub(\"glia\", \"Glia\", re.sub(\n",
    "                \"_\", \" \", j)) for j in sources]]))\n",
    "assign.loc[:, col_assignment] = assign[col_assignment].replace(rename)\n",
    "assign = assign.rename_axis(\"Gene\")\n",
    "marker_genes_dict = dict(assign.reset_index().groupby(col_assignment).apply(\n",
    "    lambda x: list(pd.unique(x.Gene))))  # to marker dictionary\n",
    "\n",
    "# Subset if Desired\n",
    "if include_stricture is False:\n",
    "    metadata = metadata[metadata.Stricture != \"Stricture\"]\n",
    "metadata\n",
    "\n",
    "# Genes\n",
    "# genes = [\"CDKN1A\", \"CDKN2A\", \"TP53\", \"PLAUR\", \"PTGER4\", \"FTL\", \"IL6ST\"]\n",
    "# cell_types = [\"ILC3\", \"LTi-like NCR+ ILC3\", \"LTi-like NCR- ILC3\",\n",
    "#               \"ILCP\", \"Macrophages\", \"Stem cells\"]\n",
    "# palette = [\"r\", \"tab:pink\", \"m\", \"b\", \"tab:brown\", \"cyan\"]\n",
    "# High in inf. vs. un\n",
    "# OSM\n",
    "# IL13\n",
    "# IL1B\n",
    "# IL6\n",
    "# TNF\n",
    "# S100A8\n",
    "# S100A9\n",
    "# ------------------------------\n",
    "# High in stricture vs inf/un\n",
    "# PDGFRA\n",
    "# IL6ST\n",
    "# PTPN1\n",
    "# IFNG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "<<< INITIALIZING SPATIAL CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "<<< INITIALIZING OMICS CLASS OBJECT >>>\n",
      "\n",
      "Unused keyword arguments: {'library_id': 'Uninflamed-50452A', 'visium': False}.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\u001b[34mINFO    \u001b[0m reading                                                                                                   \n",
      "         \u001b[35m/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/CHO-001/output-XETG00189__0010700__\u001b[0m\n",
      "         \u001b[35m50452A-TUQ97N-EA__20240126__205019/\u001b[0m\u001b[95mcell_feature_matrix.h5\u001b[0m                                                 \n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "dsfasfda AnnData object with n_obs × n_vars = 333825 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Cell Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "Observations: 333825\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Gene Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "Empty DataFrame\n",
      "Columns: [index, gene_ids, feature_types, genome]\n",
      "Index: []\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " SpatialData object with:\n",
      "├── Images\n",
      "│     ├── 'morphology_focus': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "│     └── 'morphology_mip': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 10) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (333825, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "└── Table\n",
      "      └── AnnData object with n_obs × n_vars = 333825 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts': AnnData (333825, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), morphology_mip (Images), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "None\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 333825 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts' \n",
      "\n",
      "                      gene_ids    feature_types   genome\n",
      "gene_symbols                                           \n",
      "ABCA7         ENSG00000064687  Gene Expression  Unknown\n",
      "ACTA2         ENSG00000107796  Gene Expression  Unknown\n",
      "ACTG2         ENSG00000163017  Gene Expression  Unknown\n",
      "ADAM15        ENSG00000143537  Gene Expression  Unknown\n",
      "ADAM28        ENSG00000042980  Gene Expression  Unknown \n",
      "\n",
      "       cell_id  transcript_counts  control_probe_counts  \\\n",
      "0  aaaaibki-1                168                     0   \n",
      "1  aaaaiiji-1                166                     0   \n",
      "2  aaaajape-1                135                     0   \n",
      "3  aaaamhbm-1                 99                     0   \n",
      "4  aaabdoak-1                 61                     0   \n",
      "\n",
      "   control_codeword_counts  unassigned_codeword_counts  \\\n",
      "0                        0                           0   \n",
      "1                        0                           0   \n",
      "2                        0                           0   \n",
      "3                        0                           0   \n",
      "4                        0                           0   \n",
      "\n",
      "   deprecated_codeword_counts  total_counts   cell_area  nucleus_area  \\\n",
      "0                           0           168  328.150481     13.908126   \n",
      "1                           0           166  107.607348     18.649532   \n",
      "2                           0           135  110.226410     10.250469   \n",
      "3                           0            99   46.736720     16.482032   \n",
      "4                           0            61   22.216876      6.954063   \n",
      "\n",
      "         region             Sample  \n",
      "0  cell_circles  Uninflamed-50452A  \n",
      "1  cell_circles  Uninflamed-50452A  \n",
      "2  cell_circles  Uninflamed-50452A  \n",
      "3  cell_circles  Uninflamed-50452A  \n",
      "4  cell_circles  Uninflamed-50452A  \n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "col_segment=\"region\"\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\n",
      "\n",
      "<<< RESTORING SINGLE-SAMPLE FROM h5ad >>>\n",
      "\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "SpatialData object with:\n",
      "├── Images\n",
      "│     ├── 'morphology_focus': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "│     └── 'morphology_mip': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 10) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (333825, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "└── Table\n",
      "      └── AnnData object with n_obs × n_vars = 312393 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample', 'Sample ID', 'Patient', 'Status', 'Project', 'Sample Location', 'Inflamed', 'Slide', 'GRID ID', 'Procedure Date', 'Age', 'Sex', 'Race', 'Hispanic', 'Diagnosis', 'Project.1', 'Procedure', 'Disease Status', 'Date Collected', 'Date Sectioned', 'Date of Hybridization', 'Storage in 4C', 'Created By', 'Created', 'Storage Status', 'Sample Location.1', 'Storage Row', 'Storage Col', 'Checked Out By', 'Stricture', 'Condition', 'file_path', 'n_counts', 'log_counts', 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'assay_protein', 'col_gene_symbols', 'col_cell_type', 'col_sample_id', 'col_batch', 'col_subject', 'col_condition', 'col_num_umis', 'col_segment', 'cell_filter_pmt', 'cell_filter_ncounts', 'cell_filter_ngene', 'gene_filter_ncell', 'gene_filter_ncounts', 'custom_thresholds', 'figsize', 'kws_scale', 'leiden', 'method_cluster', 'kws_pca', 'kws_umap', 'kws_neighbors', 'kws_cluster', 'resolution', 'genes_subset', 'Annotation'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
      "    uns: 'Annotation_colors', 'Patient_colors', 'Sample_colors', 'Slide_colors', 'hvg', 'leiden', 'leiden_colors', 'log1p', 'neighbors', 'original_ix', 'pca', 'rank_genes_groups', 'rank_genes_groups_leiden', 'spatialdata_attrs', 'umap', 'spatial', 'markers'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    layers: 'counts', 'log1p', 'scaled'\n",
      "    obsp: 'connectivities', 'distances': AnnData (312393, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), morphology_mip (Images), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "CPU times: user 1min 7s, sys: 4.92 s, total: 1min 12s\n",
      "Wall time: 1min 15s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Load Spatial Data\n",
    "suff = str(f\"res{re.sub('[.]', 'pt', str(resolution))}_dist\"\n",
    "           f\"{re.sub('[.]', 'pt', str(min_dist))}_npc{n_comps}\")  # file end\n",
    "selves, paths_he, file_mks, out_files = [], [], [], []\n",
    "for x in metadata.index.values:\n",
    "    self = cr.Spatial(metadata.loc[x][col_path], library_id=x, **kws_init)\n",
    "    for i in metadata:  # add metadata for subject\n",
    "        self.rna.obs.loc[:, i] = str(metadata.loc[x][i])  # add metadata\n",
    "    selves += [self]\n",
    "    paths_he += [os.path.join(metadata.loc[x][\n",
    "        col_path], \"aux_outputs/image_he.ome.tif\")]  # H&E paths\n",
    "        # out_files += [os.path.join(out_dir, f\"{x}.zarr\")]\n",
    "    if out_dir is not None:\n",
    "        out_files += [os.path.join(out_dir, f\"{x}__{suff}.zarr\")]\n",
    "    file_mks += [os.path.join(out_dir, f\"{x}__{suff}_markers.csv\")]\n",
    "\n",
    "# Reload Processed & Clustered Data (Optionally)\n",
    "if reload is True:\n",
    "    for i, s in enumerate(selves):\n",
    "        mks = file_mks[i] if os.path.exists(file_mks[i]) else None\n",
    "        s.update_from_h5ad(file=out_files[i], file_path_markers=mks)\n",
    "        print(s.adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if reload is True:\n",
    "    for i, s in enumerate(selves):\n",
    "        s.plot_spatial(color=\"leiden\")\n",
    "        # s.plot_spatial(color=col_cell_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotation & Sub-Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in selves:\n",
    "    _ = s.cluster(**kws_cluster)\n",
    "    _ = s.annotate_clusters(file_ann, col_cell_type=\"leiden\",\n",
    "                            col_annotation=col_cell_type,\n",
    "                            col_assignment=col_assignment)  # annotate & write\n",
    "    s.plot_spatial(color=col_cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cat_epi = [\"BEST2+ Goblet Cell\", \"BEST4+ Epithelial\", \"EC Cells (NPW+)\",\n",
    "#            \"EC Cells (TAC1+)\", \"Enterocyte\", \"Epithelial\", \"Goblet Cell\",\n",
    "#            \"I Cells (CCK+)\", \"Paneth\", \"D Cells (SST+)\",\n",
    "#            \"Stem Cells\", \"N Cells (NTS+)\", \"K Cells (GIP+)\",\n",
    "#            # \"Colonocyte\",\n",
    "#            \"L Cells (PYY+)\", \"Tuft\", \"Microfold Cell\"]\n",
    "# cat_epi = list(np.array()[np.where([\"epithelial\" in x.lower()\n",
    "#                                     for x in marker_genes_dict])[0]])\n",
    "# file_ann_epi = pd.concat([pd.Series(marker_genes_dict[x]) for x in cat_epi],\n",
    "#                          keys=cat_epi, names=[\"Type\"]).to_frame(\n",
    "#                              \"Gene\").reset_index(0).set_index(\"Gene\")\n",
    "file_ann_epi = file_ann\n",
    "# key_cell_type, col_ann = [\"7\"], \"Subclustering_Epi\"\n",
    "key_cell_type, col_ann = None, \"leiden_sub\"\n",
    "for s in selves:\n",
    "    ann = s.subcluster(restrict_to=(col_cell_type, key_cell_type), copy=False,\n",
    "                       key_added=\"leiden_sub\", **kws_subcluster)\n",
    "    s.annotate_clusters(kws_annotation={\"model\": file_ann_epi,\n",
    "                                        \"col_assignment\": \"Bin\"},\n",
    "                        col_annotation=[col_ann, col_cell_type])\n",
    "    s.plot_spatial(color=col_ann)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in selves:\n",
    "    s.write(out_files[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 265248 × 33234\n",
       "    obs: 'donor_id', 'Diagnosis', 'Fraction', 'batch', 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'doublet_scores', 'predicted_doublets', 'category', 'Age_group', 'author_cell_type', 'assay_ontology_term_id', 'cell_type_ontology_term_id', 'development_stage_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'sex_ontology_term_id', 'tissue_ontology_term_id', 'suspension_type', 'hca_data_portal_donor_uuid', 'hca_data_portal_cellsuspension_uuid', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage'\n",
       "    var: 'gene_ids', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'feature_types', 'feature_is_filtered', 'gene_symbols', 'feature_reference', 'feature_biotype'\n",
       "    uns: 'Integrated_colors', 'category2_colors', 'category_colors', \"dendrogram_['Integrated']\", \"dendrogram_['Integrated_05']\", 'hvg', 'neighbors', 'pca', 'schema_version', 'title', 'umap', 'rank_genes_groups_cell_type'\n",
       "    obsm: 'X_pca', 'X_umap'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workspace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Default of the method has been changed to 't-test' from 't-test_overestim_var'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "465 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "Allocate tensors for mapping.\n",
      "Begin training with 100 genes and uniform density_prior in clusters mode...\n",
      "Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.328, KL reg: 0.054\n",
      "Score: 0.605, KL reg: 0.004\n",
      "Score: 0.612, KL reg: 0.004\n",
      "Score: 0.614, KL reg: 0.004\n",
      "Score: 0.615, KL reg: 0.004\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving results..\n",
      "spatial prediction dataframe is saved in `obsm` `tangram_ct_pred` of the spatial AnnData.\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "The two AnnDatas need to have same `obs` index.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 11\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# sdata = selves[0].adata\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# # adata_sc = sc.read(os.path.join(\"/mnt/cho_lab/disk2/elizabeth/data\",\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# #                                 \"2023-05-12_CombinedCD-v2_new.h5ad\"))\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;66;03m# adata_sc = adata_sc[adata_sc.obs.tissue == \"small intestine\"]\u001b[39;00m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m# adata_sc.var = adata_sc.var.reset_index().set_index(\"feature_name\")\u001b[39;00m\n\u001b[0;32m---> 11\u001b[0m sdata_new, sdata, adata_sc, ad_map, df_compare \u001b[38;5;241m=\u001b[39m \u001b[43mcr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mintegrate_spatial\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[43m    \u001b[49m\u001b[43msdata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43madata_sc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcol_cell_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcell_type\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mclusters\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     13\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkey_added\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrank_genes_groups_cell_type\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/corescpy/corescpy/processing/spatial_pp.py:213\u001b[0m, in \u001b[0;36mintegrate_spatial\u001b[0;34m(adata_sp, adata_sc, col_cell_type, markers, gene_to_lowercase, num_epochs, device, density_prior, mode, plot, plot_genes, seed, inplace, **kwargs)\u001b[0m\n\u001b[1;32m    209\u001b[0m     adata_sp\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrna_count_based_density\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m ct_spot \u001b[38;5;241m/\u001b[39m np\u001b[38;5;241m.\u001b[39msum(ct_spot)\n\u001b[1;32m    210\u001b[0m ad_map \u001b[38;5;241m=\u001b[39m tg\u001b[38;5;241m.\u001b[39mmap_cells_to_space(\n\u001b[1;32m    211\u001b[0m     adata_sc, adata_sp, mode\u001b[38;5;241m=\u001b[39mmode, device\u001b[38;5;241m=\u001b[39mdevice, num_epochs\u001b[38;5;241m=\u001b[39mnum_epochs,\n\u001b[1;32m    212\u001b[0m     density_prior\u001b[38;5;241m=\u001b[39mdensity_prior, random_state\u001b[38;5;241m=\u001b[39mseed, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# mapping\u001b[39;00m\n\u001b[0;32m--> 213\u001b[0m tg\u001b[38;5;241m.\u001b[39mproject_cell_annotations(\n\u001b[1;32m    214\u001b[0m     ad_map, adata_sp, annotation\u001b[38;5;241m=\u001b[39mcol_cell_type)  \u001b[38;5;66;03m# clusters -> space\u001b[39;00m\n\u001b[1;32m    215\u001b[0m adata_sp_new \u001b[38;5;241m=\u001b[39m tg\u001b[38;5;241m.\u001b[39mproject_genes(adata_map\u001b[38;5;241m=\u001b[39mad_map, adata_sc\u001b[38;5;241m=\u001b[39madata_sc)\n\u001b[1;32m    216\u001b[0m df_compare \u001b[38;5;241m=\u001b[39m tg\u001b[38;5;241m.\u001b[39mcompare_spatial_geneexp(adata_sp_new, adata_sp, adata_sc)\n",
      "File \u001b[0;32m~/miniconda3/envs/py-bio/lib/python3.10/site-packages/tangram/utils.py:368\u001b[0m, in \u001b[0;36mproject_genes\u001b[0;34m(adata_map, adata_sc, cluster_label, scale)\u001b[0m\n\u001b[1;32m    365\u001b[0m     adata_sc \u001b[38;5;241m=\u001b[39m mu\u001b[38;5;241m.\u001b[39madata_to_cluster_expression(adata_sc, cluster_label, scale\u001b[38;5;241m=\u001b[39mscale)\n\u001b[1;32m    367\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m adata_map\u001b[38;5;241m.\u001b[39mobs\u001b[38;5;241m.\u001b[39mindex\u001b[38;5;241m.\u001b[39mequals(adata_sc\u001b[38;5;241m.\u001b[39mobs\u001b[38;5;241m.\u001b[39mindex):\n\u001b[0;32m--> 368\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe two AnnDatas need to have same `obs` index.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(adata_sc\u001b[38;5;241m.\u001b[39mX, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtoarray\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    370\u001b[0m     adata_sc\u001b[38;5;241m.\u001b[39mX \u001b[38;5;241m=\u001b[39m adata_sc\u001b[38;5;241m.\u001b[39mX\u001b[38;5;241m.\u001b[39mtoarray()\n",
      "\u001b[0;31mValueError\u001b[0m: The two AnnDatas need to have same `obs` index."
     ]
    }
   ],
   "source": [
    "# sdata = selves[0].adata\n",
    "# # adata_sc = sc.read(os.path.join(\"/mnt/cho_lab/disk2/elizabeth/data\",\n",
    "# #                                 \"2023-05-12_CombinedCD-v2_new.h5ad\"))\n",
    "# # adata_sc = adata_sc[adata_sc.obs.Region == \"Ileum\"]\n",
    "# # sdata_new, sdata, adata_sc, ad_map, df_compare = cr.pp.integrate_spatial(\n",
    "# #     sdata, adata_sc, col_cell_type=\"ClusterAnnotation\")\n",
    "# adata_sc = sc.read(\"/mnt/cho_lab/disk2/elizabeth/data/elmentaite.h5ad\",\n",
    "#                    col_gene_symbols=\"feature_name\")\n",
    "# adata_sc = adata_sc[adata_sc.obs.tissue == \"small intestine\"]\n",
    "# adata_sc.var = adata_sc.var.reset_index().set_index(\"feature_name\")\n",
    "sdata_new, sdata, adata_sc, ad_map, df_compare = cr.pp.integrate_spatial(\n",
    "    sdata.table, adata_sc, col_cell_type=\"cell_type\", mode=\"clusters\",\n",
    "    key_added=\"rank_genes_groups_cell_type\", markers=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'SpatialData' object has no attribute 'var_names'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28mset\u001b[39m(np\u001b[38;5;241m.\u001b[39munique(pd\u001b[38;5;241m.\u001b[39mDataFrame(adata_sc\u001b[38;5;241m.\u001b[39muns[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrank_genes_groups_cell_type\u001b[39m\u001b[38;5;124m\"\u001b[39m][\n\u001b[0;32m----> 2\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m])\u001b[38;5;241m.\u001b[39mmelt()\u001b[38;5;241m.\u001b[39mvalue\u001b[38;5;241m.\u001b[39mvalues))\u001b[38;5;241m.\u001b[39mintersection(\u001b[38;5;28mset\u001b[39m(\u001b[43msdata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar_names\u001b[49m))\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'SpatialData' object has no attribute 'var_names'"
     ]
    }
   ],
   "source": [
    "set(np.unique(pd.DataFrame(adata_sc.uns[\"rank_genes_groups_cell_type\"][\n",
    "            \"names\"]).melt().value.values)).intersection(set(sdata.var_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key_cell_type, col_ann = None, \"leiden_sub\"\n",
    "for s in selves:\n",
    "    s.plot_spatial(color=col_ann)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s.annotate_clusters(file_ann, col_cell_type=\"leiden\", col_annotation=\"Annotation\")\n",
    "s.annotate_clusters(file_ann, col_cell_type=\"leiden_subcluster\", col_annotation=\"Annotation_Subcluster\")\n",
    "s.plot_spatial(color=\"Annotation_Subcluster\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dff = pd.concat([pd.Series(marker_genes_dict[x]).to_frame(\"Marker\")\n",
    "                 for x in marker_genes_dict], keys=marker_genes_dict)\n",
    "dff[dff.Marker.isin(self.rna.var_names)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adf = pd.read_excel(file_ann)\n",
    "adf\n",
    "\n",
    "\n",
    "i = 0\n",
    "self = selves[i]\n",
    "self.add_image(paths_he[i], name=\"H_E\", file_align=paths_he_align[i])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py-bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
