{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports & Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import re\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import corescpy as cr\n",
    "\n",
    "# Computing Resources\n",
    "gpu = False\n",
    "sc.settings.n_jobs = 4\n",
    "sc.settings.max_memory = 150\n",
    "\n",
    "# Display\n",
    "pd.options.display.max_colwidth = 1000\n",
    "pd.options.display.max_columns = 100\n",
    "pd.options.display.max_rows = 500\n",
    "sc.settings.set_figure_params(dpi=100, frameon=False, figsize=(20, 20))\n",
    "\n",
    "# Column Names (from Metadata & To Be Created)\n",
    "col_sample_id_o, col_sample_id = \"Sample ID\", \"Sample\"  # in metadata, new\n",
    "col_subject = \"Patient\"  # in metadata file\n",
    "col_inflamed, col_stricture = \"Inflamed\", \"Stricture\"  # in metadata file\n",
    "col_condition = \"Condition\"  # constructed from col_inflamed & col_stricture\n",
    "col_fff = \"file_path\"  # column in metadata in which to store data file path\n",
    "col_tangram = \"tangram_prediction\"  # for future Tangram imputation annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Options & Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  Sample ID  Patient     Status  Slide Id          Project  \\\n",
      "Sample                                                                       \n",
      "Uninflamed-50336C    50336C    50336  Available   11047.0  scRNA & Spatial   \n",
      "Inflamed-50336B      50336B    50336  Available   11044.0  scRNA & Spatial   \n",
      "Stricture-50336A     50336A    50336  Available   10496.0  scRNA & Spatial   \n",
      "\n",
      "                         Location Stricture  GRID ID    Inflamed  \\\n",
      "Sample                                                             \n",
      "Uninflamed-50336C  Terminal Ileum        no  1011328  uninflamed   \n",
      "Inflamed-50336B    Terminal Ileum        no  1011328    inflamed   \n",
      "Stricture-50336A   Terminal Ileum       yes  1011328    inflamed   \n",
      "\n",
      "                  Procedure Date  Age   Sex   Race Hispanic        Diagnosis  \\\n",
      "Sample                                                                         \n",
      "Uninflamed-50336C        2023-11   24  male  white       NO  Crohn's Disease   \n",
      "Inflamed-50336B          2023-11   24  male  white       NO  Crohn's Disease   \n",
      "Stricture-50336A         2023-11   24  male  white       NO  Crohn's Disease   \n",
      "\n",
      "                             Project.1  Procedure Disease_Status  \\\n",
      "Sample                                                             \n",
      "Uninflamed-50336C  Helmsley-Senescence  Resection         Active   \n",
      "Inflamed-50336B    Helmsley-Senescence  Resection         Active   \n",
      "Stricture-50336A   Helmsley-Senescence  Resection         Active   \n",
      "\n",
      "                  Date Collected Date Sectioned Date Hybridization Storage 4c  \\\n",
      "Sample                                                                          \n",
      "Uninflamed-50336C     2024.11.27     2024.03.21                NaN        YES   \n",
      "Inflamed-50336B       2024.11.27     2024.03.21                NaN        YES   \n",
      "Stricture-50336A      2024.11.27     2024.03.21         2024.04.15        YES   \n",
      "\n",
      "                       Created By      Created  Storage Status  Location.1  \\\n",
      "Sample                                                                       \n",
      "Uninflamed-50336C  lorena tavares  15-Apr-2024  Not in storage         NaN   \n",
      "Inflamed-50336B    lorena tavares  15-Apr-2024  Not in storage         NaN   \n",
      "Stricture-50336A   lorena tavares  15-Apr-2024  Not in storage         NaN   \n",
      "\n",
      "                   Storage Row  Storage Col  Checked Out By out_file  \\\n",
      "Sample                                                                 \n",
      "Uninflamed-50336C          NaN          NaN             NaN      NaN   \n",
      "Inflamed-50336B            NaN          NaN             NaN      NaN   \n",
      "Stricture-50336A           NaN          NaN             NaN      NaN   \n",
      "\n",
      "                    Condition  \\\n",
      "Sample                          \n",
      "Uninflamed-50336C  Uninflamed   \n",
      "Inflamed-50336B      Inflamed   \n",
      "Stricture-50336A    Stricture   \n",
      "\n",
      "                                                                                                                   file_path  \n",
      "Sample                                                                                                                        \n",
      "Uninflamed-50336C  /mnt/cho_lab/bbdata2/outputs/TUQ97N/CHO-010/output-XETG00189__0011047__50336C-TUQ97N-EA__20240422__175051  \n",
      "Inflamed-50336B    /mnt/cho_lab/bbdata2/outputs/TUQ97N/CHO-010/output-XETG00189__0011044__50336B-TUQ97N-EA__20240422__175051  \n",
      "Stricture-50336A   /mnt/cho_lab/bbdata2/outputs/TUQ97N/CHO-009/output-XETG00189__0010496__50336A-TUQ97N-EA__20240418__170624  \n",
      "\n",
      "\n",
      "<<< INITIALIZING SPATIAL CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "<<< INITIALIZING OMICS CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"manual_res1pt5_dist0_npc30\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Sample\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\u001b[34mINFO    \u001b[0m reading                                                                                                   \n",
      "         \u001b[35m/mnt/cho_lab/bbdata2/outputs/TUQ97N/CHO-010/output-XETG00189__0011047__50336C-TUQ97N-EA__20240422__175051/\u001b[0m\n",
      "         \u001b[95mcell_feature_matrix.h5\u001b[0m                                                                                    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "dsfasfda AnnData object with n_obs × n_vars = 387961 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Cell Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "Observations: 387961\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Gene Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "Empty DataFrame\n",
      "Columns: [index, gene_ids, feature_types, genome]\n",
      "Index: []\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " SpatialData object with:\n",
      "├── Images\n",
      "│     └── 'morphology_focus': MultiscaleSpatialImage[cyx] (5, 74826, 48440), (5, 37413, 24220), (5, 18706, 12110), (5, 9353, 6055), (5, 4676, 3027)\n",
      "├── Labels\n",
      "│     ├── 'cell_labels': MultiscaleSpatialImage[yx] (74826, 48440), (37413, 24220), (18706, 12110), (9353, 6055), (4676, 3027)\n",
      "│     └── 'nucleus_labels': MultiscaleSpatialImage[yx] (74826, 48440), (37413, 24220), (18706, 12110), (9353, 6055), (4676, 3027)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 11) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (387961, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (387961, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (379551, 1) (2D shapes)\n",
      "└── Tables\n",
      "      └── 'table': AnnData (387961, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), cell_labels (Labels), nucleus_labels (Labels), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "None\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 387961 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts' \n",
      "\n",
      "                      gene_ids    feature_types   genome\n",
      "gene_symbols                                           \n",
      "ABCA7         ENSG00000064687  Gene Expression  Unknown\n",
      "ACTA2         ENSG00000107796  Gene Expression  Unknown\n",
      "ACTG2         ENSG00000163017  Gene Expression  Unknown\n",
      "ADAM15        ENSG00000143537  Gene Expression  Unknown\n",
      "ADAM28        ENSG00000042980  Gene Expression  Unknown \n",
      "\n",
      "       cell_id  transcript_counts  control_probe_counts  \\\n",
      "0  aaaadjpg-1                  1                     0   \n",
      "1  aaaafdgn-1                  0                     0   \n",
      "2  aaaafgfm-1                139                     0   \n",
      "3  aaaagmcb-1                 92                     0   \n",
      "4  aaaaifon-1                150                     0   \n",
      "\n",
      "   control_codeword_counts  unassigned_codeword_counts  \\\n",
      "0                        0                           0   \n",
      "1                        0                           0   \n",
      "2                        0                           0   \n",
      "3                        0                           0   \n",
      "4                        0                           0   \n",
      "\n",
      "   deprecated_codeword_counts  total_counts   cell_area  nucleus_area  \\\n",
      "0                           0             1   32.422189           NaN   \n",
      "1                           0             0   33.596251           NaN   \n",
      "2                           0           139   79.429847     25.016563   \n",
      "3                           0            92   70.488909     24.790782   \n",
      "4                           0           150  101.104847     37.660314   \n",
      "\n",
      "         region  z_level  nucleus_count  cell_labels             Sample  \n",
      "0  cell_circles      0.0            0.0            1  Uninflamed-50336C  \n",
      "1  cell_circles      0.0            0.0            2  Uninflamed-50336C  \n",
      "2  cell_circles      0.0            1.0            3  Uninflamed-50336C  \n",
      "3  cell_circles      0.0            1.0            4  Uninflamed-50336C  \n",
      "4  cell_circles      0.0            1.0            5  Uninflamed-50336C  \n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"manual_res1pt5_dist0_npc30\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Sample\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "col_segment=\"region\"\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\n",
      "\n",
      "<<< RESTORING SINGLE-SAMPLE FROM h5ad >>>\n",
      "\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "AnnData object with n_obs × n_vars = 378219 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample', 'Sample ID', 'Patient', 'Status', 'Slide Id', 'Project', 'Location', 'Stricture', 'GRID ID', 'Inflamed', 'Procedure Date', 'Age', 'Sex', 'Race', 'Hispanic', 'Diagnosis', 'Project.1', 'Procedure', 'Disease_Status', 'Date Collected', 'Date Sectioned', 'Date Hybridization', 'Storage 4c', 'Created By', 'Created', 'Storage Status', 'Location.1', 'Storage Row', 'Storage Col', 'Checked Out By', 'out_file', 'Condition', 'file_path', 'n_counts', 'log_counts', 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'assay_protein', 'col_gene_symbols', 'col_cell_type', 'col_sample_id', 'col_batch', 'col_subject', 'col_condition', 'col_num_umis', 'col_segment', 'cell_filter_pmt', 'cell_filter_ncounts', 'cell_filter_ngene', 'gene_filter_ncell', 'gene_filter_ncounts', 'custom_thresholds', 'method_norm', 'figsize', 'kws_scale', 'leiden_res1pt5_dist0_npc30', 'label_res1pt5_dist0_npc30', 'leiden_res0pt5_dist0pt5_npc30', 'label_res0pt5_dist0pt5_npc30', 'leiden_res0pt75_dist0pt3_npc30', 'label_res0pt75_dist0pt3_npc30'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
      "    uns: 'Sample_colors', 'hvg', 'label_res0pt5_dist0pt5_npc30_colors', 'label_res0pt75_dist0pt3_npc30_colors', 'label_res1pt5_dist0_npc30_colors', 'leiden_res0pt5_dist0pt5_npc30', 'leiden_res0pt75_dist0pt3_npc30', 'leiden_res1pt5_dist0_npc30', 'log1p', 'neighbors', 'original_ix', 'pca', 'rank_genes_groups_leiden_res0pt5_dist0pt5_npc30', 'rank_genes_groups_leiden_res0pt75_dist0pt3_npc30', 'rank_genes_groups_leiden_res1pt5_dist0_npc30', 'spatialdata_attrs', 'specifications', 'umap', 'spatial'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    layers: 'counts', 'log1p', 'scaled'\n",
      "    obsp: 'connectivities', 'distances'\n",
      "\n",
      "\n",
      "<<< INITIALIZING SPATIAL CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "<<< INITIALIZING OMICS CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"manual_res1pt5_dist0_npc30\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Sample\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\u001b[34mINFO    \u001b[0m reading                                                                                                   \n",
      "         \u001b[35m/mnt/cho_lab/bbdata2/outputs/TUQ97N/CHO-010/output-XETG00189__0011044__50336B-TUQ97N-EA__20240422__175051/\u001b[0m\n",
      "         \u001b[95mcell_feature_matrix.h5\u001b[0m                                                                                    \n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "dsfasfda AnnData object with n_obs × n_vars = 563427 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Cell Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "Observations: 563427\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Gene Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "Empty DataFrame\n",
      "Columns: [index, gene_ids, feature_types, genome]\n",
      "Index: []\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " SpatialData object with:\n",
      "├── Images\n",
      "│     └── 'morphology_focus': MultiscaleSpatialImage[cyx] (5, 68049, 53911), (5, 34024, 26955), (5, 17012, 13477), (5, 8506, 6738), (5, 4253, 3369)\n",
      "├── Labels\n",
      "│     ├── 'cell_labels': MultiscaleSpatialImage[yx] (68049, 53911), (34024, 26955), (17012, 13477), (8506, 6738), (4253, 3369)\n",
      "│     └── 'nucleus_labels': MultiscaleSpatialImage[yx] (68049, 53911), (34024, 26955), (17012, 13477), (8506, 6738), (4253, 3369)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 11) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (563427, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (563427, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (555843, 1) (2D shapes)\n",
      "└── Tables\n",
      "      └── 'table': AnnData (563427, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), cell_labels (Labels), nucleus_labels (Labels), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "None\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 563427 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts' \n",
      "\n",
      "                      gene_ids    feature_types   genome\n",
      "gene_symbols                                           \n",
      "ABCA7         ENSG00000064687  Gene Expression  Unknown\n",
      "ACTA2         ENSG00000107796  Gene Expression  Unknown\n",
      "ACTG2         ENSG00000163017  Gene Expression  Unknown\n",
      "ADAM15        ENSG00000143537  Gene Expression  Unknown\n",
      "ADAM28        ENSG00000042980  Gene Expression  Unknown \n",
      "\n",
      "       cell_id  transcript_counts  control_probe_counts  \\\n",
      "0  aaaaacoc-1                120                     0   \n",
      "1  aaaabdap-1                 66                     0   \n",
      "2  aaaahpgb-1                 86                     0   \n",
      "3  aaaajjap-1                 97                     0   \n",
      "4  aaaakaed-1                134                     0   \n",
      "\n",
      "   control_codeword_counts  unassigned_codeword_counts  \\\n",
      "0                        0                           0   \n",
      "1                        0                           0   \n",
      "2                        0                           0   \n",
      "3                        0                           0   \n",
      "4                        0                           0   \n",
      "\n",
      "   deprecated_codeword_counts  total_counts  cell_area  nucleus_area  \\\n",
      "0                           0           120  72.701565     28.132345   \n",
      "1                           0            66  44.388595     26.235782   \n",
      "2                           0            86  33.686564           NaN   \n",
      "3                           0            97  61.502815     30.841720   \n",
      "4                           0           134  55.948596     24.068282   \n",
      "\n",
      "         region  z_level  nucleus_count  cell_labels           Sample  \n",
      "0  cell_circles      0.0            2.0            1  Inflamed-50336B  \n",
      "1  cell_circles      0.0            1.0            2  Inflamed-50336B  \n",
      "2  cell_circles      0.0            0.0            3  Inflamed-50336B  \n",
      "3  cell_circles      0.0            1.0            4  Inflamed-50336B  \n",
      "4  cell_circles      0.0            1.0            5  Inflamed-50336B  \n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"manual_res1pt5_dist0_npc30\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Sample\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "col_segment=\"region\"\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\n",
      "\n",
      "<<< RESTORING SINGLE-SAMPLE FROM h5ad >>>\n",
      "\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "AnnData object with n_obs × n_vars = 530116 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample', 'Sample ID', 'Patient', 'Status', 'Slide Id', 'Project', 'Location', 'Stricture', 'GRID ID', 'Inflamed', 'Procedure Date', 'Age', 'Sex', 'Race', 'Hispanic', 'Diagnosis', 'Project.1', 'Procedure', 'Disease_Status', 'Date Collected', 'Date Sectioned', 'Date Hybridization', 'Storage 4c', 'Created By', 'Created', 'Storage Status', 'Location.1', 'Storage Row', 'Storage Col', 'Checked Out By', 'out_file', 'Condition', 'file_path', 'n_counts', 'log_counts', 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'assay_protein', 'col_gene_symbols', 'col_cell_type', 'col_sample_id', 'col_batch', 'col_subject', 'col_condition', 'col_num_umis', 'col_segment', 'cell_filter_pmt', 'cell_filter_ncounts', 'cell_filter_ngene', 'gene_filter_ncell', 'gene_filter_ncounts', 'custom_thresholds', 'method_norm', 'figsize', 'kws_scale', 'leiden_res1pt5_dist0_npc30', 'label_res1pt5_dist0_npc30', 'leiden_res0pt5_dist0pt5_npc30', 'label_res0pt5_dist0pt5_npc30', 'leiden_res0pt75_dist0pt3_npc30', 'label_res0pt75_dist0pt3_npc30'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
      "    uns: 'Sample_colors', 'hvg', 'label_res0pt5_dist0pt5_npc30_colors', 'label_res0pt75_dist0pt3_npc30_colors', 'label_res1pt5_dist0_npc30_colors', 'leiden_res0pt5_dist0pt5_npc30', 'leiden_res0pt75_dist0pt3_npc30', 'leiden_res1pt5_dist0_npc30', 'log1p', 'neighbors', 'original_ix', 'pca', 'rank_genes_groups_leiden_res0pt5_dist0pt5_npc30', 'rank_genes_groups_leiden_res0pt75_dist0pt3_npc30', 'rank_genes_groups_leiden_res1pt5_dist0_npc30', 'spatialdata_attrs', 'specifications', 'umap', 'spatial'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    layers: 'counts', 'log1p', 'scaled'\n",
      "    obsp: 'connectivities', 'distances'\n",
      "\n",
      "\n",
      "<<< INITIALIZING SPATIAL CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "<<< INITIALIZING OMICS CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"manual_res1pt5_dist0_npc30\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Sample\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\u001b[34mINFO    \u001b[0m reading                                                                                                   \n",
      "         \u001b[35m/mnt/cho_lab/bbdata2/outputs/TUQ97N/CHO-009/output-XETG00189__0010496__50336A-TUQ97N-EA__20240418__170624/\u001b[0m\n",
      "         \u001b[95mcell_feature_matrix.h5\u001b[0m                                                                                    \n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "dsfasfda AnnData object with n_obs × n_vars = 465118 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Cell Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "Observations: 465118\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Gene Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "Empty DataFrame\n",
      "Columns: [index, gene_ids, feature_types, genome]\n",
      "Index: []\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " SpatialData object with:\n",
      "├── Images\n",
      "│     └── 'morphology_focus': MultiscaleSpatialImage[cyx] (5, 74906, 51104), (5, 37453, 25552), (5, 18726, 12776), (5, 9363, 6388), (5, 4681, 3194)\n",
      "├── Labels\n",
      "│     ├── 'cell_labels': MultiscaleSpatialImage[yx] (74906, 51104), (37453, 25552), (18726, 12776), (9363, 6388), (4681, 3194)\n",
      "│     └── 'nucleus_labels': MultiscaleSpatialImage[yx] (74906, 51104), (37453, 25552), (18726, 12776), (9363, 6388), (4681, 3194)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 11) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (465118, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (465118, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (459750, 1) (2D shapes)\n",
      "└── Tables\n",
      "      └── 'table': AnnData (465118, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), cell_labels (Labels), nucleus_labels (Labels), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "None\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 465118 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts' \n",
      "\n",
      "                      gene_ids    feature_types   genome\n",
      "gene_symbols                                           \n",
      "ABCA7         ENSG00000064687  Gene Expression  Unknown\n",
      "ACTA2         ENSG00000107796  Gene Expression  Unknown\n",
      "ACTG2         ENSG00000163017  Gene Expression  Unknown\n",
      "ADAM15        ENSG00000143537  Gene Expression  Unknown\n",
      "ADAM28        ENSG00000042980  Gene Expression  Unknown \n",
      "\n",
      "       cell_id  transcript_counts  control_probe_counts  \\\n",
      "0  aaaaapek-1                  0                     0   \n",
      "1  aaaabcdk-1                  0                     0   \n",
      "2  aaaadmae-1                  0                     0   \n",
      "3  aaaajcdb-1                 96                     0   \n",
      "4  aaaakmmm-1                 49                     0   \n",
      "\n",
      "   control_codeword_counts  unassigned_codeword_counts  \\\n",
      "0                        0                           0   \n",
      "1                        0                           0   \n",
      "2                        0                           0   \n",
      "3                        0                           0   \n",
      "4                        0                           0   \n",
      "\n",
      "   deprecated_codeword_counts  total_counts   cell_area  nucleus_area  \\\n",
      "0                           0             0   73.424065           NaN   \n",
      "1                           0             0   38.834376           NaN   \n",
      "2                           0             0  119.754379           NaN   \n",
      "3                           0            96   61.006096     24.835938   \n",
      "4                           0            49   42.537189     26.280938   \n",
      "\n",
      "         region  z_level  nucleus_count  cell_labels            Sample  \n",
      "0  cell_circles      0.0            0.0            1  Stricture-50336A  \n",
      "1  cell_circles      0.0            0.0            2  Stricture-50336A  \n",
      "2  cell_circles      0.0            0.0            3  Stricture-50336A  \n",
      "3  cell_circles      0.0            1.0            4  Stricture-50336A  \n",
      "4  cell_circles      0.0            1.0            5  Stricture-50336A  \n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"manual_res1pt5_dist0_npc30\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Sample\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "col_segment=\"region\"\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\n",
      "\n",
      "<<< RESTORING SINGLE-SAMPLE FROM h5ad >>>\n",
      "\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "AnnData object with n_obs × n_vars = 444414 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample', 'Sample ID', 'Patient', 'Status', 'Slide Id', 'Project', 'Location', 'Stricture', 'GRID ID', 'Inflamed', 'Procedure Date', 'Age', 'Sex', 'Race', 'Hispanic', 'Diagnosis', 'Project.1', 'Procedure', 'Disease_Status', 'Date Collected', 'Date Sectioned', 'Date Hybridization', 'Storage 4c', 'Created By', 'Created', 'Storage Status', 'Location.1', 'Storage Row', 'Storage Col', 'Checked Out By', 'out_file', 'Condition', 'file_path', 'n_counts', 'log_counts', 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'assay_protein', 'col_gene_symbols', 'col_cell_type', 'col_sample_id', 'col_batch', 'col_subject', 'col_condition', 'col_num_umis', 'col_segment', 'cell_filter_pmt', 'cell_filter_ncounts', 'cell_filter_ngene', 'gene_filter_ncell', 'gene_filter_ncounts', 'custom_thresholds', 'method_norm', 'figsize', 'kws_scale', 'leiden_res1pt5_dist0_npc30', 'label_res1pt5_dist0_npc30', 'leiden_res0pt5_dist0pt5_npc30', 'label_res0pt5_dist0pt5_npc30', 'leiden_res0pt75_dist0pt3_npc30', 'label_res0pt75_dist0pt3_npc30'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
      "    uns: 'Sample_colors', 'hvg', 'label_res0pt5_dist0pt5_npc30_colors', 'label_res0pt75_dist0pt3_npc30_colors', 'label_res1pt5_dist0_npc30_colors', 'leiden_res0pt5_dist0pt5_npc30', 'leiden_res0pt75_dist0pt3_npc30', 'leiden_res1pt5_dist0_npc30', 'log1p', 'neighbors', 'original_ix', 'pca', 'rank_genes_groups_leiden_res0pt5_dist0pt5_npc30', 'rank_genes_groups_leiden_res0pt75_dist0pt3_npc30', 'rank_genes_groups_leiden_res1pt5_dist0_npc30', 'spatialdata_attrs', 'specifications', 'umap', 'spatial'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    layers: 'counts', 'log1p', 'scaled'\n",
      "    obsp: 'connectivities', 'distances'\n"
     ]
    }
   ],
   "source": [
    "# Directories & Metadata\n",
    "load, reannotate = True, True\n",
    "run = [\"CHO-010\", \"CHO-010\", \"CHO-009\"]\n",
    "# run = [\"CHO-008\", \"CHO-008\", \"CHO-009\"]\n",
    "# run = [\"CHO-001\", \"CHO-002\"]\n",
    "# run = [\"CHO-001\", \"CHO-002\", \"CHO-005\", \"CHO-006\", \"CHO-007\"]\n",
    "# samples = \"all\"\n",
    "# samples = [\"50564A4\", \"50618B5\"]\n",
    "# samples = [\"50007B2\"]\n",
    "# samples = [\"50006B\", \"50006A\", \"50006C\"]\n",
    "# samples = [\"50217B\", \"50217A\", \"50217C\"]\n",
    "samples = [\"50336C\", \"50336B\", \"50336A\"]\n",
    "# samples = [\"50452A\", \"50452B\", \"50452C\"]\n",
    "# samples = [\"50564A4\", \"50618B5\"]\n",
    "\n",
    "# Optionally, Define Manual Annotation Versions\n",
    "# should be stored in (\"<out_dir>/annotations_dictionaries\")\n",
    "# in format <selves[i]._library_id>___leiden_<man_anns[i]>_dictionary.xlsx\n",
    "# with first column = leiden cluster and second column = annotation\n",
    "man_anns = True  # load manual annotations according to clustering kws\n",
    "# man_anns = [\"res0pt5_dist0pt5_npc30\", \"res0pt75_dist0pt3_npc30\",\n",
    "#             \"res1pt5_dist0_npc30\"]  # choose manual annotations to load\n",
    "# man_anns = None  # do not load manual annotations\n",
    "\n",
    "# Main Directories\n",
    "# Replace manually or mirror my file/directory tree in your home (`ddu`)\n",
    "ddu = os.path.expanduser(\"~\")\n",
    "ddm = \"/mnt/cho_lab\" if os.path.exists(\"/mnt/cho_lab\") else \"/mnt\"  # Spark?\n",
    "ddl = f\"{ddm}/disk2/{os.getlogin()}/data/shared-xenium-library\" if (\n",
    "    \"cho\" in ddm) else os.path.join(ddu, \"shared-xenium-library\")\n",
    "ddx = f\"{ddm}/bbdata2\"  # mounted drive Xenium folder\n",
    "out_dir = os.path.join(ddl, \"outputs\", \"TUQ97N\", \"nebraska\")  # None = no save\n",
    "d_path = os.path.join(ddm, \"disk2\" if \"cho\" in ddm else \"\",\n",
    "                      os.getlogin(), \"data\")  # other, e.g., Tangram data\n",
    "anf = pd.read_csv(os.path.join(ddu, \"corescpy/examples/markers_lineages.csv\"))\n",
    "file_mdf = os.path.join(ddl, \"samples.csv\")  # metadata\n",
    "\n",
    "# Annotation & Tangram Imputation\n",
    "col_assignment = \"Bin\"  # which column from annotation file to use\n",
    "# col_cell_type_sc, file_sc = \"ClusterAnnotation\", str(\n",
    "#     f\"{d_path}/2023-05-12_CombinedCD-v2_ileal_new.h5ad\")\n",
    "col_cell_type_sc, file_sc = \"cell_type\", f\"{d_path}/elmentaite_ileal.h5ad\"\n",
    "# file_sc = None  # to skip Tangram imputation/label transfer\n",
    "\n",
    "# Processing & Clustering Options\n",
    "kws_pp = dict(cell_filter_pmt=None, cell_filter_ncounts=[15, None],\n",
    "              cell_filter_ngene=[3, None], gene_filter_ncell=[3, None],\n",
    "              gene_filter_ncounts=[3, None], custom_thresholds=None,\n",
    "              kws_scale=dict(max_value=10, zero_center=True),\n",
    "              method_norm=\"log\")  # preprocessing keyword arguments\n",
    "kws_cluster = dict(kws_umap=dict(method=\"rapids\" if gpu else \"umap\"),\n",
    "                   genes_subset=list(anf.iloc[:, 0]),  # use only markers\n",
    "                   use_gpu=gpu, use_highly_variable=False)\n",
    "kws_clustering, col_assignment = {}, []\n",
    "for i in zip([0.5, 0.75, 1.5], [0.5, 0.3, 0], [30, 30, 30]):\n",
    "    kws = {**kws_cluster, \"resolution\": i[0], \"kws_umap\": {\n",
    "        **kws_cluster[\"kws_umap\"], \"min_dist\": i[1]}, \"n_comps\": i[2]}\n",
    "    suff = str(f\"res{re.sub('[.]', 'pt', str(kws['resolution']))}_dist\"\n",
    "               f\"{re.sub('[.]', 'pt', str(kws['kws_umap']['min_dist']))}\"\n",
    "               f\"_npc{kws['n_comps']}\")  # file path suffix\n",
    "    kws_clustering.update({suff: kws})\n",
    "    col_assignment += [\"group\" if kws[\"resolution\"] >= 0.7 else \"Bucket\"]\n",
    "if man_anns is True:\n",
    "    man_anns = list(kws_clustering.keys())\n",
    "col_cell_type = list(kws_clustering.keys())[-1] if (\n",
    "    man_anns is None) else f\"manual_{man_anns[-1]}\"  # default cell labels\n",
    "\n",
    "# After this point, no more options to specify\n",
    "# Just code to infer the data file path from your specifications\n",
    "# and construct argument dictionaries and manipulate metadata and such.\n",
    "\n",
    "# Read Metadata & Other Information\n",
    "metadata = (pd.read_excel if file_mdf[-4:] == \"xlsx\" else pd.read_csv)(\n",
    "    file_mdf, dtype={\"Slide ID\": str}).rename({\n",
    "        \"Name\": col_subject, \"Inflammation\": col_inflamed}, axis=1)\n",
    "if samples not in [\"all\", None]:  # subset by sample ID?\n",
    "    metadata = metadata.set_index(col_sample_id_o).loc[samples].reset_index()\n",
    "\n",
    "# Revise Metadata & Construct Variables from Options\n",
    "metadata.loc[:, col_condition] = metadata.apply(lambda x: \"Stricture\" if x[\n",
    "    col_stricture].lower() in [\"stricture\", \"yes\"] else x[\n",
    "        col_inflamed].capitalize(), axis=1)  # inflamation/stricture condition\n",
    "metadata.loc[:, col_sample_id] = metadata[[col_condition, col_sample_id_o]\n",
    "                                          ].apply(\"-\".join, axis=1)\n",
    "metadata = metadata.set_index(col_sample_id)\n",
    "fff = np.array(cr.pp.construct_file(run=run, directory=ddx))\n",
    "bff = np.array([os.path.basename(i) for i in fff])  # base path names\n",
    "samps = np.array([i.split(\"__\")[2].split(\"-\")[0] for i in fff])\n",
    "for x in metadata[col_sample_id_o]:\n",
    "    m_f = metadata[metadata[col_sample_id_o] == x][\n",
    "        \"out_file\"].iloc[0]  # ...use to find unconventionally-named files\n",
    "    locx = np.where(samps == x)[0] if pd.isnull(\n",
    "        m_f) else np.where(bff == m_f)[0]\n",
    "    metadata.loc[metadata[col_sample_id_o] == x, col_fff] = fff[locx[0]] if (\n",
    "        len(locx) > 0) else np.nan  # assign output file to metadata row\n",
    "metadata = metadata.dropna(subset=[col_fff]).drop_duplicates()\n",
    "\n",
    "# Annotation File\n",
    "assign = anf.dropna(subset=col_assignment).set_index(\n",
    "    \"gene\").rename_axis(\"Gene\")  # markers\n",
    "# assign = assign[~assign.Quality.isin([-1])]  # drop low-quality markers\n",
    "\n",
    "# Print Metadata & Make Output Directory (If Not Present)\n",
    "print(metadata)\n",
    "if not os.path.exists(out_dir):\n",
    "    os.makedirs(out_dir)\n",
    "\n",
    "# Load Data\n",
    "kws_init = dict(col_sample_id=col_sample_id, col_subject=col_subject,\n",
    "                col_cell_type=col_cell_type)  # object creation arguments\n",
    "selves = [None] * metadata.shape[0]  # to hold different samples\n",
    "for i, x in enumerate(metadata.index.values):\n",
    "    selves[i] = cr.Spatial(metadata.loc[x][col_fff], library_id=x, **kws_init)\n",
    "    for j in metadata:  # iterate metadata columns\n",
    "        selves[i].rna.obs.loc[:, j] = str(metadata.loc[x][j])  # add to object\n",
    "    selves[i].rna.obs.loc[:, \"out_file\"] = os.path.join(\n",
    "        out_dir, selves[i]._library_id)  # output path (to save object)\n",
    "    if load is True:\n",
    "        if os.path.exists(str(selves[i].rna.obs.out_file.iloc[0]) + \".h5ad\"):\n",
    "            selves[i].update_from_h5ad(selves[i].rna.obs.out_file.iloc[0])\n",
    "        print(selves[i].rna)\n",
    "\n",
    "# Marker Gene Dictionary (for Scanpy Plotting)\n",
    "marker_genes_dict = dict(assign[\"Bucket\"].reset_index().groupby(\n",
    "    \"Bucket\").apply(lambda x: list(pd.unique(list(set(\n",
    "        x.Gene).intersection(selves[0].rna.var_names))))))  # to dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manual Annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Manual Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if man_anns is not None and man_anns is not False:\n",
    "    for i, s in enumerate(selves):\n",
    "        for r in man_anns:  # iterate Leiden clusterings\n",
    "            fmr = os.path.join(out_dir, \"annotation_dictionaries\", str(\n",
    "                f\"{s._library_id}___leiden_{r}_dictionary.xlsx\"))  # mappings\n",
    "            if os.path.exists(fmr) is False:\n",
    "                print(f\"{fmr} file NOT found.\")\n",
    "                continue\n",
    "            else:\n",
    "                print(f\"{fmr} file found.\")\n",
    "            fmr = pd.read_excel(fmr).iloc[:, :2].astype(str)\n",
    "            mans = dict(fmr.set_index(fmr.columns[0]).iloc[:, 0])\n",
    "            s.rna.obs.loc[:, f\"manual_{r}\"] = s.rna.obs[f\"leiden_{r}\"].astype(\n",
    "                int).astype(str).replace(mans)  # Leiden -> manual annotation\n",
    "            s.rna.obs.loc[s.rna.obs[f\"manual_{r}\"].isnull(\n",
    "                ), f\"manual_{r}\"] = s.rna.obs.loc[s.rna.obs[\n",
    "                    f\"manual_{r}\"].isnull(), f\"leiden_{r}\"].astype(\n",
    "                        str)  # missing annotations replaced with Leiden\n",
    "            s.rna.obs.loc[:, f\"manual_{r}\"] = s.rna.obs[\n",
    "                f\"manual_{r}\"].astype(\"category\")  # as categorical\n",
    "            # s.plot_spatial(f\"manual_{r}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write Manual Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if man_anns not in [None, False] and out_dir is not None:\n",
    "    for i, s in enumerate(selves):\n",
    "        for r in man_anns:\n",
    "            if f\"manual_{r}\" in s.rna.obs:\n",
    "                s.write_clusters(out_dir, col_cell_type=f\"manual_{r}\",\n",
    "                                 overwrite=True,\n",
    "                                 file_prefix=f\"{s._library_id}__\")\n",
    "        s.write(s.rna.obs.out_file.iloc[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing, Leiden, Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "print(\"\\n\\n\", kws_pp, \"\\n\\n\")\n",
    "for i, s in enumerate(selves):\n",
    "    f_o = str(selves[i].rna.obs.out_file.iloc[0])\n",
    "\n",
    "    # Preprocessing\n",
    "    if load is False:\n",
    "        _ = s.preprocess(**kws_pp, figsize=(15, 15))  # preprocess\n",
    "\n",
    "    # Clustering at Different Resolutions & Minimum Distances\n",
    "    for j, x in enumerate(kws_clustering):\n",
    "\n",
    "        # Variables & Output Files\n",
    "        print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\\n\\n\")\n",
    "        cct, cca = f\"leiden_{x}\", f\"label_{x}\"  # Leiden & annotation columns\n",
    "        annot = assign[[col_assignment[j]]]  # gene-annotation dictionary\n",
    "\n",
    "        # Clustering, Markers, Annotation, & Writing Output\n",
    "        if load is False:\n",
    "            _ = s.cluster(**kws_clustering[x], key_added=cct, out_file=f_o)\n",
    "        if load is False or f\"rank_genes_groups_{cct}\" not in s.rna.uns:\n",
    "            _ = s.find_markers(col_cell_type=cct, kws_plot=False)  # DEGs\n",
    "        if reannotate is True or load is False:  # annotate; Explorer files\n",
    "            _ = s.annotate_clusters(annot, col_cell_type=cct,\n",
    "                                    col_annotation=cca)  # annotate\n",
    "            for c in [k for k in [cct, cca] if k in s.rna.obs]:  # Explorer\n",
    "                s.write_clusters(out_dir, col_cell_type=c, overwrite=True,\n",
    "                                 file_prefix=f\"{s._library_id}__\", n_top=True)\n",
    "\n",
    "        # Write Final Object\n",
    "        if load is False or reannotate is True and f_o is not None:\n",
    "            s.write(f_o)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "\n",
    "# print(\"\\n\\n\", kws_pp, \"\\n\\n\")\n",
    "# for i, s in enumerate(selves):\n",
    "#     f_o = str(selves[i].rna.obs.out_file.iloc[0])\n",
    "\n",
    "#     # Preprocessing\n",
    "#     if load is False:\n",
    "#         _ = s.preprocess(**kws_pp, figsize=(15, 15))  # preprocess\n",
    "\n",
    "#     # Clustering at Different Resolutions & Minimum Distances\n",
    "#     for j, x in enumerate(kws_clustering):\n",
    "\n",
    "#         # Variables & Output Files\n",
    "#         print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\\n\\n\")\n",
    "#         cct, cca = f\"leiden_{x}\", f\"label_{x}\"  # Leiden & annotation columns\n",
    "#         annot = assign[[col_assignment[j]]]  # gene-annotation dictionary\n",
    "\n",
    "#         # Clustering, Markers, Annotation, & Writing Output\n",
    "#         if load is False:\n",
    "#             _ = s.cluster(**kws_clustering[x], key_added=cct, out_file=f_o)\n",
    "#         if load is False or f\"rank_genes_groups_{cct}\" not in s.rna.uns:\n",
    "#             _ = s.find_markers(col_cell_type=cct, kws_plot=False)  # DEGs\n",
    "#         if reannotate is True or load is False:  # annotate; Explorer files\n",
    "#             _ = s.annotate_clusters(annot, col_cell_type=cct,\n",
    "#                                     col_annotation=cca)  # annotate\n",
    "#             for c in [k for k in [cct, cca] if k in s.rna.obs]:  # Explorer\n",
    "#                 s.write_clusters(out_dir, col_cell_type=c, overwrite=True,\n",
    "#                                  file_prefix=f\"{s._library_id}__\", n_top=True)\n",
    "\n",
    "#         # Write Final Object\n",
    "#         if load is False or reannotate is True and f_o is not None:\n",
    "#             s.write(f_o)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tangram Imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "if file_sc is not None:\n",
    "    adata_sc = sc.read(file_sc)  # read whole tx'ome data for imputation\n",
    "    if load is False:\n",
    "        for i, s in enumerate(selves):\n",
    "            out = s.impute(\n",
    "                adata_sc.copy(), col_cell_type=col_cell_type_sc,\n",
    "                mode=\"clusters\", markers=None, plot=False, plot_density=False,\n",
    "                plot_genes=None, col_annotation=col_tangram, out_file=None)\n",
    "            out[0].write_h5ad(os.path.splitext(selves[\n",
    "                0].rna.obs.out_file.iloc[0])[0] + \"___tangram.h5ad\")  # write\n",
    "            s.write(s.rna.obs.out_file.iloc[0])\n",
    "            s.write_clusters(out_dir, file_prefix=f\"{s._library_id}___\",\n",
    "                             col_cell_type=col_tangram,\n",
    "                             overwrite=True, n_top=True)\n",
    "    s.plot_spatial(color=col_tangram)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in selves:\n",
    "    s.plot_spatial(color=col_tangram)\n",
    "    for j, x in enumerate(kws_clustering):\n",
    "        _ = s.plot_spatial(color=[f\"leiden_{x}\",, f\"label_{x}\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ToppGene & Marker Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "self = selves[2]\n",
    "self.get_layer(\"counts\", inplace=True)\n",
    "\n",
    "srcs = [\"Cells of the human intestinal tract mapped across space and time\",\n",
    "        \"Human Ileal Epithelial cells from Crohn’s Disease\",\n",
    "        \"Human Ileal Immune cells from Crohn’s Disease\"]\n",
    "# cct = \"leiden_res0pt5_dist0pt5_npc30\"\n",
    "# cct = \"leiden_res0pt75_dist0pt3_npc30\"\n",
    "c_t = \"leiden_res1pt5_dist0_npc30\"\n",
    "tgdf, mks = self.annotate_clusters(\n",
    "        None, sources=srcs, col_cell_type=c_t, max_results=10000,\n",
    "        name_pattern={srcs[0]: \"SmallIntestine\"}, p_threshold=1e-15,\n",
    "        lfc_threshold=1.5, n_top_genes=20, n_top_annotations=40)\n",
    "tgdf.loc[:, \"Name\"] = tgdf.Name.str.strip(\n",
    "        \" / Per Region, Age_group, Lineage, cell class, cell type   \").apply(\n",
    "                lambda x: re.sub(\"SmallIntestine\", \"SmInt\", \" --- \".join(\n",
    "                        x.split(\"|\")[0].split(\"-\"))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### By Top Markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                Name  \\\n",
      "0     SmInt_GW_trimst --- 2 --- Neuronal --- Glial_immature --- ENCC/glia_Progenitor   \n",
      "1                                                 SmInt_GW_trimst --- 2 --- Neuronal   \n",
      "2           SmInt_GW_trimst --- 2 --- Neuronal --- Glial_immature --- Glia_3_(BCAN+)   \n",
      "3                              SmInt_GW_trimst --- 2 --- Neuronal --- Glial_immature   \n",
      "4                                               SmInt_GW_trimst --- 1.5 --- Neuronal   \n",
      "5   SmInt_GW_trimst --- 1.5 --- Neuronal --- neuron_precursor --- cycling_neuroblast   \n",
      "6         SmInt_GW_trimst --- 1.5 --- Neuronal --- Glial_immature --- Glia_3_(BCAN+)   \n",
      "7           SmInt_GW_trimst --- 1.5 --- Neuronal --- neuron_precursor --- Neuroblast   \n",
      "8                  SmInt_Child04 --- 06 --- Neuronal --- Glial_mature --- Adult_Glia   \n",
      "9                                                  SmInt_Child04 --- 06 --- Neuronal   \n",
      "10                                SmInt_Child04 --- 06 --- Neuronal --- Glial_mature   \n",
      "11                         SmInt_GW_trimst --- 1.5 --- Neuronal --- neuron_precursor   \n",
      "12                           SmInt_GW_trimst --- 1.5 --- Neuronal --- Glial_immature   \n",
      "13           SmInt_GW_trimst --- 2 --- Neuronal --- Glial_immature --- Glia_2_(ELN+)   \n",
      "14    SmInt_GW_trimst --- 1 --- Neuronal --- Glial_immature --- ENCC/glia_Progenitor   \n",
      "15                           SmInt_GW_trimst --- 1 --- Neuronal --- neuron_precursor   \n",
      "16                                                SmInt_GW_trimst --- 1 --- Neuronal   \n",
      "17  SmInt_GW_trimst --- 1.5 --- Neuronal --- Glial_immature --- Differentiating_glia   \n",
      "18         SmInt_GW_trimst --- 1.5 --- Neuronal --- Glial_immature --- Glia_2_(ELN+)   \n",
      "19                          SmInt_Adult --- Neuronal --- Glial_mature --- Adult_Glia   \n",
      "20                                         SmInt_Adult --- Neuronal --- Glial_mature   \n",
      "21                                                          SmInt_Adult --- Neuronal   \n",
      "\n",
      "                                         Genes  GenesInTerm  GenesInQuery  \\\n",
      "0    [SYT1, S100B, LINC00682, SORBS2, TPD52L1]          182            20   \n",
      "1      [SYT11, SYT1, S100B, LINC00682, SORBS2]          195            20   \n",
      "2         [SYT1, S100B, SORBS2, TPD52L1, GPX3]          195            20   \n",
      "3   [SYT11, S100B, LINC00682, SORBS2, TPD52L1]          196            20   \n",
      "4     [SYT11, SYT1, S100B, LINC00682, TPD52L1]          198            20   \n",
      "5    [TUBA1B, SYT1, S100B, LINC00682, TPD52L1]          199            20   \n",
      "6          [S100B, LINC00682, SORBS2, TPD52L1]          176            20   \n",
      "7            [SYT1, S100B, LINC00682, TPD52L1]          190            20   \n",
      "8             [S100B, LINC00682, SORBS2, GPX3]          191            20   \n",
      "9             [S100B, LINC00682, SORBS2, GPX3]          191            20   \n",
      "10            [S100B, LINC00682, SORBS2, GPX3]          191            20   \n",
      "11           [SYT1, S100B, LINC00682, TPD52L1]          194            20   \n",
      "12         [S100B, LINC00682, SORBS2, TPD52L1]          195            20   \n",
      "13           [SYT11, S100B, LINC00682, SORBS2]          196            20   \n",
      "14          [SYT11, S100B, LINC00682, TPD52L1]          196            20   \n",
      "15             [SYT11, SYT1, S100B, LINC00682]          197            20   \n",
      "16             [SYT11, SYT1, S100B, LINC00682]          198            20   \n",
      "17                  [S100B, LINC00682, SORBS2]          146            20   \n",
      "18                  [S100B, LINC00682, SORBS2]          162            20   \n",
      "19                      [SYT11, S100B, SORBS2]          172            20   \n",
      "20                      [SYT11, S100B, SORBS2]          172            20   \n",
      "21                      [SYT11, S100B, SORBS2]          172            20   \n",
      "\n",
      "    GenesInTermInQuery  TotalGenes        PValue  \n",
      "0                    5       45650  1.407850e-08  \n",
      "1                    5       45650  1.988106e-08  \n",
      "2                    5       45650  1.988106e-08  \n",
      "3                    5       45650  2.039591e-08  \n",
      "4                    5       45650  2.145742e-08  \n",
      "5                    5       45650  2.200441e-08  \n",
      "6                    4       45650  9.857625e-07  \n",
      "7                    4       45650  1.337008e-06  \n",
      "8                    4       45650  1.365223e-06  \n",
      "9                    4       45650  1.365223e-06  \n",
      "10                   4       45650  1.365223e-06  \n",
      "11                   4       45650  1.452529e-06  \n",
      "12                   4       45650  1.482532e-06  \n",
      "13                   4       45650  1.512994e-06  \n",
      "14                   4       45650  1.512994e-06  \n",
      "15                   4       45650  1.543917e-06  \n",
      "16                   4       45650  1.575308e-06  \n",
      "17                   3       45650  3.510320e-05  \n",
      "18                   3       45650  4.783915e-05  \n",
      "19                   3       45650  5.715879e-05  \n",
      "20                   3       45650  5.715879e-05  \n",
      "21                   3       45650  5.715879e-05  \n",
      "\n",
      "================================================================================\n",
      "Count Threshold: 1\n",
      "================================================================================\n",
      "\n",
      "\n",
      " group\n",
      "Adult Glia                                                 SORBS2, GPX3, IGFBP7\n",
      "B (Plasma RBC)                                                            COX8A\n",
      "Branch A4 (IN)                                                             BCL2\n",
      "Branch B1 (eMN)                                                LINC00682, RAMP1\n",
      "Branch B3 (IPAN)                                                          SKAP2\n",
      "CLDN10+ cells                                                            IGFBP7\n",
      "Contractile pericyte (PLN+)                                              SORBS2\n",
      "Cycling B cell                                                    TUBA1B, COX8A\n",
      "Cycling Stromal                                                          TUBA1B\n",
      "EC cells (NPW+)                                                           PEG10\n",
      "Endothelial                                                                AQP1\n",
      "Glia 1 (DHH+)                                                             S100B\n",
      "Glia 2 (ELN+)                                                              AQP1\n",
      "Glia 3 (BCAN+)                                                    TPD52L1, APOE\n",
      "ICC                                                                       RAMP1\n",
      "ILC1                                                                   HSP90AA1\n",
      "ILCP                                                      CAMTA1, PEG10, IGFBP7\n",
      "Immature B                                                                JAZF1\n",
      "LTi-like NCR- ILC3                                                       NCKAP1\n",
      "LZ GC cell                                                             MARCKSL1\n",
      "Lymphoid DC                                                            MARCKSL1\n",
      "MMP9+ Inflammatory macrophage                                              APOE\n",
      "MPO+ mono-neutrophil                                                      COX8A\n",
      "Macrophages                                                       HSPA1A, KCNA2\n",
      "Mature venous EC                                                        TPD52L1\n",
      "Mesenchymal                                                              SORBS2\n",
      "Neural                           S100B, LINC00682, TPD52L1, IGFBP7, APOE, RAMP1\n",
      "Pericyte                                                                 IGFBP7\n",
      "Proximal progenitor                                                        APOE\n",
      "SELL+ CD8 T                                                               S100B\n",
      "SMC (PLPP2+)                                                              RIMS1\n",
      "Stromal 1 (CCL11+)                                                         APOE\n",
      "Stromal 3 (C7+)                                                           PEG10\n",
      "TRGV4 gdT                                                          TUBA1B, CREM\n",
      "TRGV5/7 gdT                                                       HSPA1A, STIP1\n",
      "arterial capillary                                                       SORBS2\n",
      "cDC1                                                                S100B, SNX3\n",
      "cycling EC                                                               TUBA1B\n",
      "cycling neuroblast                                                       TUBA1B\n",
      "cycling stromal                                                          TUBA1B\n",
      "mLN Stroma (FMO2+)                                                   APOD, GPX3\n",
      "myofibroblast                                                            HSPA1A\n",
      "myofibroblast (RSPO2+)                                     SYT1, HSPA1A, HSPA1B\n",
      "dtype: object \n",
      "\n",
      "              31   Other   Total  Percent_Total\n",
      "Gene                                          \n",
      "LINC00682   518    1423    1941             26\n",
      "SYT1       1209    3546    4755             25\n",
      "RIMS1       303    3758    4061              7\n",
      "PEG10       399   10408   10807              3\n",
      "TPD52L1     488   13387   13875              3\n",
      "KCNA2       294   10125   10419              2\n",
      "S100B       970   41524   42494              2\n",
      "APOD        695   51008   51703              1\n",
      "HSPA1B      429   32699   33128              1\n",
      "JAZF1       310   24527   24837              1\n",
      "SORBS2      625   49933   50558              1\n",
      "CAMTA1      553   46306   46859              1\n",
      "CREM        363   36072   36435              0\n",
      "BCL2        646   65084   65730              0\n",
      "RAMP1       394   44726   45120              0\n",
      "APOE        411   48293   48704              0\n",
      "HSPA1A      580   68300   68880              0\n",
      "STIP1       463   54836   55299              0\n",
      "SKAP2       410   49517   49927              0\n",
      "MARCKSL1    493   63549   64042              0\n",
      "AQP1        652   85200   85852              0\n",
      "COX8A       499   70919   71418              0\n",
      "GPX3        769  109503  110272              0\n",
      "NCKAP1      749  122361  123110              0\n",
      "HSP90AA1    936  165275  166211              0\n",
      "IGFBP7      741  132212  132953              0\n",
      "SNX3        915  202386  203301              0\n",
      "TUBA1B     1124  286143  287267              0 \n",
      "\n",
      "\n",
      "Represents 26% of all LINC00682+ cells, 25% of all SYT1+ cells. 97% SYT1+; 90% TUBA1B+; 78% S100B+; 75% HSP90AA1+; 73% SNX3+; 61% GPX3+; 60% NCKAP1+; 59% IGFBP7+; 56% APOD+; 52% AQP1+; 52% BCL2+; 50% SORBS2+; 46% HSPA1A+; 44% CAMTA1+; 41% LINC00682+; 40% COX8A+; 39% MARCKSL1+; 39% TPD52L1+; 37% STIP1+; 34% HSPA1B+; 33% APOE+; 33% SKAP2+; 32% PEG10+; 31% RAMP1+; 29% CREM+; 24% JAZF1+; 24% RIMS1+; 23% KCNA2+ (>=1 counts)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adult Glia</th>\n",
       "      <th>B (Plasma RBC)</th>\n",
       "      <th>Branch A4 (IN)</th>\n",
       "      <th>Branch B1 (eMN)</th>\n",
       "      <th>Branch B3 (IPAN)</th>\n",
       "      <th>CLDN10+ cells</th>\n",
       "      <th>Contractile pericyte (PLN+)</th>\n",
       "      <th>Cycling B cell</th>\n",
       "      <th>Cycling Stromal</th>\n",
       "      <th>EC cells (NPW+)</th>\n",
       "      <th>Endothelial</th>\n",
       "      <th>Glia 1 (DHH+)</th>\n",
       "      <th>Glia 2 (ELN+)</th>\n",
       "      <th>Glia 3 (BCAN+)</th>\n",
       "      <th>ICC</th>\n",
       "      <th>ILC1</th>\n",
       "      <th>ILCP</th>\n",
       "      <th>Immature B</th>\n",
       "      <th>LTi-like NCR- ILC3</th>\n",
       "      <th>LZ GC cell</th>\n",
       "      <th>Lymphoid DC</th>\n",
       "      <th>MMP9+ Inflammatory macrophage</th>\n",
       "      <th>MPO+ mono-neutrophil</th>\n",
       "      <th>Macrophages</th>\n",
       "      <th>Mature venous EC</th>\n",
       "      <th>Mesenchymal</th>\n",
       "      <th>Neural</th>\n",
       "      <th>Pericyte</th>\n",
       "      <th>Proximal progenitor</th>\n",
       "      <th>SELL+ CD8 T</th>\n",
       "      <th>SMC (PLPP2+)</th>\n",
       "      <th>Stromal 1 (CCL11+)</th>\n",
       "      <th>Stromal 3 (C7+)</th>\n",
       "      <th>TRGV4 gdT</th>\n",
       "      <th>TRGV5/7 gdT</th>\n",
       "      <th>arterial capillary</th>\n",
       "      <th>cDC1</th>\n",
       "      <th>cycling EC</th>\n",
       "      <th>cycling neuroblast</th>\n",
       "      <th>cycling stromal</th>\n",
       "      <th>mLN Stroma (FMO2+)</th>\n",
       "      <th>myofibroblast</th>\n",
       "      <th>myofibroblast (RSPO2+)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>APOD</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>56%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOE</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AQP1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>52%</td>\n",
       "      <td></td>\n",
       "      <td>52%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>52%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAMTA1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>44%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COX8A</th>\n",
       "      <td></td>\n",
       "      <td>40%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>40%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>40%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CREM</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>29%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPX3</th>\n",
       "      <td>61%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>61%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSP90AA1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>75%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSPA1A</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td>46%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSPA1B</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGFBP7</th>\n",
       "      <td>59%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>59%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>59%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>59%</td>\n",
       "      <td>59%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAZF1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>24%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KCNA2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>23%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00682</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>41%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>41%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MARCKSL1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>39%</td>\n",
       "      <td>39%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCKAP1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>60%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PEG10</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>32%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>32%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>32%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAMP1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>31%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>31%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>31%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIMS1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>24%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S100B</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>78%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>78%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>78%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>78%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SKAP2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNX3</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>73%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SORBS2</th>\n",
       "      <td>50%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>50%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>50%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>50%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STIP1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>37%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYT1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>97%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TPD52L1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>39%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>39%</td>\n",
       "      <td></td>\n",
       "      <td>39%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TUBA1B</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>90%</td>\n",
       "      <td>90%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>90%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>90%</td>\n",
       "      <td>90%</td>\n",
       "      <td>90%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Adult Glia B (Plasma RBC) Branch A4 (IN) Branch B1 (eMN)  \\\n",
       "Gene                                                                 \n",
       "APOD                                                                 \n",
       "APOE                                                                 \n",
       "AQP1                                                                 \n",
       "BCL2                                           52%                   \n",
       "CAMTA1                                                               \n",
       "COX8A                           40%                                  \n",
       "CREM                                                                 \n",
       "GPX3             61%                                                 \n",
       "HSP90AA1                                                             \n",
       "HSPA1A                                                               \n",
       "HSPA1B                                                               \n",
       "IGFBP7           59%                                                 \n",
       "JAZF1                                                                \n",
       "KCNA2                                                                \n",
       "LINC00682                                                      41%   \n",
       "MARCKSL1                                                             \n",
       "NCKAP1                                                               \n",
       "PEG10                                                                \n",
       "RAMP1                                                          31%   \n",
       "RIMS1                                                                \n",
       "S100B                                                                \n",
       "SKAP2                                                                \n",
       "SNX3                                                                 \n",
       "SORBS2           50%                                                 \n",
       "STIP1                                                                \n",
       "SYT1                                                                 \n",
       "TPD52L1                                                              \n",
       "TUBA1B                                                               \n",
       "\n",
       "          Branch B3 (IPAN) CLDN10+ cells Contractile pericyte (PLN+)  \\\n",
       "Gene                                                                   \n",
       "APOD                                                                   \n",
       "APOE                                                                   \n",
       "AQP1                                                                   \n",
       "BCL2                                                                   \n",
       "CAMTA1                                                                 \n",
       "COX8A                                                                  \n",
       "CREM                                                                   \n",
       "GPX3                                                                   \n",
       "HSP90AA1                                                               \n",
       "HSPA1A                                                                 \n",
       "HSPA1B                                                                 \n",
       "IGFBP7                               59%                               \n",
       "JAZF1                                                                  \n",
       "KCNA2                                                                  \n",
       "LINC00682                                                              \n",
       "MARCKSL1                                                               \n",
       "NCKAP1                                                                 \n",
       "PEG10                                                                  \n",
       "RAMP1                                                                  \n",
       "RIMS1                                                                  \n",
       "S100B                                                                  \n",
       "SKAP2                  33%                                             \n",
       "SNX3                                                                   \n",
       "SORBS2                                                           50%   \n",
       "STIP1                                                                  \n",
       "SYT1                                                                   \n",
       "TPD52L1                                                                \n",
       "TUBA1B                                                                 \n",
       "\n",
       "          Cycling B cell Cycling Stromal EC cells (NPW+) Endothelial  \\\n",
       "Gene                                                                   \n",
       "APOD                                                                   \n",
       "APOE                                                                   \n",
       "AQP1                                                             52%   \n",
       "BCL2                                                                   \n",
       "CAMTA1                                                                 \n",
       "COX8A                40%                                               \n",
       "CREM                                                                   \n",
       "GPX3                                                                   \n",
       "HSP90AA1                                                               \n",
       "HSPA1A                                                                 \n",
       "HSPA1B                                                                 \n",
       "IGFBP7                                                                 \n",
       "JAZF1                                                                  \n",
       "KCNA2                                                                  \n",
       "LINC00682                                                              \n",
       "MARCKSL1                                                               \n",
       "NCKAP1                                                                 \n",
       "PEG10                                                32%               \n",
       "RAMP1                                                                  \n",
       "RIMS1                                                                  \n",
       "S100B                                                                  \n",
       "SKAP2                                                                  \n",
       "SNX3                                                                   \n",
       "SORBS2                                                                 \n",
       "STIP1                                                                  \n",
       "SYT1                                                                   \n",
       "TPD52L1                                                                \n",
       "TUBA1B               90%             90%                               \n",
       "\n",
       "          Glia 1 (DHH+) Glia 2 (ELN+) Glia 3 (BCAN+)  ICC ILC1 ILCP  \\\n",
       "Gene                                                                  \n",
       "APOD                                                                  \n",
       "APOE                                             33%                  \n",
       "AQP1                              52%                                 \n",
       "BCL2                                                                  \n",
       "CAMTA1                                                          44%   \n",
       "COX8A                                                                 \n",
       "CREM                                                                  \n",
       "GPX3                                                                  \n",
       "HSP90AA1                                                   75%        \n",
       "HSPA1A                                                                \n",
       "HSPA1B                                                                \n",
       "IGFBP7                                                          59%   \n",
       "JAZF1                                                                 \n",
       "KCNA2                                                                 \n",
       "LINC00682                                                             \n",
       "MARCKSL1                                                              \n",
       "NCKAP1                                                                \n",
       "PEG10                                                           32%   \n",
       "RAMP1                                                 31%             \n",
       "RIMS1                                                                 \n",
       "S100B               78%                                               \n",
       "SKAP2                                                                 \n",
       "SNX3                                                                  \n",
       "SORBS2                                                                \n",
       "STIP1                                                                 \n",
       "SYT1                                                                  \n",
       "TPD52L1                                          39%                  \n",
       "TUBA1B                                                                \n",
       "\n",
       "          Immature B LTi-like NCR- ILC3 LZ GC cell Lymphoid DC  \\\n",
       "Gene                                                             \n",
       "APOD                                                             \n",
       "APOE                                                             \n",
       "AQP1                                                             \n",
       "BCL2                                                             \n",
       "CAMTA1                                                           \n",
       "COX8A                                                            \n",
       "CREM                                                             \n",
       "GPX3                                                             \n",
       "HSP90AA1                                                         \n",
       "HSPA1A                                                           \n",
       "HSPA1B                                                           \n",
       "IGFBP7                                                           \n",
       "JAZF1            24%                                             \n",
       "KCNA2                                                            \n",
       "LINC00682                                                        \n",
       "MARCKSL1                                       39%         39%   \n",
       "NCKAP1                              60%                          \n",
       "PEG10                                                            \n",
       "RAMP1                                                            \n",
       "RIMS1                                                            \n",
       "S100B                                                            \n",
       "SKAP2                                                            \n",
       "SNX3                                                             \n",
       "SORBS2                                                           \n",
       "STIP1                                                            \n",
       "SYT1                                                             \n",
       "TPD52L1                                                          \n",
       "TUBA1B                                                           \n",
       "\n",
       "          MMP9+ Inflammatory macrophage MPO+ mono-neutrophil Macrophages  \\\n",
       "Gene                                                                       \n",
       "APOD                                                                       \n",
       "APOE                                33%                                    \n",
       "AQP1                                                                       \n",
       "BCL2                                                                       \n",
       "CAMTA1                                                                     \n",
       "COX8A                                                    40%               \n",
       "CREM                                                                       \n",
       "GPX3                                                                       \n",
       "HSP90AA1                                                                   \n",
       "HSPA1A                                                               46%   \n",
       "HSPA1B                                                                     \n",
       "IGFBP7                                                                     \n",
       "JAZF1                                                                      \n",
       "KCNA2                                                                23%   \n",
       "LINC00682                                                                  \n",
       "MARCKSL1                                                                   \n",
       "NCKAP1                                                                     \n",
       "PEG10                                                                      \n",
       "RAMP1                                                                      \n",
       "RIMS1                                                                      \n",
       "S100B                                                                      \n",
       "SKAP2                                                                      \n",
       "SNX3                                                                       \n",
       "SORBS2                                                                     \n",
       "STIP1                                                                      \n",
       "SYT1                                                                       \n",
       "TPD52L1                                                                    \n",
       "TUBA1B                                                                     \n",
       "\n",
       "          Mature venous EC Mesenchymal Neural Pericyte Proximal progenitor  \\\n",
       "Gene                                                                         \n",
       "APOD                                                                         \n",
       "APOE                                      33%                          33%   \n",
       "AQP1                                                                         \n",
       "BCL2                                                                         \n",
       "CAMTA1                                                                       \n",
       "COX8A                                                                        \n",
       "CREM                                                                         \n",
       "GPX3                                                                         \n",
       "HSP90AA1                                                                     \n",
       "HSPA1A                                                                       \n",
       "HSPA1B                                                                       \n",
       "IGFBP7                                    59%      59%                       \n",
       "JAZF1                                                                        \n",
       "KCNA2                                                                        \n",
       "LINC00682                                 41%                                \n",
       "MARCKSL1                                                                     \n",
       "NCKAP1                                                                       \n",
       "PEG10                                                                        \n",
       "RAMP1                                     31%                                \n",
       "RIMS1                                                                        \n",
       "S100B                                     78%                                \n",
       "SKAP2                                                                        \n",
       "SNX3                                                                         \n",
       "SORBS2                             50%                                       \n",
       "STIP1                                                                        \n",
       "SYT1                                                                         \n",
       "TPD52L1                39%                39%                                \n",
       "TUBA1B                                                                       \n",
       "\n",
       "          SELL+ CD8 T SMC (PLPP2+) Stromal 1 (CCL11+) Stromal 3 (C7+)  \\\n",
       "Gene                                                                    \n",
       "APOD                                                                    \n",
       "APOE                                              33%                   \n",
       "AQP1                                                                    \n",
       "BCL2                                                                    \n",
       "CAMTA1                                                                  \n",
       "COX8A                                                                   \n",
       "CREM                                                                    \n",
       "GPX3                                                                    \n",
       "HSP90AA1                                                                \n",
       "HSPA1A                                                                  \n",
       "HSPA1B                                                                  \n",
       "IGFBP7                                                                  \n",
       "JAZF1                                                                   \n",
       "KCNA2                                                                   \n",
       "LINC00682                                                               \n",
       "MARCKSL1                                                                \n",
       "NCKAP1                                                                  \n",
       "PEG10                                                             32%   \n",
       "RAMP1                                                                   \n",
       "RIMS1                          24%                                      \n",
       "S100B             78%                                                   \n",
       "SKAP2                                                                   \n",
       "SNX3                                                                    \n",
       "SORBS2                                                                  \n",
       "STIP1                                                                   \n",
       "SYT1                                                                    \n",
       "TPD52L1                                                                 \n",
       "TUBA1B                                                                  \n",
       "\n",
       "          TRGV4 gdT TRGV5/7 gdT arterial capillary cDC1 cycling EC  \\\n",
       "Gene                                                                 \n",
       "APOD                                                                 \n",
       "APOE                                                                 \n",
       "AQP1                                                                 \n",
       "BCL2                                                                 \n",
       "CAMTA1                                                               \n",
       "COX8A                                                                \n",
       "CREM            29%                                                  \n",
       "GPX3                                                                 \n",
       "HSP90AA1                                                             \n",
       "HSPA1A                      46%                                      \n",
       "HSPA1B                                                               \n",
       "IGFBP7                                                               \n",
       "JAZF1                                                                \n",
       "KCNA2                                                                \n",
       "LINC00682                                                            \n",
       "MARCKSL1                                                             \n",
       "NCKAP1                                                               \n",
       "PEG10                                                                \n",
       "RAMP1                                                                \n",
       "RIMS1                                                                \n",
       "S100B                                               78%              \n",
       "SKAP2                                                                \n",
       "SNX3                                                73%              \n",
       "SORBS2                                         50%                   \n",
       "STIP1                       37%                                      \n",
       "SYT1                                                                 \n",
       "TPD52L1                                                              \n",
       "TUBA1B          90%                                            90%   \n",
       "\n",
       "          cycling neuroblast cycling stromal mLN Stroma (FMO2+) myofibroblast  \\\n",
       "Gene                                                                            \n",
       "APOD                                                        56%                 \n",
       "APOE                                                                            \n",
       "AQP1                                                                            \n",
       "BCL2                                                                            \n",
       "CAMTA1                                                                          \n",
       "COX8A                                                                           \n",
       "CREM                                                                            \n",
       "GPX3                                                        61%                 \n",
       "HSP90AA1                                                                        \n",
       "HSPA1A                                                                    46%   \n",
       "HSPA1B                                                                          \n",
       "IGFBP7                                                                          \n",
       "JAZF1                                                                           \n",
       "KCNA2                                                                           \n",
       "LINC00682                                                                       \n",
       "MARCKSL1                                                                        \n",
       "NCKAP1                                                                          \n",
       "PEG10                                                                           \n",
       "RAMP1                                                                           \n",
       "RIMS1                                                                           \n",
       "S100B                                                                           \n",
       "SKAP2                                                                           \n",
       "SNX3                                                                            \n",
       "SORBS2                                                                          \n",
       "STIP1                                                                           \n",
       "SYT1                                                                            \n",
       "TPD52L1                                                                         \n",
       "TUBA1B                   90%             90%                                    \n",
       "\n",
       "          myofibroblast (RSPO2+)  \n",
       "Gene                              \n",
       "APOD                              \n",
       "APOE                              \n",
       "AQP1                              \n",
       "BCL2                              \n",
       "CAMTA1                            \n",
       "COX8A                             \n",
       "CREM                              \n",
       "GPX3                              \n",
       "HSP90AA1                          \n",
       "HSPA1A                       46%  \n",
       "HSPA1B                       34%  \n",
       "IGFBP7                            \n",
       "JAZF1                             \n",
       "KCNA2                             \n",
       "LINC00682                         \n",
       "MARCKSL1                          \n",
       "NCKAP1                            \n",
       "PEG10                             \n",
       "RAMP1                             \n",
       "RIMS1                             \n",
       "S100B                             \n",
       "SKAP2                             \n",
       "SNX3                              \n",
       "SORBS2                            \n",
       "STIP1                             \n",
       "SYT1                         97%  \n",
       "TPD52L1                           \n",
       "TUBA1B                            "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key_cluster = \"31\"\n",
    "count_threshold = 1\n",
    "\n",
    "print(tgdf.loc[key_cluster].iloc[:, :-3])\n",
    "_, percs_exp, n_exp, genes, msg = self.print_markers(\n",
    "    key_cluster, assign, col_cell_type=c_t, lfc_threshold=2,\n",
    "    count_threshold=count_threshold, p_threshold=1e-15,\n",
    "    print_threshold=15, n_top_genes=40)\n",
    "print(\"\\n\\n\", genes, \"\\n\\n\", n_exp.applymap(int), \"\\n\\n\")\n",
    "print(msg)\n",
    "percs_exp.applymap(lambda x: x if x == \"\" else str(int(x)) + \"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### By Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key_cluster = \"8\"\n",
    "print(tgdf.loc[key_cluster].iloc[:, :-3])\n",
    "_, percs_exp, n_exp, genes, msg = self.print_markers(\n",
    "    key_cluster, assign, col_cell_type=c_t, lfc_threshold=None,\n",
    "    count_threshold=1, p_threshold=1, n_top_genes=[\n",
    "        \"LGR5\", \"CCL25\", \"CDCA7\", \"TOP2A\",\n",
    "        \"MUC2\", \"MUC5B\", \"MUC1\", \"MUC4\", \"KLK1\", \"ZG16\", \"BEST2\",\n",
    "        \"BEST4\", \"CCL20\"])\n",
    "print(\"\\n\\n\", genes, \"\\n\\n\", n_exp.applymap(int), \"\\n\\n\")\n",
    "print(msg)\n",
    "print(percs_exp.stack().replace(\"\", np.nan).dropna().rename_axis([\n",
    "    \"Gene\", \"Annotation\"]).groupby(\"Annotation\").apply(\n",
    "        lambda x: x.sort_values(ascending=False).apply(\n",
    "            lambda y: f\"{int(y)}%\")).reset_index(2, drop=True))\n",
    "percs_exp.applymap(lambda x: x if x == \"\" else str(int(x)) + \"%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key_cluster = \"22\"\n",
    "_, percs_exp, n_exp, genes, msg = self.print_markers(\n",
    "    key_cluster, assign, col_cell_type=c_t, lfc_threshold=None,\n",
    "    count_threshold=1, p_threshold=1, n_top_genes=list(assign[\n",
    "        assign.Bin == \"SMC\"].index.intersection(self.rna.var_names)))\n",
    "print(\"\\n\\n\", genes, \"\\n\\n\", n_exp.applymap(int), \"\\n\\n\")\n",
    "print(msg)\n",
    "print(percs_exp.stack().replace(\"\", np.nan).dropna().rename_axis([\n",
    "    \"Gene\", \"Annotation\"]).groupby(\"Annotation\").apply(\n",
    "        lambda x: x.sort_values(ascending=False).apply(\n",
    "            lambda y: f\"{int(y)}%\")).reset_index(2, drop=True))\n",
    "percs_exp.applymap(lambda x: x if x == \"\" else str(int(x)) + \"%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key_cluster = \"15\"\n",
    "_, percs_exp, n_exp, genes, msg = self.print_markers(\n",
    "    key_cluster, assign, col_cell_type=c_t, lfc_threshold=None,\n",
    "    count_threshold=1, p_threshold=1, n_top_genes=list(assign[\n",
    "        assign.Bucket == \"Granulocyte\"].index.intersection(\n",
    "            self.rna.var_names)))\n",
    "print(\"\\n\\n\", genes, \"\\n\\n\", n_exp.applymap(int), \"\\n\\n\")\n",
    "print(msg)\n",
    "print(percs_exp.stack().replace(\"\", np.nan).dropna().rename_axis([\n",
    "    \"Gene\", \"Annotation\"]).groupby(\"Annotation\").apply(\n",
    "        lambda x: x.sort_values(ascending=False).apply(\n",
    "            lambda y: f\"{int(y)}%\")).reset_index(2, drop=True))\n",
    "percs_exp.applymap(lambda x: x if x == \"\" else str(int(x)) + \"%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "================================================================================\n",
      "Count Threshold: 1\n",
      "================================================================================\n",
      "\n",
      "\n",
      " group\n",
      "Adult Glia                                                 IGFBP7, MT1E, SELENOM, SORBS2, ZFP36, GPX3\n",
      "BEST2+ Goblet cell                                                                             RNASE1\n",
      "Branch A4 (IN)                                                                    RNASE1, BCL2, UNC5B\n",
      "Branch B1 (eMN)                                                                      LINC00682, RAMP1\n",
      "Branch B3 (IPAN)                                                                                SKAP2\n",
      "CLDN10+ cells                                                                                  IGFBP7\n",
      "Contractile pericyte (PLN+)                                                                    SORBS2\n",
      "Cycling B cell                                                                          HMGB2, TUBA1B\n",
      "Cycling Stromal                                                                  TOP2A, HMGB2, TUBA1B\n",
      "Differentiating glia                                                                             CCL2\n",
      "Endothelial                                                                               UNC5B, AQP1\n",
      "Epithelial                                                                                     RNASE1\n",
      "Epithelial (TA)                                                                                 TOP2A\n",
      "Glia 1 (DHH+)                                            S100B, COL4A1, COL4A2, COL1A1, COL1A2, NR4A1\n",
      "Glia 2 (ELN+)                                                                 CD44, ETS1, TGFB1, AQP1\n",
      "Glia 3 (BCAN+)                                                                    BCAN, APOE, TPD52L1\n",
      "ICC                                                                                       RAMP1, BCAN\n",
      "ILCP                                                                                           IGFBP7\n",
      "IgG plasma cell                                                                               SELENOM\n",
      "LEC2 (MADCAM1+)                                                                               SELENOM\n",
      "LEC6 (ADAMTS4+)                                                                                  MT1E\n",
      "LYVE1+ Macrophage                                                                        RNASE1, CCL2\n",
      "MMP9+ Inflammatory macrophage                                                                    APOE\n",
      "MPO+ mono-neutrophil                                                                            HMGB2\n",
      "Mature venous EC                                                                              TPD52L1\n",
      "Mesenchymal                                                                                    SORBS2\n",
      "Mesothelium (PRG4+)                                                                              MT1E\n",
      "Monocytes                                                                                        CD44\n",
      "Neural                           IGFBP7, LINC00682, RAMP1, S100B, COL1A1, COL1A2, BCAN, APOE, TPD52L1\n",
      "Pericyte                                                                                       IGFBP7\n",
      "Proximal progenitor                                                                              APOE\n",
      "RBC                                                                                              MT1E\n",
      "SELL+ CD8 T                                                                                     S100B\n",
      "SMC (PLPP2+)                                                                                  C5orf30\n",
      "Stem cells                                                                                      CDCA7\n",
      "Stromal 1 (CCL11+)                                                                               APOE\n",
      "Stromal 3 (C7+)                                                                                COL1A1\n",
      "Stromal 4 (MMP1+)                                                                             SELENOM\n",
      "TRGV4 gdT                                                                                      TUBA1B\n",
      "TRGV5/7 gdT                                                                                     HMGB2\n",
      "angiogenic pericyte                                                                    COL4A1, COL4A2\n",
      "arterial capillary                                                                             SORBS2\n",
      "cDC1                                                                                            S100B\n",
      "cycling EC                                                                       HMGB2, H2AFX, TUBA1B\n",
      "cycling ENCC/glia                                                                      CDCA7, C5orf30\n",
      "cycling neuroblast                                                        TOP2A, HMGB2, H2AFX, TUBA1B\n",
      "cycling stromal                                                           TOP2A, HMGB2, H2AFX, TUBA1B\n",
      "mLN Stroma (FMO2+)                                                                        GPX3, NR4A1\n",
      "dtype: object \n",
      "\n",
      "              31   Other   Total  Percent_Total\n",
      "Gene                                          \n",
      "LINC00682   518    1423    1941             26\n",
      "TPD52L1     488   13387   13875              3\n",
      "UNC5B       230    8330    8560              2\n",
      "S100B       970   41524   42494              2\n",
      "C5orf30     153   11148   11301              1\n",
      "SORBS2      625   49933   50558              1\n",
      "BCL2        646   65084   65730              0\n",
      "NR4A1       209   23467   23676              0\n",
      "RAMP1       394   44726   45120              0\n",
      "APOE        411   48293   48704              0\n",
      "SKAP2       410   49517   49927              0\n",
      "CCL2        341   44052   44393              0\n",
      "AQP1        652   85200   85852              0\n",
      "GPX3        769  109503  110272              0\n",
      "SELENOM     275   40136   40411              0\n",
      "BCAN         17    2772    2789              0\n",
      "IGFBP7      741  132212  132953              0\n",
      "H2AFX        57   10926   10983              0\n",
      "CD44        574  130427  131001              0\n",
      "HMGB2       582  137561  138143              0\n",
      "TUBA1B     1124  286143  287267              0\n",
      "COL1A1      857  224424  225281              0\n",
      "COL1A2      958  256624  257582              0\n",
      "TGFB1       392  107279  107671              0\n",
      "MT1E         72   21102   21174              0\n",
      "COL4A1      498  153580  154078              0\n",
      "COL4A2      590  195601  196191              0\n",
      "ETS1        438  163501  163939              0\n",
      "RNASE1      264  100976  101240              0\n",
      "ZFP36       529  255924  256453              0\n",
      "TOP2A         8   11590   11598              0\n",
      "CDCA7         7   18418   18425              0 \n",
      "\n",
      "\n",
      "90% TUBA1B+; 78% S100B+; 77% COL1A2+; 69% COL1A1+ (>=1 counts)\n",
      "Annotation                     Gene     \n",
      "Adult Glia                     GPX3         61%\n",
      "                               IGFBP7       59%\n",
      "                               SORBS2       50%\n",
      "                               ZFP36        42%\n",
      "                               SELENOM      22%\n",
      "                               MT1E          5%\n",
      "BEST2+ Goblet cell             RNASE1       21%\n",
      "Branch A4 (IN)                 BCL2         52%\n",
      "                               RNASE1       21%\n",
      "                               UNC5B        18%\n",
      "Branch B1 (eMN)                LINC00682    41%\n",
      "                               RAMP1        31%\n",
      "Branch B3 (IPAN)               SKAP2        33%\n",
      "CLDN10+ cells                  IGFBP7       59%\n",
      "Contractile pericyte (PLN+)    SORBS2       50%\n",
      "Cycling B cell                 TUBA1B       90%\n",
      "                               HMGB2        46%\n",
      "Cycling Stromal                TUBA1B       90%\n",
      "                               HMGB2        46%\n",
      "                               TOP2A         0%\n",
      "Differentiating glia           CCL2         27%\n",
      "Endothelial                    AQP1         52%\n",
      "                               UNC5B        18%\n",
      "Epithelial                     RNASE1       21%\n",
      "Epithelial (TA)                TOP2A         0%\n",
      "Glia 1 (DHH+)                  S100B        78%\n",
      "                               COL1A2       77%\n",
      "                               COL1A1       69%\n",
      "                               COL4A2       47%\n",
      "                               COL4A1       40%\n",
      "                               NR4A1        16%\n",
      "Glia 2 (ELN+)                  AQP1         52%\n",
      "                               CD44         46%\n",
      "                               ETS1         35%\n",
      "                               TGFB1        31%\n",
      "Glia 3 (BCAN+)                 TPD52L1      39%\n",
      "                               APOE         33%\n",
      "                               BCAN          1%\n",
      "ICC                            RAMP1        31%\n",
      "                               BCAN          1%\n",
      "ILCP                           IGFBP7       59%\n",
      "IgG plasma cell                SELENOM      22%\n",
      "LEC2 (MADCAM1+)                SELENOM      22%\n",
      "LEC6 (ADAMTS4+)                MT1E          5%\n",
      "LYVE1+ Macrophage              CCL2         27%\n",
      "                               RNASE1       21%\n",
      "MMP9+ Inflammatory macrophage  APOE         33%\n",
      "MPO+ mono-neutrophil           HMGB2        46%\n",
      "Mature venous EC               TPD52L1      39%\n",
      "Mesenchymal                    SORBS2       50%\n",
      "Mesothelium (PRG4+)            MT1E          5%\n",
      "Monocytes                      CD44         46%\n",
      "Neural                         S100B        78%\n",
      "                               COL1A2       77%\n",
      "                               COL1A1       69%\n",
      "                               IGFBP7       59%\n",
      "                               LINC00682    41%\n",
      "                               TPD52L1      39%\n",
      "                               APOE         33%\n",
      "                               RAMP1        31%\n",
      "                               BCAN          1%\n",
      "Pericyte                       IGFBP7       59%\n",
      "Proximal progenitor            APOE         33%\n",
      "RBC                            MT1E          5%\n",
      "SELL+ CD8 T                    S100B        78%\n",
      "SMC (PLPP2+)                   C5orf30      12%\n",
      "Stem cells                     CDCA7         0%\n",
      "Stromal 1 (CCL11+)             APOE         33%\n",
      "Stromal 3 (C7+)                COL1A1       69%\n",
      "Stromal 4 (MMP1+)              SELENOM      22%\n",
      "TRGV4 gdT                      TUBA1B       90%\n",
      "TRGV5/7 gdT                    HMGB2        46%\n",
      "angiogenic pericyte            COL4A2       47%\n",
      "                               COL4A1       40%\n",
      "arterial capillary             SORBS2       50%\n",
      "cDC1                           S100B        78%\n",
      "cycling EC                     TUBA1B       90%\n",
      "                               HMGB2        46%\n",
      "                               H2AFX         4%\n",
      "cycling ENCC/glia              C5orf30      12%\n",
      "                               CDCA7         0%\n",
      "cycling neuroblast             TUBA1B       90%\n",
      "                               HMGB2        46%\n",
      "                               H2AFX         4%\n",
      "                               TOP2A         0%\n",
      "cycling stromal                TUBA1B       90%\n",
      "                               HMGB2        46%\n",
      "                               H2AFX         4%\n",
      "                               TOP2A         0%\n",
      "mLN Stroma (FMO2+)             GPX3         61%\n",
      "                               NR4A1        16%\n",
      "dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adult Glia</th>\n",
       "      <th>BEST2+ Goblet cell</th>\n",
       "      <th>Branch A4 (IN)</th>\n",
       "      <th>Branch B1 (eMN)</th>\n",
       "      <th>Branch B3 (IPAN)</th>\n",
       "      <th>CLDN10+ cells</th>\n",
       "      <th>Contractile pericyte (PLN+)</th>\n",
       "      <th>Cycling B cell</th>\n",
       "      <th>Cycling Stromal</th>\n",
       "      <th>Differentiating glia</th>\n",
       "      <th>Endothelial</th>\n",
       "      <th>Epithelial</th>\n",
       "      <th>Epithelial (TA)</th>\n",
       "      <th>Glia 1 (DHH+)</th>\n",
       "      <th>Glia 2 (ELN+)</th>\n",
       "      <th>Glia 3 (BCAN+)</th>\n",
       "      <th>ICC</th>\n",
       "      <th>ILCP</th>\n",
       "      <th>IgG plasma cell</th>\n",
       "      <th>LEC2 (MADCAM1+)</th>\n",
       "      <th>LEC6 (ADAMTS4+)</th>\n",
       "      <th>LYVE1+ Macrophage</th>\n",
       "      <th>MMP9+ Inflammatory macrophage</th>\n",
       "      <th>MPO+ mono-neutrophil</th>\n",
       "      <th>Mature venous EC</th>\n",
       "      <th>Mesenchymal</th>\n",
       "      <th>Mesothelium (PRG4+)</th>\n",
       "      <th>Monocytes</th>\n",
       "      <th>Neural</th>\n",
       "      <th>Pericyte</th>\n",
       "      <th>Proximal progenitor</th>\n",
       "      <th>RBC</th>\n",
       "      <th>SELL+ CD8 T</th>\n",
       "      <th>SMC (PLPP2+)</th>\n",
       "      <th>Stem cells</th>\n",
       "      <th>Stromal 1 (CCL11+)</th>\n",
       "      <th>Stromal 3 (C7+)</th>\n",
       "      <th>Stromal 4 (MMP1+)</th>\n",
       "      <th>TRGV4 gdT</th>\n",
       "      <th>TRGV5/7 gdT</th>\n",
       "      <th>angiogenic pericyte</th>\n",
       "      <th>arterial capillary</th>\n",
       "      <th>cDC1</th>\n",
       "      <th>cycling EC</th>\n",
       "      <th>cycling ENCC/glia</th>\n",
       "      <th>cycling neuroblast</th>\n",
       "      <th>cycling stromal</th>\n",
       "      <th>mLN Stroma (FMO2+)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>APOE</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AQP1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>52%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>52%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCAN</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1%</td>\n",
       "      <td>1%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>52%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C5orf30</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>12%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>12%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>27%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>27%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD44</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDCA7</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>69%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>69%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>69%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>77%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>77%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL4A1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>40%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>40%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL4A2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>47%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>47%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETS1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>35%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPX3</th>\n",
       "      <td>61%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>61%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2AFX</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4%</td>\n",
       "      <td></td>\n",
       "      <td>4%</td>\n",
       "      <td>4%</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMGB2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td>46%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td></td>\n",
       "      <td>46%</td>\n",
       "      <td>46%</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGFBP7</th>\n",
       "      <td>59%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>59%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>59%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>59%</td>\n",
       "      <td>59%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00682</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>41%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>41%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT1E</th>\n",
       "      <td>5%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NR4A1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>16%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>16%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAMP1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>31%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>31%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>31%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNASE1</th>\n",
       "      <td></td>\n",
       "      <td>21%</td>\n",
       "      <td>21%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>21%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>21%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S100B</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>78%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>78%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>78%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>78%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SELENOM</th>\n",
       "      <td>22%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>22%</td>\n",
       "      <td>22%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>22%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SKAP2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SORBS2</th>\n",
       "      <td>50%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>50%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>50%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>50%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFB1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>31%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TOP2A</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0%</td>\n",
       "      <td>0%</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TPD52L1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>39%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>39%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>39%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TUBA1B</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>90%</td>\n",
       "      <td>90%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>90%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>90%</td>\n",
       "      <td></td>\n",
       "      <td>90%</td>\n",
       "      <td>90%</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNC5B</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>18%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>18%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFP36</th>\n",
       "      <td>42%</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Adult Glia BEST2+ Goblet cell Branch A4 (IN) Branch B1 (eMN)  \\\n",
       "Gene                                                                     \n",
       "APOE                                                                     \n",
       "AQP1                                                                     \n",
       "BCAN                                                                     \n",
       "BCL2                                               52%                   \n",
       "C5orf30                                                                  \n",
       "CCL2                                                                     \n",
       "CD44                                                                     \n",
       "CDCA7                                                                    \n",
       "COL1A1                                                                   \n",
       "COL1A2                                                                   \n",
       "COL4A1                                                                   \n",
       "COL4A2                                                                   \n",
       "ETS1                                                                     \n",
       "GPX3             61%                                                     \n",
       "H2AFX                                                                    \n",
       "HMGB2                                                                    \n",
       "IGFBP7           59%                                                     \n",
       "LINC00682                                                          41%   \n",
       "MT1E              5%                                                     \n",
       "NR4A1                                                                    \n",
       "RAMP1                                                              31%   \n",
       "RNASE1                              21%            21%                   \n",
       "S100B                                                                    \n",
       "SELENOM          22%                                                     \n",
       "SKAP2                                                                    \n",
       "SORBS2           50%                                                     \n",
       "TGFB1                                                                    \n",
       "TOP2A                                                                    \n",
       "TPD52L1                                                                  \n",
       "TUBA1B                                                                   \n",
       "UNC5B                                              18%                   \n",
       "ZFP36            42%                                                     \n",
       "\n",
       "          Branch B3 (IPAN) CLDN10+ cells Contractile pericyte (PLN+)  \\\n",
       "Gene                                                                   \n",
       "APOE                                                                   \n",
       "AQP1                                                                   \n",
       "BCAN                                                                   \n",
       "BCL2                                                                   \n",
       "C5orf30                                                                \n",
       "CCL2                                                                   \n",
       "CD44                                                                   \n",
       "CDCA7                                                                  \n",
       "COL1A1                                                                 \n",
       "COL1A2                                                                 \n",
       "COL4A1                                                                 \n",
       "COL4A2                                                                 \n",
       "ETS1                                                                   \n",
       "GPX3                                                                   \n",
       "H2AFX                                                                  \n",
       "HMGB2                                                                  \n",
       "IGFBP7                               59%                               \n",
       "LINC00682                                                              \n",
       "MT1E                                                                   \n",
       "NR4A1                                                                  \n",
       "RAMP1                                                                  \n",
       "RNASE1                                                                 \n",
       "S100B                                                                  \n",
       "SELENOM                                                                \n",
       "SKAP2                  33%                                             \n",
       "SORBS2                                                           50%   \n",
       "TGFB1                                                                  \n",
       "TOP2A                                                                  \n",
       "TPD52L1                                                                \n",
       "TUBA1B                                                                 \n",
       "UNC5B                                                                  \n",
       "ZFP36                                                                  \n",
       "\n",
       "          Cycling B cell Cycling Stromal Differentiating glia Endothelial  \\\n",
       "Gene                                                                        \n",
       "APOE                                                                        \n",
       "AQP1                                                                  52%   \n",
       "BCAN                                                                        \n",
       "BCL2                                                                        \n",
       "C5orf30                                                                     \n",
       "CCL2                                                      27%               \n",
       "CD44                                                                        \n",
       "CDCA7                                                                       \n",
       "COL1A1                                                                      \n",
       "COL1A2                                                                      \n",
       "COL4A1                                                                      \n",
       "COL4A2                                                                      \n",
       "ETS1                                                                        \n",
       "GPX3                                                                        \n",
       "H2AFX                                                                       \n",
       "HMGB2                46%             46%                                    \n",
       "IGFBP7                                                                      \n",
       "LINC00682                                                                   \n",
       "MT1E                                                                        \n",
       "NR4A1                                                                       \n",
       "RAMP1                                                                       \n",
       "RNASE1                                                                      \n",
       "S100B                                                                       \n",
       "SELENOM                                                                     \n",
       "SKAP2                                                                       \n",
       "SORBS2                                                                      \n",
       "TGFB1                                                                       \n",
       "TOP2A                                 0%                                    \n",
       "TPD52L1                                                                     \n",
       "TUBA1B               90%             90%                                    \n",
       "UNC5B                                                                 18%   \n",
       "ZFP36                                                                       \n",
       "\n",
       "          Epithelial Epithelial (TA) Glia 1 (DHH+) Glia 2 (ELN+)  \\\n",
       "Gene                                                               \n",
       "APOE                                                               \n",
       "AQP1                                                         52%   \n",
       "BCAN                                                               \n",
       "BCL2                                                               \n",
       "C5orf30                                                            \n",
       "CCL2                                                               \n",
       "CD44                                                         46%   \n",
       "CDCA7                                                              \n",
       "COL1A1                                         69%                 \n",
       "COL1A2                                         77%                 \n",
       "COL4A1                                         40%                 \n",
       "COL4A2                                         47%                 \n",
       "ETS1                                                         35%   \n",
       "GPX3                                                               \n",
       "H2AFX                                                              \n",
       "HMGB2                                                              \n",
       "IGFBP7                                                             \n",
       "LINC00682                                                          \n",
       "MT1E                                                               \n",
       "NR4A1                                          16%                 \n",
       "RAMP1                                                              \n",
       "RNASE1           21%                                               \n",
       "S100B                                          78%                 \n",
       "SELENOM                                                            \n",
       "SKAP2                                                              \n",
       "SORBS2                                                             \n",
       "TGFB1                                                        31%   \n",
       "TOP2A                             0%                               \n",
       "TPD52L1                                                            \n",
       "TUBA1B                                                             \n",
       "UNC5B                                                              \n",
       "ZFP36                                                              \n",
       "\n",
       "          Glia 3 (BCAN+)  ICC ILCP IgG plasma cell LEC2 (MADCAM1+)  \\\n",
       "Gene                                                                 \n",
       "APOE                 33%                                             \n",
       "AQP1                                                                 \n",
       "BCAN                  1%   1%                                        \n",
       "BCL2                                                                 \n",
       "C5orf30                                                              \n",
       "CCL2                                                                 \n",
       "CD44                                                                 \n",
       "CDCA7                                                                \n",
       "COL1A1                                                               \n",
       "COL1A2                                                               \n",
       "COL4A1                                                               \n",
       "COL4A2                                                               \n",
       "ETS1                                                                 \n",
       "GPX3                                                                 \n",
       "H2AFX                                                                \n",
       "HMGB2                                                                \n",
       "IGFBP7                         59%                                   \n",
       "LINC00682                                                            \n",
       "MT1E                                                                 \n",
       "NR4A1                                                                \n",
       "RAMP1                     31%                                        \n",
       "RNASE1                                                               \n",
       "S100B                                                                \n",
       "SELENOM                                        22%             22%   \n",
       "SKAP2                                                                \n",
       "SORBS2                                                               \n",
       "TGFB1                                                                \n",
       "TOP2A                                                                \n",
       "TPD52L1              39%                                             \n",
       "TUBA1B                                                               \n",
       "UNC5B                                                                \n",
       "ZFP36                                                                \n",
       "\n",
       "          LEC6 (ADAMTS4+) LYVE1+ Macrophage MMP9+ Inflammatory macrophage  \\\n",
       "Gene                                                                        \n",
       "APOE                                                                  33%   \n",
       "AQP1                                                                        \n",
       "BCAN                                                                        \n",
       "BCL2                                                                        \n",
       "C5orf30                                                                     \n",
       "CCL2                                    27%                                 \n",
       "CD44                                                                        \n",
       "CDCA7                                                                       \n",
       "COL1A1                                                                      \n",
       "COL1A2                                                                      \n",
       "COL4A1                                                                      \n",
       "COL4A2                                                                      \n",
       "ETS1                                                                        \n",
       "GPX3                                                                        \n",
       "H2AFX                                                                       \n",
       "HMGB2                                                                       \n",
       "IGFBP7                                                                      \n",
       "LINC00682                                                                   \n",
       "MT1E                   5%                                                   \n",
       "NR4A1                                                                       \n",
       "RAMP1                                                                       \n",
       "RNASE1                                  21%                                 \n",
       "S100B                                                                       \n",
       "SELENOM                                                                     \n",
       "SKAP2                                                                       \n",
       "SORBS2                                                                      \n",
       "TGFB1                                                                       \n",
       "TOP2A                                                                       \n",
       "TPD52L1                                                                     \n",
       "TUBA1B                                                                      \n",
       "UNC5B                                                                       \n",
       "ZFP36                                                                       \n",
       "\n",
       "          MPO+ mono-neutrophil Mature venous EC Mesenchymal  \\\n",
       "Gene                                                          \n",
       "APOE                                                          \n",
       "AQP1                                                          \n",
       "BCAN                                                          \n",
       "BCL2                                                          \n",
       "C5orf30                                                       \n",
       "CCL2                                                          \n",
       "CD44                                                          \n",
       "CDCA7                                                         \n",
       "COL1A1                                                        \n",
       "COL1A2                                                        \n",
       "COL4A1                                                        \n",
       "COL4A2                                                        \n",
       "ETS1                                                          \n",
       "GPX3                                                          \n",
       "H2AFX                                                         \n",
       "HMGB2                      46%                                \n",
       "IGFBP7                                                        \n",
       "LINC00682                                                     \n",
       "MT1E                                                          \n",
       "NR4A1                                                         \n",
       "RAMP1                                                         \n",
       "RNASE1                                                        \n",
       "S100B                                                         \n",
       "SELENOM                                                       \n",
       "SKAP2                                                         \n",
       "SORBS2                                                  50%   \n",
       "TGFB1                                                         \n",
       "TOP2A                                                         \n",
       "TPD52L1                                     39%               \n",
       "TUBA1B                                                        \n",
       "UNC5B                                                         \n",
       "ZFP36                                                         \n",
       "\n",
       "          Mesothelium (PRG4+) Monocytes Neural Pericyte Proximal progenitor  \\\n",
       "Gene                                                                          \n",
       "APOE                                       33%                          33%   \n",
       "AQP1                                                                          \n",
       "BCAN                                        1%                                \n",
       "BCL2                                                                          \n",
       "C5orf30                                                                       \n",
       "CCL2                                                                          \n",
       "CD44                                46%                                       \n",
       "CDCA7                                                                         \n",
       "COL1A1                                     69%                                \n",
       "COL1A2                                     77%                                \n",
       "COL4A1                                                                        \n",
       "COL4A2                                                                        \n",
       "ETS1                                                                          \n",
       "GPX3                                                                          \n",
       "H2AFX                                                                         \n",
       "HMGB2                                                                         \n",
       "IGFBP7                                     59%      59%                       \n",
       "LINC00682                                  41%                                \n",
       "MT1E                       5%                                                 \n",
       "NR4A1                                                                         \n",
       "RAMP1                                      31%                                \n",
       "RNASE1                                                                        \n",
       "S100B                                      78%                                \n",
       "SELENOM                                                                       \n",
       "SKAP2                                                                         \n",
       "SORBS2                                                                        \n",
       "TGFB1                                                                         \n",
       "TOP2A                                                                         \n",
       "TPD52L1                                    39%                                \n",
       "TUBA1B                                                                        \n",
       "UNC5B                                                                         \n",
       "ZFP36                                                                         \n",
       "\n",
       "          RBC SELL+ CD8 T SMC (PLPP2+) Stem cells Stromal 1 (CCL11+)  \\\n",
       "Gene                                                                   \n",
       "APOE                                                             33%   \n",
       "AQP1                                                                   \n",
       "BCAN                                                                   \n",
       "BCL2                                                                   \n",
       "C5orf30                            12%                                 \n",
       "CCL2                                                                   \n",
       "CD44                                                                   \n",
       "CDCA7                                          0%                      \n",
       "COL1A1                                                                 \n",
       "COL1A2                                                                 \n",
       "COL4A1                                                                 \n",
       "COL4A2                                                                 \n",
       "ETS1                                                                   \n",
       "GPX3                                                                   \n",
       "H2AFX                                                                  \n",
       "HMGB2                                                                  \n",
       "IGFBP7                                                                 \n",
       "LINC00682                                                              \n",
       "MT1E       5%                                                          \n",
       "NR4A1                                                                  \n",
       "RAMP1                                                                  \n",
       "RNASE1                                                                 \n",
       "S100B                 78%                                              \n",
       "SELENOM                                                                \n",
       "SKAP2                                                                  \n",
       "SORBS2                                                                 \n",
       "TGFB1                                                                  \n",
       "TOP2A                                                                  \n",
       "TPD52L1                                                                \n",
       "TUBA1B                                                                 \n",
       "UNC5B                                                                  \n",
       "ZFP36                                                                  \n",
       "\n",
       "          Stromal 3 (C7+) Stromal 4 (MMP1+) TRGV4 gdT TRGV5/7 gdT  \\\n",
       "Gene                                                                \n",
       "APOE                                                                \n",
       "AQP1                                                                \n",
       "BCAN                                                                \n",
       "BCL2                                                                \n",
       "C5orf30                                                             \n",
       "CCL2                                                                \n",
       "CD44                                                                \n",
       "CDCA7                                                               \n",
       "COL1A1                69%                                           \n",
       "COL1A2                                                              \n",
       "COL4A1                                                              \n",
       "COL4A2                                                              \n",
       "ETS1                                                                \n",
       "GPX3                                                                \n",
       "H2AFX                                                               \n",
       "HMGB2                                                         46%   \n",
       "IGFBP7                                                              \n",
       "LINC00682                                                           \n",
       "MT1E                                                                \n",
       "NR4A1                                                               \n",
       "RAMP1                                                               \n",
       "RNASE1                                                              \n",
       "S100B                                                               \n",
       "SELENOM                                 22%                         \n",
       "SKAP2                                                               \n",
       "SORBS2                                                              \n",
       "TGFB1                                                               \n",
       "TOP2A                                                               \n",
       "TPD52L1                                                             \n",
       "TUBA1B                                            90%               \n",
       "UNC5B                                                               \n",
       "ZFP36                                                               \n",
       "\n",
       "          angiogenic pericyte arterial capillary cDC1 cycling EC  \\\n",
       "Gene                                                               \n",
       "APOE                                                               \n",
       "AQP1                                                               \n",
       "BCAN                                                               \n",
       "BCL2                                                               \n",
       "C5orf30                                                            \n",
       "CCL2                                                               \n",
       "CD44                                                               \n",
       "CDCA7                                                              \n",
       "COL1A1                                                             \n",
       "COL1A2                                                             \n",
       "COL4A1                    40%                                      \n",
       "COL4A2                    47%                                      \n",
       "ETS1                                                               \n",
       "GPX3                                                               \n",
       "H2AFX                                                         4%   \n",
       "HMGB2                                                        46%   \n",
       "IGFBP7                                                             \n",
       "LINC00682                                                          \n",
       "MT1E                                                               \n",
       "NR4A1                                                              \n",
       "RAMP1                                                              \n",
       "RNASE1                                                             \n",
       "S100B                                             78%              \n",
       "SELENOM                                                            \n",
       "SKAP2                                                              \n",
       "SORBS2                                       50%                   \n",
       "TGFB1                                                              \n",
       "TOP2A                                                              \n",
       "TPD52L1                                                            \n",
       "TUBA1B                                                       90%   \n",
       "UNC5B                                                              \n",
       "ZFP36                                                              \n",
       "\n",
       "          cycling ENCC/glia cycling neuroblast cycling stromal  \\\n",
       "Gene                                                             \n",
       "APOE                                                             \n",
       "AQP1                                                             \n",
       "BCAN                                                             \n",
       "BCL2                                                             \n",
       "C5orf30                 12%                                      \n",
       "CCL2                                                             \n",
       "CD44                                                             \n",
       "CDCA7                    0%                                      \n",
       "COL1A1                                                           \n",
       "COL1A2                                                           \n",
       "COL4A1                                                           \n",
       "COL4A2                                                           \n",
       "ETS1                                                             \n",
       "GPX3                                                             \n",
       "H2AFX                                       4%              4%   \n",
       "HMGB2                                      46%             46%   \n",
       "IGFBP7                                                           \n",
       "LINC00682                                                        \n",
       "MT1E                                                             \n",
       "NR4A1                                                            \n",
       "RAMP1                                                            \n",
       "RNASE1                                                           \n",
       "S100B                                                            \n",
       "SELENOM                                                          \n",
       "SKAP2                                                            \n",
       "SORBS2                                                           \n",
       "TGFB1                                                            \n",
       "TOP2A                                       0%              0%   \n",
       "TPD52L1                                                          \n",
       "TUBA1B                                     90%             90%   \n",
       "UNC5B                                                            \n",
       "ZFP36                                                            \n",
       "\n",
       "          mLN Stroma (FMO2+)  \n",
       "Gene                          \n",
       "APOE                          \n",
       "AQP1                          \n",
       "BCAN                          \n",
       "BCL2                          \n",
       "C5orf30                       \n",
       "CCL2                          \n",
       "CD44                          \n",
       "CDCA7                         \n",
       "COL1A1                        \n",
       "COL1A2                        \n",
       "COL4A1                        \n",
       "COL4A2                        \n",
       "ETS1                          \n",
       "GPX3                     61%  \n",
       "H2AFX                         \n",
       "HMGB2                         \n",
       "IGFBP7                        \n",
       "LINC00682                     \n",
       "MT1E                          \n",
       "NR4A1                    16%  \n",
       "RAMP1                         \n",
       "RNASE1                        \n",
       "S100B                         \n",
       "SELENOM                       \n",
       "SKAP2                         \n",
       "SORBS2                        \n",
       "TGFB1                         \n",
       "TOP2A                         \n",
       "TPD52L1                       \n",
       "TUBA1B                        \n",
       "UNC5B                         \n",
       "ZFP36                         "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key_cluster = \"31\"\n",
    "_, percs_exp, n_exp, genes, msg = self.print_markers(\n",
    "    key_cluster, assign, col_cell_type=c_t, lfc_threshold=None,\n",
    "    count_threshold=1, p_threshold=1, n_top_genes=list(assign[\n",
    "        assign.Bucket == \"Neural-Glia\"].index.intersection(self.rna.var_names)))\n",
    "print(\"\\n\\n\", genes, \"\\n\\n\", n_exp.applymap(int), \"\\n\\n\")\n",
    "print(msg)\n",
    "print(percs_exp.stack().replace(\"\", np.nan).dropna().rename_axis([\n",
    "    \"Gene\", \"Annotation\"]).groupby(\"Annotation\").apply(\n",
    "        lambda x: x.sort_values(ascending=False).apply(\n",
    "            lambda y: f\"{int(y)}%\")).reset_index(2, drop=True))\n",
    "percs_exp.applymap(lambda x: x if x == \"\" else str(int(x)) + \"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sensitivity: % of Total Cells (Restricted to Comparison)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# comp = [\"0\", \"5\", \"8\", \"10\", \"18\", \"20\", \"34\"]\n",
    "comp = [\"2\", \"7\", \"11\", \"21\", \"28\"]\n",
    "count_threshold = 1\n",
    "key_cluster = \"2\"\n",
    "_, percs_exp, n_exp, genes, msg = self.print_markers(\n",
    "    key_cluster, assign, col_cell_type=c_t, lfc_threshold=None,\n",
    "    key_compare=comp, count_threshold=count_threshold, p_threshold=1,\n",
    "    n_top_genes=list(assign[assign.Lump == \"Epithelial\"].index.intersection(\n",
    "        self.rna.var_names)) + [\"TP53\", \"PLAUR\"])\n",
    "print(\"\\n\\n\", genes, \"\\n\\n\", n_exp.applymap(int), \"\\n\\n\")\n",
    "print(msg)\n",
    "percs_exp.applymap(lambda x: x if x == \"\" else str(int(x)) + \"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyze"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Centrality Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "for s in selves:\n",
    "    s.calculate_centrality(n_jobs=sc.settings.n_jobs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Neighborhood Enrichment Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "for s in selves:\n",
    "    _ = s.calculate_neighborhood(figsize=(60, 30))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell Type Co-Occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "for s in selves:\n",
    "    _ = s.find_cooccurrence(figsize=(60, 20), kws_plot=dict(wspace=3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spatial Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in selves:\n",
    "    cct = f\"leiden_spatial_{list(kws_clustering.keys())[-1]}\"\n",
    "    _ = s.cluster_spatial(key_added=cct,\n",
    "                          **kws_clustering[list(kws_clustering.keys())[-1]])\n",
    "    _ = s.find_markers(col_cell_type=cct, kws_plot=False)\n",
    "    _ = s.annotate_clusters(assign[[col_assignment[-1]]], col_cell_type=cct,\n",
    "                            col_annotation=f\"annotation_{cct}\")\n",
    "    for c in [cct, f\"annotation_{cct}\"]:\n",
    "        s.plot_spatial(c)\n",
    "        if out_dir is not None:\n",
    "            s.write_clusters(out_dir, col_cell_type=c,\n",
    "                             n_top=True, overwrite=True,\n",
    "                             file_prefix=f\"{s._library_id}___\")\n",
    "    if out_dir is not None:\n",
    "        s.write(str(s.rna.obs.out_file.iloc[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spatially-Variable Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "kws = dict(kws_plot=dict(legend_fontsize=\"large\"), figsize=(15, 15))\n",
    "for s in selves:\n",
    "    _ = s.find_svgs(genes=15, method=\"moran\", n_perms=10, **kws)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Receptor-Ligand Interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "for s in selves:\n",
    "    kss, ktt = None, None\n",
    "    _ = s.calculate_receptor_ligand(\n",
    "        col_condition=False, p_threshold=0.01, remove_ns=True,\n",
    "        figsize=(30, 20), top_n=25, key_sources=kss, key_targets=ktt)\n",
    "    # s.calculate_receptor_ligand_spatial()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell Type Composition "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in selves:\n",
    "    s.run_composition_analysis()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GEX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in selves:\n",
    "    s.plot_spatial(color=[\"TNF\", \"IL23\", col_cell_type])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = self.calculate_spatial_distance(\"LTi-like NCR+ ILC3\", genes=\"CSF2RB\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers[\"counts\"]\n",
    "sq.tl.var_by_distance(\n",
    "    adata, \"LTi-like NCR+ ILC3\", col_cell_type, library_key=col_sample_id,\n",
    "    design_matrix_key=\"design_matrix\", covariates=[\"Patient\", \"Inflamed\"],\n",
    "    metric=\"euclidean\", spatial_key=self._spatial_key, copy=False)\n",
    "sq.pl.var_by_distance(\n",
    "    adata=adata, var=[\"CSF2RB\", \"CSF2RA\", \"IL7R\"],\n",
    "    anchor_key=\"LTi-like NCR+ ILC3\",\n",
    "    covariate=\"Inflamed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workspace"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manual Annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Uninflamed-50452A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "s.get_layer(\"counts\", inplace=True)\n",
    "print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"9\"][:, \"CD79A\"].X > 0))\n",
    "print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"9\"][:, \"CD79B\"].X > 0))\n",
    "print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"6\"][:, \"CD79A\"].X > 0))\n",
    "print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"6\"][:, \"CD79B\"].X > 0))\n",
    "print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"6\"][\n",
    "    :, \"ADAMDEC1\"].X > 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in [\"COL1A1\", \"PEG10\", \"RCN3\"]:\n",
    "    print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"10\"][:, x].X > 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in [\"BMP5\", \"F3\", \"MMP11\", \"CCL8\", \"NPY\", \"CH25H\"]:\n",
    "    print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"15\"][:, x].X > 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in [\"LGR5\"]:\n",
    "    print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"16\"][:, x].X > 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in [\"CRIP2\", \"KCNA5\", \"LBH\", \"NET1\", \"PLN\", \"RERGL\"]:\n",
    "    print(round(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"20\"][:, x].X > 0) * 100, 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in [\"TPSAB1\", \"CMA1\"]:\n",
    "    print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"21\"][:, x].X > 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in [\"GNLY\", \"GZMB\"]:\n",
    "    print(np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] != \"23\"][:, x].X > 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in marker_genes_dict[\"Pericyte\"]:\n",
    "    print(int(100 * np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == \"25\"][\n",
    "        :, x].X > 0)), \" vs. \", int(100 * np.mean(s.rna[s.rna.obs[\n",
    "            f\"leiden_{r}\"] != \"25\"][:, x].X > 0)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comp = \"2\"\n",
    "print(\", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "    f\"leiden_{r}\"] == \"11\"][:, x].X > 0))) + \" vs. \" + str(int(\n",
    "        100 * np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] == comp][\n",
    "            :, x].X > 0))) + \"%\" for x in [\"COL4A1\", \"COL4A2\", \"DLC1\"]]))\n",
    "print(f\"\\n...compared to {comp}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "    f\"leiden_{r}\"] == \"9\"][:, x].X > 0))) + \" vs. \" + str(int(\n",
    "        100 * np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"] != \"9\"][\n",
    "            :, x].X > 0))) + \"%\" for x in [\n",
    "                \"ZBTB38\", \"IFNAR2\", \"SELENOM\", \"XBP1\"]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Stricture-50452C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Memory B\n",
    "# Stricture-50452C___leiden_res0.75_dist0.3_npc30\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res0.75_dist0.3_npc30\"\n",
    "s.rna.obs.loc[:, f\"leiden_{r}\"] = s.rna.obs[\n",
    "    f\"leiden_{r}\"].astype(float).astype(int).astype(str)\n",
    "clid = \"21\"\n",
    "comp = [\"2\", \"16\", \"18\"]\n",
    "print(\", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "    f\"leiden_{r}\"] == clid][:, x].X > 0))) + \" vs. \" + str(int(\n",
    "        100 * np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"].isin(comp)][\n",
    "            :, x].X > 0))) + \"%\" for x in [\n",
    "                \"GPR183\", \"HHEX\", \"INPP5D\", \"LRRK2\"]]))\n",
    "print(f\"\\n...compared to {comp}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Naive B\n",
    "# Stricture-50452C___leiden_res0.75_dist0.3_npc30\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res0.75_dist0.3_npc30\"\n",
    "s.rna.obs.loc[:, f\"leiden_{r}\"] = s.rna.obs[\n",
    "    f\"leiden_{r}\"].astype(float).astype(int).astype(str)\n",
    "clid = \"21\"\n",
    "comp = [\"2\", \"16\", \"18\"]\n",
    "print(\", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "    f\"leiden_{r}\"] == clid][:, x].X > 0))) + \" vs. \" + str(int(\n",
    "        100 * np.mean(s.rna[s.rna.obs[f\"leiden_{r}\"].isin(comp)][\n",
    "            :, x].X > 0))) + \"%\" for x in [\n",
    "                \"IRF1\", \"LNPEP\", \"PLAC8\", \"IFITM1\"]]))\n",
    "print(f\"\\n...compared to {comp}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Uninflamed-50006B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Neural-Glia\n",
    "# Uninflamed-50006B\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"24\"\n",
    "comp = [\"19\"]\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.Bin == \"Neural\"].index.intersection(\n",
    "                s.rna.var_names).difference(assign[\n",
    "                    assign.Bin == \"Glia\"].index)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Epithelial\n",
    "# Uninflamed-50006B\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"26\"\n",
    "# comp = [\"10\", \"28\"]\n",
    "# comp = [\"1\", \"3\", \"4\", \"7\", \"8\", \"10\", \"20\", \"26\"]\n",
    "# comp = [\"3\", \"4\", \"7\", \"8\", \"10\", \"26\"]\n",
    "comp = [\"4\", \"7\", \"8\", \"9\"]  # suspected goblets\n",
    "# comp = [\"8\"]\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in [\n",
    "            \"MUC2\", \"MUC4\", \"MUC1\", \"KLK1\", \"ZG16\",  # \"FAM3D\",\n",
    "            \"BEST2\", \"BEST4\",\"LGR5\", \"CDCA7\", \"CCL20\", \"LYZ\",\n",
    "            \"HCK\", \"MATK\", \"ZFHX3\"]])  # Tuft markers\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Eptithelial\n",
    "# Uninflamed-50006B\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"26\"\n",
    "comp = [\"10\", \"28\"]\n",
    "# comp = [\"1\", \"3\", \"4\", \"7\", \"8\", \"10\", \"20\", \"26\"]\n",
    "# comp = [\"3\", \"4\", \"7\", \"8\", \"10\", \"26\"]\n",
    "# comp = [\"4\", \"7\", \"8\", \"9\"]  # suspected goblets\n",
    "# comp = [\"8\"]\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in [\n",
    "            \"MUC5B\", \"KIT\", \"MARCKSL1\", \"NACA\", \"FAM3D\", \"LCN2\", \"IRF2BP2\",\n",
    "            \"MUC4\", \"SATB2\", \"UQCRC1\", \"SLC39A8\", \"SERBP1\", \"TP53\", \"MYO10\",\n",
    "            \"KLK1\", \"HMGB1\", \"HOXA3\", \"NR5A2\", \"CAT\", \"LIMA1\", \"COX8A\",\n",
    "            \"LGALS9\", \"HMGB2\", \"CDCA7\", \"ZFP91\", \"MUC1\", \"ATF6B\"]])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ggc = [\"HCK\", \"MATK\", \"ZFHX3\"]\n",
    "subs = s.rna[s.rna.obs[r] == clus].copy()\n",
    "subs = sc.tl.score_genes(subs, ggc, score_name=\"|\".join(ggc),\n",
    "                         use_raw=False, copy=True)  # score co-GEX\n",
    "np.mean(subs.obs[\"|\".join(ggc)] > 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Stricture-50006C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# T Cell\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"22\"\n",
    "comp = list(s.rna.obs[r].unique())\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.Bin == \"T Cell\"].index.intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# gDT\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"27\"\n",
    "comp = list(s.rna.obs[r].unique())\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.group.apply(lambda x: \"gdT\" in x)\n",
    "            ].index.intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# LEC\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"27\"\n",
    "comp = list(s.rna.obs[r].unique())\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.group.apply(lambda x: \"LEC\" in x)\n",
    "            ].index.intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# cDC\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"25\"\n",
    "# comp = list(s.rna.obs[r].unique())\n",
    "comp = [\"6\"]\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.group == \"cDC1\"].index.intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"21\"\n",
    "comp = list(s.rna.obs[r].unique())\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.Bin == \"Neural\"].index.intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"11\"\n",
    "comp = list(s.rna.obs[r].unique())\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in [\"LYVE1\"]])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"27\"\n",
    "comp = list(s.rna.obs[r].unique())\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.Bin == \"Endothelial\"].index.intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "i = -1\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"31\"\n",
    "comp = [\"4\"]\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.Bin == \"Glia\"].index.intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Uninflamed-50336C\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"9\"\n",
    "# comp = [\"4\"]\n",
    "comp = list(s.rna.obs[r].unique())\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.group == \"LYVE1+ Macrophage\"].index.intersection(\n",
    "                s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "clus = \"0\"\n",
    "# comp = [\"4\"]\n",
    "comp = list(s.rna.obs[r].unique())\n",
    "comp = list(set(comp).difference(set([clus])))\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[\n",
    "        r].isin(comp)][:, x].X > 0))) + \"%\" for x in assign[\n",
    "            assign.Bucket == \"B Cell\"].index.intersection(\n",
    "                s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# Stem/TA\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "ggx = assign[assign[\"group\"].isin([\"Stem cells\",\n",
    "                                   \"Proximal progenitor\"])].index\n",
    "# ggx = [\"LGR5\"]\n",
    "clus = \"13\"\n",
    "comp = [\"2\"]\n",
    "# comp = list(s.rna.obs[r].unique())\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[r].isin(\n",
    "        list(set(comp).difference(set([clus]))))][:, x].X > 0))) + \"%\"\n",
    "    for x in set(ggx).intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# Goblet\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "ggx = assign[assign[\"Bin\"].isin([\"Goblet\"])].index\n",
    "# ggx = [\"LGR5\"]\n",
    "clus = \"14\"\n",
    "comp = [\"3\"]\n",
    "# comp = list(s.rna.obs[r].unique())\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[r].isin(\n",
    "        list(set(comp).difference(set([clus]))))][:, x].X > 0))) + \"%\"\n",
    "    for x in set(ggx).intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# Stem/Paneth\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "ggx = [\"LYZ\", \"LCN2\", \"RAMP1\"]\n",
    "clus = \"19\"\n",
    "comp = [\"3\"]\n",
    "# comp = list(s.rna.obs[r].unique())\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[r].isin(\n",
    "        list(set(comp).difference(set([clus]))))][:, x].X > 0))) + \"%\"\n",
    "    for x in set(ggx).intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# Stem/Paneth\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "ggx = assign[assign[\"group\"].isin([\"Mast Cell\"])].index\n",
    "clus = \"24\"\n",
    "comp = list(s.rna.obs[r].unique())\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[r].isin(\n",
    "        list(set(comp).difference(set([clus]))))][:, x].X > 0))) + \"%\"\n",
    "    for x in set(ggx).intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# MF/DP\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "ggx = assign[assign[\"Bin\"].isin([\"Myofibroblast\"])].index\n",
    "clus = \"31\"\n",
    "# comp = list(s.rna.obs[r].unique())\n",
    "# comp = [\"1\", \"6\"]\n",
    "comp = [\"17\", \"22\"]\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[r].isin(\n",
    "        list(set(comp).difference(set([clus]))))][:, x].X > 0))) + \"%\"\n",
    "    for x in set(ggx).intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stricture-50006C\n",
    "# SMC/MF/DP\n",
    "i = 0\n",
    "s = selves[i]\n",
    "r = \"leiden_res1pt5_dist0_npc30\"\n",
    "ggx = assign[assign[\"Bin\"].isin([\"SMC\"])].index\n",
    "clus = \"1\"\n",
    "comp = [\"6\"]\n",
    "ccc = \", \".join([x + \": \" + str(int(100 * np.mean(s.rna[s.rna.obs[r] == clus][\n",
    "    :, x].X > 0))) + \" vs. \" + str(int(100 * np.mean(s.rna[s.rna.obs[r].isin(\n",
    "        list(set(comp).difference(set([clus]))))][:, x].X > 0))) + \"%\"\n",
    "    for x in set(ggx).intersection(s.rna.var_names)])\n",
    "f\"{clus} compared to {comp}: {ccc}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CD34"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "gene = \"CD34\"\n",
    "thr = np.arange(1, 15)  # thresholds\n",
    "fig, axes = plt.subplots(1, len(selves))\n",
    "for i, s in enumerate(selves):\n",
    "    print(f\"{s._library_id}: {np.sum(s.rna[:, gene].X > 0)} / {s.rna.n_obs}\")\n",
    "    a_x = axes[i].bar(thr, [np.sum(s.rna[:, gene].X >= i) for i in thr])\n",
    "    axes[i].set_title(s._library_id)\n",
    "    # for x in s.rna.obs[f\"leiden_{r}\"].unique():\n",
    "    #     print(str(x) + \": \" + np.mean(s.rna[s.rna.obs[\n",
    "    #         f\"leiden_{r}\"] == x][:, \"CD34\"].X > 0))\n",
    "fig.suptitle(\"Total Cells above Count Thresholds for CD34\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CSF1/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in [\"CSF1\", \"CSF2\"]:\n",
    "    print(f\"\\n\\n{'=' * 80}\\n{g}\\n{'=' * 80}\")\n",
    "    for i, s in enumerate(selves):\n",
    "        print(f\"{s._library_id}: {np.sum(s.rna[:, g].X > 0)} / {s.rna.n_obs}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aligned Images"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "landmarks = ShapesModel.parse(\n",
    "    np.array([[10556.699, 7829.764], [13959.155, 13522.025], [10621.200, 17392.116]]), geometry=0, radius=500\n",
    ")\n",
    "visium_sdata[\"visium_landmarks\"] = visium_landmarks\n",
    "\n",
    "xenium_landmarks = ShapesModel.parse(\n",
    "    np.array([[9438.385, 13933.017], [24847.866, 5948.002], [34082.584, 15234.235]]), geometry=0, radius=500\n",
    ")\n",
    "xenium_sdata[\"xenium_landmarks\"] = xenium_landmarks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import spatialdata_io as sdio\n",
    "from spatialdata.models import ShapesModel\n",
    "\n",
    "fai = os.path.join(\n",
    "    \"~/corescpy/examples/data/spatial\",\n",
    "    \"Stricture-50452C__P53Red_P16Green_DAPIBlue.ome.tif\")\n",
    "faa = os.path.join(\n",
    "    \"~/corescpy/examples/data/spatial\",\n",
    "    \"Stricture-50452C__P53Red_P16Green_DAPIBlue_matrix.csv\")\n",
    "\n",
    "img = sdio.xenium_aligned_image(fai, faa)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py-bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
