{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports and Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #808000; text-decoration-color: #808000; font-weight: bold\">ryp2 is not installed. Install with </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">pip install rpy2 </span><span style=\"color: #808000; text-decoration-color: #808000; font-weight: bold\">to run tools with R support.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;33mryp2 is not installed. Install with \u001b[0m\u001b[1;32mpip install rpy2 \u001b[0m\u001b[1;33mto run tools with R support.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading data from `https://omnipathdb.org/queries/enzsub?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/interactions?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/complexes?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/annotations?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/intercell?format=json`\n",
      "Downloading data from `https://omnipathdb.org/about?format=text`\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import re\n",
    "import json\n",
    "from datetime import datetime\n",
    "import functools\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sb\n",
    "from anndata import AnnData\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import corescpy as cr\n",
    "\n",
    "pd.options.display.max_columns = 100\n",
    "pd.options.display.max_rows = 200\n",
    "sc.settings.set_figure_params(dpi=100, frameon=False, figsize=(20, 20))\n",
    "\n",
    "palette = \"tab20\"\n",
    "\n",
    "\n",
    "def construct_file(sample, slide, date=None, timestamp=None,\n",
    "                   panel_id=\"TUQ97N\", prefix=\"output-XETG00189\",\n",
    "                   project_owner=\"EA\", run=\"CHO-001\", directory=None):\n",
    "    \"\"\"Construct file path from information.\"\"\"\n",
    "    if isinstance(sample, str):\n",
    "        sample = [sample]\n",
    "    if \"outputs\" not in directory and os.path.exists(\n",
    "            os.path.join(directory, \"outputs\")):\n",
    "        directory = os.path.join(directory, \"outputs\")\n",
    "    print(directory)\n",
    "    panel_id, prefix, project_owner, slide, date, timestamp = [\n",
    "        [x] * len(sample) if isinstance(x, str) else list(x) if x else x\n",
    "        for x in [panel_id, prefix, project_owner, slide, date, timestamp]]\n",
    "    run = [run] * len(sample) if isinstance(run, (str, int, float)) else run\n",
    "    block = [\"-\".join(i) for i in zip(sample, panel_id, project_owner)]\n",
    "    fff = [f\"{prefix[i]}__{slide[i]}__{block[i]}\" for i in range(len(sample))]\n",
    "    if date is None or timestamp is None:\n",
    "        for i, x in enumerate(fff):  # iterate current file stems\n",
    "            ddd = os.path.join(directory, panel_id[i], run[i])\n",
    "            print(ddd)\n",
    "            matches = sum([x in d for d in os.listdir(ddd)])\n",
    "            if  matches != 1:\n",
    "                raise ValueError(f\"{x} found in 0 or multiple file paths\",\n",
    "                                 f\"\\n\\n{os.listdir(ddd)}\")\n",
    "            fff[i] = os.path.join(ddd, np.array(os.listdir(ddd))[np.where([\n",
    "                x in d for d in os.listdir(ddd)])[0][0]])  # find match\n",
    "    else:\n",
    "        fff = [os.path.join(directory, panel_id[i], run[i],\n",
    "                            f\"{x}__{date[i]}__{timestamp[i]}\")\n",
    "               for i, x in enumerate(fff)]\n",
    "    return fff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs\n",
      "/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/CHO-001\n",
      "/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/CHO-001\n",
      "/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/CHO-001\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Status</th>\n",
       "      <th>Project</th>\n",
       "      <th>Sample Location</th>\n",
       "      <th>Inflamed</th>\n",
       "      <th>Slide</th>\n",
       "      <th>GRID ID</th>\n",
       "      <th>Procedure Date</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Hispanic</th>\n",
       "      <th>Diagnosis</th>\n",
       "      <th>Project.1</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Disease Status</th>\n",
       "      <th>Date Collected</th>\n",
       "      <th>Date Sectioned</th>\n",
       "      <th>Date of Hybridization</th>\n",
       "      <th>Storage in 4C</th>\n",
       "      <th>Created By</th>\n",
       "      <th>Created</th>\n",
       "      <th>Storage Status</th>\n",
       "      <th>Sample Location.1</th>\n",
       "      <th>Storage Row</th>\n",
       "      <th>Storage Col</th>\n",
       "      <th>Checked Out By</th>\n",
       "      <th>Stricture</th>\n",
       "      <th>Condition</th>\n",
       "      <th>file_path</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Uninflamed-50452A</th>\n",
       "      <td>50452A</td>\n",
       "      <td>50452</td>\n",
       "      <td>Available</td>\n",
       "      <td>scRNA &amp; Spatial</td>\n",
       "      <td>Terminal Ileum</td>\n",
       "      <td>uninflamed</td>\n",
       "      <td>0010700</td>\n",
       "      <td>1011335</td>\n",
       "      <td>2023-12-04</td>\n",
       "      <td>69</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>NO</td>\n",
       "      <td>Crohn's Disease</td>\n",
       "      <td>Helmsley-Senescence</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2024-01-22 15:25:38</td>\n",
       "      <td>2024-01-23 15:25:58</td>\n",
       "      <td>YES</td>\n",
       "      <td>ksenija sabic</td>\n",
       "      <td>2024-02-06 15:26:47.887</td>\n",
       "      <td>Not in storage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>Uninflamed</td>\n",
       "      <td>/mnt/cho_lab/disk2/elizabeth/data/shared-xeniu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inflamed-50452B</th>\n",
       "      <td>50452B</td>\n",
       "      <td>50452</td>\n",
       "      <td>Available</td>\n",
       "      <td>scRNA &amp; Spatial</td>\n",
       "      <td>Terminal Ileum</td>\n",
       "      <td>inflamed</td>\n",
       "      <td>0010700</td>\n",
       "      <td>1011335</td>\n",
       "      <td>2023-12-04</td>\n",
       "      <td>69</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>NO</td>\n",
       "      <td>Crohn's Disease</td>\n",
       "      <td>Helmsley-Senescence</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2024-01-22 00:00:00</td>\n",
       "      <td>2024-01-23 00:00:00</td>\n",
       "      <td>YES</td>\n",
       "      <td>ksenija sabic</td>\n",
       "      <td>2024-02-06 15:26:47.887</td>\n",
       "      <td>Not in storage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>Inflamed</td>\n",
       "      <td>/mnt/cho_lab/disk2/elizabeth/data/shared-xeniu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stricture-50452C</th>\n",
       "      <td>50452C</td>\n",
       "      <td>50452</td>\n",
       "      <td>Available</td>\n",
       "      <td>scRNA &amp; Spatial</td>\n",
       "      <td>Terminal Ileum (stricture)</td>\n",
       "      <td>inflamed</td>\n",
       "      <td>0010663</td>\n",
       "      <td>1011335</td>\n",
       "      <td>2023-12-04</td>\n",
       "      <td>69</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>NO</td>\n",
       "      <td>Crohn's Disease</td>\n",
       "      <td>Helmsley-Senescence</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2024-01-22 00:00:00</td>\n",
       "      <td>2024-01-23 00:00:00</td>\n",
       "      <td>YES</td>\n",
       "      <td>ksenija sabic</td>\n",
       "      <td>2024-02-06 15:26:47.887</td>\n",
       "      <td>Not in storage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stricture</td>\n",
       "      <td>Stricture</td>\n",
       "      <td>/mnt/cho_lab/disk2/elizabeth/data/shared-xeniu...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Sample ID  Patient     Status          Project  \\\n",
       "Sample                                                             \n",
       "Uninflamed-50452A    50452A    50452  Available  scRNA & Spatial   \n",
       "Inflamed-50452B      50452B    50452  Available  scRNA & Spatial   \n",
       "Stricture-50452C     50452C    50452  Available  scRNA & Spatial   \n",
       "\n",
       "                              Sample Location    Inflamed    Slide  GRID ID  \\\n",
       "Sample                                                                        \n",
       "Uninflamed-50452A              Terminal Ileum  uninflamed  0010700  1011335   \n",
       "Inflamed-50452B                Terminal Ileum    inflamed  0010700  1011335   \n",
       "Stricture-50452C   Terminal Ileum (stricture)    inflamed  0010663  1011335   \n",
       "\n",
       "                  Procedure Date  Age     Sex   Race Hispanic  \\\n",
       "Sample                                                          \n",
       "Uninflamed-50452A     2023-12-04   69  female  white       NO   \n",
       "Inflamed-50452B       2023-12-04   69  female  white       NO   \n",
       "Stricture-50452C      2023-12-04   69  female  white       NO   \n",
       "\n",
       "                         Diagnosis            Project.1  Procedure  \\\n",
       "Sample                                                               \n",
       "Uninflamed-50452A  Crohn's Disease  Helmsley-Senescence  Resection   \n",
       "Inflamed-50452B    Crohn's Disease  Helmsley-Senescence  Resection   \n",
       "Stricture-50452C   Crohn's Disease  Helmsley-Senescence  Resection   \n",
       "\n",
       "                  Disease Status  Date Collected      Date Sectioned  \\\n",
       "Sample                                                                 \n",
       "Uninflamed-50452A         Active             NaN 2024-01-22 15:25:38   \n",
       "Inflamed-50452B           Active             NaN 2024-01-22 00:00:00   \n",
       "Stricture-50452C          Active             NaN 2024-01-22 00:00:00   \n",
       "\n",
       "                  Date of Hybridization Storage in 4C     Created By  \\\n",
       "Sample                                                                 \n",
       "Uninflamed-50452A   2024-01-23 15:25:58           YES  ksenija sabic   \n",
       "Inflamed-50452B     2024-01-23 00:00:00           YES  ksenija sabic   \n",
       "Stricture-50452C    2024-01-23 00:00:00           YES  ksenija sabic   \n",
       "\n",
       "                                  Created  Storage Status  Sample Location.1  \\\n",
       "Sample                                                                         \n",
       "Uninflamed-50452A 2024-02-06 15:26:47.887  Not in storage                NaN   \n",
       "Inflamed-50452B   2024-02-06 15:26:47.887  Not in storage                NaN   \n",
       "Stricture-50452C  2024-02-06 15:26:47.887  Not in storage                NaN   \n",
       "\n",
       "                   Storage Row  Storage Col  Checked Out By  Stricture  \\\n",
       "Sample                                                                   \n",
       "Uninflamed-50452A          NaN          NaN             NaN       None   \n",
       "Inflamed-50452B            NaN          NaN             NaN       None   \n",
       "Stricture-50452C           NaN          NaN             NaN  Stricture   \n",
       "\n",
       "                    Condition  \\\n",
       "Sample                          \n",
       "Uninflamed-50452A  Uninflamed   \n",
       "Inflamed-50452B      Inflamed   \n",
       "Stricture-50452C    Stricture   \n",
       "\n",
       "                                                           file_path  \n",
       "Sample                                                                \n",
       "Uninflamed-50452A  /mnt/cho_lab/disk2/elizabeth/data/shared-xeniu...  \n",
       "Inflamed-50452B    /mnt/cho_lab/disk2/elizabeth/data/shared-xeniu...  \n",
       "Stricture-50452C   /mnt/cho_lab/disk2/elizabeth/data/shared-xeniu...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Column Names\n",
    "col_sample_id_original = \"Sample ID\"\n",
    "col_sample_id = \"Sample\"\n",
    "col_subject = \"Patient\"\n",
    "col_batch = \"Slide\"\n",
    "col_path = \"file_path\"\n",
    "# col_date_time, col_date = \"Date Sectioned\", \"Date\"\n",
    "col_date_time, col_date = None, None\n",
    "col_inflamed, col_stricture = \"Inflamed\", \"Stricture\"\n",
    "col_condition = \"Condition\"\n",
    "meta_rn = {\"Name\": col_subject, \"Slide ID\": col_batch,\n",
    "           \"Inflammation Status\": col_inflamed}\n",
    "\n",
    "# Directories & Metadata\n",
    "# Replace manually or mirror my file/directory tree in your home\n",
    "include_stricture = True\n",
    "run = \"CHO-001\"\n",
    "samples = [\"50452A\", \"50452B\", \"50452C\"]\n",
    "# samples = [\"50452A\", \"50452B\"]\n",
    "# samples = [\"50452A\", \"50452B\", \"50452C\", \"50564A4\", \"50618B5\"]\n",
    "# run = [\"CHO-001\"] * 3 + [\"CHO-002\"] * 3\n",
    "panel_id = \"TUQ97N\"\n",
    "prefix = \"output-XETG00189\"\n",
    "project_owner = \"EA\"\n",
    "ddu = os.path.expanduser(\"~\")\n",
    "ddl = \"/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library\"\n",
    "ddd = os.path.join(ddl, \"outputs\", panel_id)\n",
    "panel = os.path.join(ddu, \"projects/senescence/ProposedGenePanel.xlsx\")\n",
    "file_ann = os.path.join(ddu, \"corescpy/examples/annotation_guide.xlsx\")\n",
    "\n",
    "# Input/Output Options\n",
    "reload = True\n",
    "out_dir = os.path.join(ddd, \"nebraska\")  # set to None to avoid saving\n",
    "if not os.path.exists(out_dir):\n",
    "    os.makedirs(out_dir)\n",
    "\n",
    "# Computing Resources\n",
    "gpu = False\n",
    "sc.settings.n_jobs = 4\n",
    "sc.settings.max_memory = 150\n",
    "\n",
    "# Read Metadata & Other Information\n",
    "annot_df = pd.read_excel(file_ann)\n",
    "metadata = pd.read_excel(os.path.join(ddl, \"Xenium_Samples_02092024.xlsx\"),\n",
    "                         dtype={\"Slide ID\": str})\n",
    "metadata = metadata.rename(meta_rn, axis=1)\n",
    "if samples not in [\"all\", None]:  # subset by sample ID?\n",
    "    metadata = metadata.set_index(col_sample_id_original).loc[\n",
    "        samples].reset_index()\n",
    "\n",
    "# Processing & Clustering Options\n",
    "resolution = 0.5\n",
    "min_dist = 1\n",
    "n_comps = 20\n",
    "# col_qscore = ?\n",
    "# custom_thresholds = {col_qscore: [, None]}\n",
    "custom_thresholds = None\n",
    "genes_subset = list(annot_df.iloc[:, 0])\n",
    "kws_pp = dict(cell_filter_pmt=None, cell_filter_ncounts=[50, None],\n",
    "              cell_filter_ngene=[30, None], gene_filter_ncell=[3, None],\n",
    "              gene_filter_ncounts=[3, None],\n",
    "              custom_thresholds=custom_thresholds,\n",
    "              kws_scale=dict(max_value=10, zero_center=True))\n",
    "kws_umap = dict(min_dist=min_dist, method=\"rapids\") if gpu is True else dict(\n",
    "    min_dist=min_dist)\n",
    "kws_cluster = dict(use_gpu=gpu, kws_umap=kws_umap, kws_neighbors=None,\n",
    "                   use_highly_variable=False, n_comps=n_comps,\n",
    "                   genes_subset=genes_subset, resolution=resolution)\n",
    "# kws_subcluster = dict(use_gpu=gpu, kws_umap=dict(min_dist=0.3),\n",
    "#                       kws_neighbors=None, use_highly_variable=False,\n",
    "#                       n_comps=n_comps, genes_subset=genes_subset,\n",
    "#                       method_cluster=\"leiden\", resolution=0.5)\n",
    "kws_subcluster = dict(method_cluster=\"leiden\", resolution=0.5)\n",
    "\n",
    "# Revise Metadata & Construct Variables from Options\n",
    "if col_stricture not in metadata.columns:\n",
    "    metadata.loc[:, col_stricture] = metadata[\"Sample Location\"].apply(\n",
    "        lambda x: \"Stricture\" if \"stricture\" in x.lower() else \"None\")\n",
    "metadata.loc[:, col_condition] = metadata.apply(\n",
    "    lambda x: \"Stricture\" if \"stricture\" in x[col_stricture].lower() else x[\n",
    "        col_inflamed].capitalize() , axis=1)\n",
    "if col_date_time:\n",
    "    metadata.loc[:, col_date] = metadata[col_date_time].apply(\n",
    "        lambda x: datetime.strftime(x, \"%Y%m%d\"))\n",
    "    dates = list(metadata[col_date])\n",
    "else:\n",
    "    dates = None\n",
    "metadata.loc[:, col_sample_id] = metadata.apply(\n",
    "    lambda x: f\"{x[col_condition]}-{x[col_sample_id_original]}\" , axis=1)\n",
    "metadata = metadata.set_index(col_sample_id)\n",
    "metadata.loc[:, col_path] = construct_file(\n",
    "    list(metadata[col_sample_id_original]), list(metadata[col_batch]),\n",
    "    dates, panel_id=panel_id, prefix=prefix,\n",
    "    project_owner=project_owner, run=run, directory=ddl)\n",
    "col_cell_type = \"Annotation\"\n",
    "file_path_dict = dict(zip(metadata.index.values, metadata[\"file_path\"]))\n",
    "kws_init = dict(col_batch=col_batch, col_subject=col_subject,\n",
    "                col_sample_id=col_sample_id, col_cell_type=col_cell_type)\n",
    "\n",
    "# Annotation File\n",
    "col_assignment = \"Bin\"\n",
    "assign = pd.read_excel(file_ann, index_col=0).dropna(subset=col_assignment)\n",
    "sources = assign[col_assignment].unique()\n",
    "rename = dict(zip(sources, [\" \".join([i.capitalize() if i and i[\n",
    "    0] != \"(\" and not i.isupper() and i not in [\n",
    "        \"IgG\", \"IgA\"] else i for i in x.split(\" \")]) if len(x.split(\n",
    "            \" \")) > 1 else x for x in [re.sub(\"glia\", \"Glia\", re.sub(\n",
    "                \"_\", \" \", j)) for j in sources]]))\n",
    "assign.loc[:, col_assignment] = assign[col_assignment].replace(rename)\n",
    "assign = assign.rename_axis(\"Gene\")\n",
    "marker_genes_dict = dict(assign.reset_index().groupby(col_assignment).apply(\n",
    "    lambda x: list(pd.unique(x.Gene))))  # to marker dictionary\n",
    "\n",
    "# Subset if Desired\n",
    "if include_stricture is False:\n",
    "    metadata = metadata[metadata.Stricture != \"Stricture\"]\n",
    "metadata\n",
    "\n",
    "# Genes\n",
    "# genes = [\"CDKN1A\", \"CDKN2A\", \"TP53\", \"PLAUR\", \"PTGER4\", \"FTL\", \"IL6ST\"]\n",
    "# cell_types = [\"ILC3\", \"LTi-like NCR+ ILC3\", \"LTi-like NCR- ILC3\",\n",
    "#               \"ILCP\", \"Macrophages\", \"Stem cells\"]\n",
    "# palette = [\"r\", \"tab:pink\", \"m\", \"b\", \"tab:brown\", \"cyan\"]\n",
    "# High in inf. vs. un\n",
    "# OSM\n",
    "# IL13\n",
    "# IL1B\n",
    "# IL6\n",
    "# TNF\n",
    "# S100A8\n",
    "# S100A9\n",
    "# ------------------------------\n",
    "# High in stricture vs inf/un\n",
    "# PDGFRA\n",
    "# IL6ST\n",
    "# PTPN1\n",
    "# IFNG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Re-Name Files in Standard Form Using Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fff = list(pd.Series([x if \"h5ad\" in x else np.nan for x in os.listdir(\n",
    "#     out_dir)]).dropna())\n",
    "# frn = []\n",
    "# for x in fff:\n",
    "#     ann = sc.read(os.path.join(out_dir, x))\n",
    "#     kwu = re.sub(\"{\", \"\", re.sub(\"}\", \"\", str(ann.obs.iloc[0].kws_umap))\n",
    "#                  ).split(\", \")\n",
    "#     kwu = np.array(kwu)[np.where([\"min_dist\" in x for x in kwu])[\n",
    "#         0]][0].split(\": \")[1]\n",
    "#     frn += [str(f\"{str(ann.obs.iloc[0][col_sample_id])}__\"\n",
    "#                 f\"res{str(ann.obs.iloc[0].resolution)}\"\n",
    "#                 f\"_dist{kwu}_npc{ann.varm['PCs'].shape[1]}\")]\n",
    "# frd = dict(zip([os.path.join(out_dir, i) for i in fff], [os.path.join(\n",
    "#     out_dir, re.sub(\"[.]\", \"pt\", re.sub(\"[.]h5ad\", \"\", i)) + \".h5ad\")\n",
    "#                                                          for i in frn]))\n",
    "# frd\n",
    "# # for x in frd:\n",
    "# #     os.system(f\"mv {x} {frd[x]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "<<< INITIALIZING SPATIAL CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "<<< INITIALIZING OMICS CLASS OBJECT >>>\n",
      "\n",
      "Unused keyword arguments: {'library_id': 'Uninflamed-50452A', 'visium': False}.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\u001b[34mINFO    \u001b[0m reading                                                                                                   \n",
      "         \u001b[35m/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/CHO-001/output-XETG00189__0010700__\u001b[0m\n",
      "         \u001b[35m50452A-TUQ97N-EA__20240126__205019/\u001b[0m\u001b[95mcell_feature_matrix.h5\u001b[0m                                                 \n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "dsfasfda AnnData object with n_obs × n_vars = 333825 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Cell Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "Observations: 333825\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Gene Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "Empty DataFrame\n",
      "Columns: [index, gene_ids, feature_types, genome]\n",
      "Index: []\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " SpatialData object with:\n",
      "├── Images\n",
      "│     ├── 'morphology_focus': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "│     └── 'morphology_mip': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 10) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (333825, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "└── Table\n",
      "      └── AnnData object with n_obs × n_vars = 333825 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts': AnnData (333825, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), morphology_mip (Images), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "None\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 333825 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts' \n",
      "\n",
      "                      gene_ids    feature_types   genome\n",
      "gene_symbols                                           \n",
      "ABCA7         ENSG00000064687  Gene Expression  Unknown\n",
      "ACTA2         ENSG00000107796  Gene Expression  Unknown\n",
      "ACTG2         ENSG00000163017  Gene Expression  Unknown\n",
      "ADAM15        ENSG00000143537  Gene Expression  Unknown\n",
      "ADAM28        ENSG00000042980  Gene Expression  Unknown \n",
      "\n",
      "       cell_id  transcript_counts  control_probe_counts  \\\n",
      "0  aaaaibki-1                168                     0   \n",
      "1  aaaaiiji-1                166                     0   \n",
      "2  aaaajape-1                135                     0   \n",
      "3  aaaamhbm-1                 99                     0   \n",
      "4  aaabdoak-1                 61                     0   \n",
      "\n",
      "   control_codeword_counts  unassigned_codeword_counts  \\\n",
      "0                        0                           0   \n",
      "1                        0                           0   \n",
      "2                        0                           0   \n",
      "3                        0                           0   \n",
      "4                        0                           0   \n",
      "\n",
      "   deprecated_codeword_counts  total_counts   cell_area  nucleus_area  \\\n",
      "0                           0           168  328.150481     13.908126   \n",
      "1                           0           166  107.607348     18.649532   \n",
      "2                           0           135  110.226410     10.250469   \n",
      "3                           0            99   46.736720     16.482032   \n",
      "4                           0            61   22.216876      6.954063   \n",
      "\n",
      "         region             Sample  \n",
      "0  cell_circles  Uninflamed-50452A  \n",
      "1  cell_circles  Uninflamed-50452A  \n",
      "2  cell_circles  Uninflamed-50452A  \n",
      "3  cell_circles  Uninflamed-50452A  \n",
      "4  cell_circles  Uninflamed-50452A  \n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "col_segment=\"region\"\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\n",
      "\n",
      "<<< INITIALIZING SPATIAL CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "<<< INITIALIZING OMICS CLASS OBJECT >>>\n",
      "\n",
      "Unused keyword arguments: {'library_id': 'Inflamed-50452B', 'visium': False}.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\u001b[34mINFO    \u001b[0m reading                                                                                                   \n",
      "         \u001b[35m/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/CHO-001/output-XETG00189__0010700__\u001b[0m\n",
      "         \u001b[35m50452B-TUQ97N-EA__20240126__205019/\u001b[0m\u001b[95mcell_feature_matrix.h5\u001b[0m                                                 \n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "dsfasfda AnnData object with n_obs × n_vars = 276881 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Cell Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "Observations: 276881\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Gene Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "Empty DataFrame\n",
      "Columns: [index, gene_ids, feature_types, genome]\n",
      "Index: []\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " SpatialData object with:\n",
      "├── Images\n",
      "│     ├── 'morphology_focus': MultiscaleSpatialImage[cyx] (1, 37492, 51118), (1, 18746, 25559), (1, 9373, 12779), (1, 4686, 6389), (1, 2343, 3194)\n",
      "│     └── 'morphology_mip': MultiscaleSpatialImage[cyx] (1, 37492, 51118), (1, 18746, 25559), (1, 9373, 12779), (1, 4686, 6389), (1, 2343, 3194)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 10) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (276881, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (276881, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (276881, 1) (2D shapes)\n",
      "└── Table\n",
      "      └── AnnData object with n_obs × n_vars = 276881 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts': AnnData (276881, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), morphology_mip (Images), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "None\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 276881 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts' \n",
      "\n",
      "                      gene_ids    feature_types   genome\n",
      "gene_symbols                                           \n",
      "ABCA7         ENSG00000064687  Gene Expression  Unknown\n",
      "ACTA2         ENSG00000107796  Gene Expression  Unknown\n",
      "ACTG2         ENSG00000163017  Gene Expression  Unknown\n",
      "ADAM15        ENSG00000143537  Gene Expression  Unknown\n",
      "ADAM28        ENSG00000042980  Gene Expression  Unknown \n",
      "\n",
      "       cell_id  transcript_counts  control_probe_counts  \\\n",
      "0  aaaahemo-1                145                     0   \n",
      "1  aaaamioh-1                 28                     0   \n",
      "2  aaabdmfg-1                 28                     0   \n",
      "3  aaabjabe-1                145                     0   \n",
      "4  aaabkpll-1                166                     0   \n",
      "\n",
      "   control_codeword_counts  unassigned_codeword_counts  \\\n",
      "0                        0                           0   \n",
      "1                        0                           0   \n",
      "2                        0                           0   \n",
      "3                        0                           0   \n",
      "4                        0                           0   \n",
      "\n",
      "   deprecated_codeword_counts  total_counts  cell_area  nucleus_area  \\\n",
      "0                           0           145  92.434847     14.675782   \n",
      "1                           0            28  25.152032      9.798907   \n",
      "2                           0            28  31.428751     17.249688   \n",
      "3                           0           145  60.057815     28.087189   \n",
      "4                           0           166  80.649065     21.539532   \n",
      "\n",
      "         region           Sample  \n",
      "0  cell_circles  Inflamed-50452B  \n",
      "1  cell_circles  Inflamed-50452B  \n",
      "2  cell_circles  Inflamed-50452B  \n",
      "3  cell_circles  Inflamed-50452B  \n",
      "4  cell_circles  Inflamed-50452B  \n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "col_segment=\"region\"\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\n",
      "\n",
      "<<< INITIALIZING SPATIAL CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "<<< INITIALIZING OMICS CLASS OBJECT >>>\n",
      "\n",
      "Unused keyword arguments: {'library_id': 'Stricture-50452C', 'visium': False}.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\u001b[34mINFO    \u001b[0m reading                                                                                                   \n",
      "         \u001b[35m/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/CHO-001/output-XETG00189__0010663__\u001b[0m\n",
      "         \u001b[35m50452C-TUQ97N-EA__20240126__205019/\u001b[0m\u001b[95mcell_feature_matrix.h5\u001b[0m                                                 \n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "dsfasfda AnnData object with n_obs × n_vars = 613026 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Cell Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "Observations: 613026\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Gene Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "Empty DataFrame\n",
      "Columns: [index, gene_ids, feature_types, genome]\n",
      "Index: []\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " SpatialData object with:\n",
      "├── Images\n",
      "│     ├── 'morphology_focus': MultiscaleSpatialImage[cyx] (1, 98585, 48410), (1, 49292, 24205), (1, 24646, 12102), (1, 12323, 6051), (1, 6161, 3025)\n",
      "│     └── 'morphology_mip': MultiscaleSpatialImage[cyx] (1, 98585, 48410), (1, 49292, 24205), (1, 24646, 12102), (1, 12323, 6051), (1, 6161, 3025)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 10) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (613026, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (613026, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (613026, 1) (2D shapes)\n",
      "└── Table\n",
      "      └── AnnData object with n_obs × n_vars = 613026 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts': AnnData (613026, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), morphology_mip (Images), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "None\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 613026 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts' \n",
      "\n",
      "                      gene_ids    feature_types   genome\n",
      "gene_symbols                                           \n",
      "ABCA7         ENSG00000064687  Gene Expression  Unknown\n",
      "ACTA2         ENSG00000107796  Gene Expression  Unknown\n",
      "ACTG2         ENSG00000163017  Gene Expression  Unknown\n",
      "ADAM15        ENSG00000143537  Gene Expression  Unknown\n",
      "ADAM28        ENSG00000042980  Gene Expression  Unknown \n",
      "\n",
      "       cell_id  transcript_counts  control_probe_counts  \\\n",
      "0  aaaafddg-1                 12                     0   \n",
      "1  aaaafdmg-1                306                     0   \n",
      "2  aaaafkdp-1                203                     0   \n",
      "3  aaaafnae-1                248                     0   \n",
      "4  aaaahffg-1                280                     0   \n",
      "\n",
      "   control_codeword_counts  unassigned_codeword_counts  \\\n",
      "0                        0                           0   \n",
      "1                        0                           0   \n",
      "2                        0                           0   \n",
      "3                        0                           0   \n",
      "4                        0                           0   \n",
      "\n",
      "   deprecated_codeword_counts  total_counts   cell_area  nucleus_area  \\\n",
      "0                           0            12  833.900499      6.547656   \n",
      "1                           0           306  595.610959     23.797345   \n",
      "2                           0           203  353.844388     13.817813   \n",
      "3                           0           248  440.183141     10.882657   \n",
      "4                           0           280  371.726263     54.684221   \n",
      "\n",
      "         region            Sample  \n",
      "0  cell_circles  Stricture-50452C  \n",
      "1  cell_circles  Stricture-50452C  \n",
      "2  cell_circles  Stricture-50452C  \n",
      "3  cell_circles  Stricture-50452C  \n",
      "4  cell_circles  Stricture-50452C  \n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "col_segment=\"region\"\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\n",
      "\n",
      "<<< RESTORING SINGLE-SAMPLE FROM h5ad >>>\n",
      "\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "SpatialData object with:\n",
      "├── Images\n",
      "│     ├── 'morphology_focus': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "│     └── 'morphology_mip': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 10) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (333825, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "└── Table\n",
      "      └── AnnData object with n_obs × n_vars = 312393 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample', 'Sample ID', 'Patient', 'Status', 'Project', 'Sample Location', 'Inflamed', 'Slide', 'GRID ID', 'Procedure Date', 'Age', 'Sex', 'Race', 'Hispanic', 'Diagnosis', 'Project.1', 'Procedure', 'Disease Status', 'Date Collected', 'Date Sectioned', 'Date of Hybridization', 'Storage in 4C', 'Created By', 'Created', 'Storage Status', 'Sample Location.1', 'Storage Row', 'Storage Col', 'Checked Out By', 'Stricture', 'Condition', 'file_path', 'n_counts', 'log_counts', 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'assay_protein', 'col_gene_symbols', 'col_cell_type', 'col_sample_id', 'col_batch', 'col_subject', 'col_condition', 'col_num_umis', 'col_segment', 'cell_filter_pmt', 'cell_filter_ncounts', 'cell_filter_ngene', 'gene_filter_ncell', 'gene_filter_ncounts', 'custom_thresholds', 'figsize', 'kws_scale', 'leiden', 'method_cluster', 'kws_pca', 'kws_umap', 'kws_neighbors', 'kws_cluster', 'resolution', 'genes_subset', 'Annotation'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
      "    uns: 'Annotation_colors', 'Patient_colors', 'Sample_colors', 'Slide_colors', 'hvg', 'leiden', 'leiden_colors', 'log1p', 'neighbors', 'original_ix', 'pca', 'rank_genes_groups', 'rank_genes_groups_leiden', 'spatialdata_attrs', 'umap', 'spatial', 'markers'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    layers: 'counts', 'log1p', 'scaled'\n",
      "    obsp: 'connectivities', 'distances': AnnData (312393, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), morphology_mip (Images), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "\n",
      "\n",
      "<<< RESTORING SINGLE-SAMPLE FROM h5ad >>>\n",
      "\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "SpatialData object with:\n",
      "├── Images\n",
      "│     ├── 'morphology_focus': MultiscaleSpatialImage[cyx] (1, 37492, 51118), (1, 18746, 25559), (1, 9373, 12779), (1, 4686, 6389), (1, 2343, 3194)\n",
      "│     └── 'morphology_mip': MultiscaleSpatialImage[cyx] (1, 37492, 51118), (1, 18746, 25559), (1, 9373, 12779), (1, 4686, 6389), (1, 2343, 3194)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 10) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (276881, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (276881, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (276881, 1) (2D shapes)\n",
      "└── Table\n",
      "      └── AnnData object with n_obs × n_vars = 256836 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample', 'Sample ID', 'Patient', 'Status', 'Project', 'Sample Location', 'Inflamed', 'Slide', 'GRID ID', 'Procedure Date', 'Age', 'Sex', 'Race', 'Hispanic', 'Diagnosis', 'Project.1', 'Procedure', 'Disease Status', 'Date Collected', 'Date Sectioned', 'Date of Hybridization', 'Storage in 4C', 'Created By', 'Created', 'Storage Status', 'Sample Location.1', 'Storage Row', 'Storage Col', 'Checked Out By', 'Stricture', 'Condition', 'file_path', 'n_counts', 'log_counts', 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'assay_protein', 'col_gene_symbols', 'col_cell_type', 'col_sample_id', 'col_batch', 'col_subject', 'col_condition', 'col_num_umis', 'col_segment', 'cell_filter_pmt', 'cell_filter_ncounts', 'cell_filter_ngene', 'gene_filter_ncell', 'gene_filter_ncounts', 'custom_thresholds', 'figsize', 'kws_scale', 'leiden', 'method_cluster', 'kws_pca', 'kws_umap', 'kws_neighbors', 'kws_cluster', 'resolution', 'genes_subset', 'Annotation'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
      "    uns: 'Annotation_colors', 'Patient_colors', 'Sample_colors', 'Slide_colors', 'hvg', 'leiden', 'leiden_colors', 'log1p', 'neighbors', 'original_ix', 'pca', 'rank_genes_groups', 'rank_genes_groups_leiden', 'spatialdata_attrs', 'umap', 'spatial', 'markers'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    layers: 'counts', 'log1p', 'scaled'\n",
      "    obsp: 'connectivities', 'distances': AnnData (256836, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), morphology_mip (Images), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] Unable to synchronously open file (unable to open file: name = '/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/Stricture-50452C__res0pt5_dist1_npc20.h5ad', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m<timed exec>:21\u001b[0m\n",
      "File \u001b[0;32m~/corescpy/corescpy/class_spatial.py:118\u001b[0m, in \u001b[0;36mSpatial.update_from_h5ad\u001b[0;34m(self, file, file_path_markers, method_cluster)\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Update SpatialData object `.table` from h5ad file.\"\"\"\u001b[39;00m\n\u001b[1;32m    116\u001b[0m original_ix \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrna\u001b[38;5;241m.\u001b[39muns[\n\u001b[1;32m    117\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124moriginal_ix\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124moriginal_ix\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrna\u001b[38;5;241m.\u001b[39muns \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 118\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrna \u001b[38;5;241m=\u001b[39m \u001b[43msc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msplitext\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m.h5ad\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    119\u001b[0m csid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_columns[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcol_sample_id\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    120\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39madata, spatialdata\u001b[38;5;241m.\u001b[39mSpatialData) \u001b[38;5;129;01mand\u001b[39;00m (\n\u001b[1;32m    121\u001b[0m         \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrna\u001b[38;5;241m.\u001b[39mobs[csid]\u001b[38;5;241m.\u001b[39munique()) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m):  \u001b[38;5;66;03m# multi-sample\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/py-bio/lib/python3.10/site-packages/scanpy/readwrite.py:111\u001b[0m, in \u001b[0;36mread\u001b[0;34m(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, cache_compression, **kwargs)\u001b[0m\n\u001b[1;32m    109\u001b[0m filename \u001b[38;5;241m=\u001b[39m Path(filename)  \u001b[38;5;66;03m# allow passing strings\u001b[39;00m\n\u001b[1;32m    110\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_valid_filename(filename):\n\u001b[0;32m--> 111\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    112\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    113\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbacked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbacked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    114\u001b[0m \u001b[43m        \u001b[49m\u001b[43msheet\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msheet\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    115\u001b[0m \u001b[43m        \u001b[49m\u001b[43mext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mext\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    116\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdelimiter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdelimiter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    117\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfirst_column_names\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfirst_column_names\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    118\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbackup_url\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbackup_url\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    119\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    120\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcache_compression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache_compression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    121\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    122\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    123\u001b[0m \u001b[38;5;66;03m# generate filename and read to dict\u001b[39;00m\n\u001b[1;32m    124\u001b[0m filekey \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(filename)\n",
      "File \u001b[0;32m~/miniconda3/envs/py-bio/lib/python3.10/site-packages/scanpy/readwrite.py:721\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, cache_compression, suppress_cache_warning, **kwargs)\u001b[0m\n\u001b[1;32m    719\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ext \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mh5\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mh5ad\u001b[39m\u001b[38;5;124m'\u001b[39m}:\n\u001b[1;32m    720\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m sheet \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 721\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mread_h5ad\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbacked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbacked\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    722\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    723\u001b[0m         logg\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mreading sheet \u001b[39m\u001b[38;5;132;01m{\u001b[39;00msheet\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m from file \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfilename\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/py-bio/lib/python3.10/site-packages/anndata/__init__.py:56\u001b[0m, in \u001b[0;36mread\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     49\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mwarnings\u001b[39;00m\n\u001b[1;32m     51\u001b[0m warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m     52\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m`anndata.read` is deprecated, use `anndata.read_h5ad` instead. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     53\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m`ad.read` will be removed in mid 2024.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     54\u001b[0m     \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m     55\u001b[0m )\n\u001b[0;32m---> 56\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mread_h5ad\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/py-bio/lib/python3.10/site-packages/anndata/_io/h5ad.py:237\u001b[0m, in \u001b[0;36mread_h5ad\u001b[0;34m(filename, backed, as_sparse, as_sparse_fmt, chunk_size)\u001b[0m\n\u001b[1;32m    229\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m(\n\u001b[1;32m    230\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCurrently only `X` and `raw/X` can be read as sparse.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    231\u001b[0m         )\n\u001b[1;32m    233\u001b[0m rdasp \u001b[38;5;241m=\u001b[39m partial(\n\u001b[1;32m    234\u001b[0m     read_dense_as_sparse, sparse_format\u001b[38;5;241m=\u001b[39mas_sparse_fmt, axis_chunk\u001b[38;5;241m=\u001b[39mchunk_size\n\u001b[1;32m    235\u001b[0m )\n\u001b[0;32m--> 237\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mh5py\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mFile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mr\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[1;32m    239\u001b[0m     \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcallback\u001b[39m(func, elem_name: \u001b[38;5;28mstr\u001b[39m, elem, iospec):\n\u001b[1;32m    240\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m iospec\u001b[38;5;241m.\u001b[39mencoding_type \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124manndata\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m elem_name\u001b[38;5;241m.\u001b[39mendswith(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n",
      "File \u001b[0;32m~/miniconda3/envs/py-bio/lib/python3.10/site-packages/h5py/_hl/files.py:562\u001b[0m, in \u001b[0;36mFile.__init__\u001b[0;34m(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, fs_page_size, page_buf_size, min_meta_keep, min_raw_keep, locking, alignment_threshold, alignment_interval, meta_block_size, **kwds)\u001b[0m\n\u001b[1;32m    553\u001b[0m     fapl \u001b[38;5;241m=\u001b[39m make_fapl(driver, libver, rdcc_nslots, rdcc_nbytes, rdcc_w0,\n\u001b[1;32m    554\u001b[0m                      locking, page_buf_size, min_meta_keep, min_raw_keep,\n\u001b[1;32m    555\u001b[0m                      alignment_threshold\u001b[38;5;241m=\u001b[39malignment_threshold,\n\u001b[1;32m    556\u001b[0m                      alignment_interval\u001b[38;5;241m=\u001b[39malignment_interval,\n\u001b[1;32m    557\u001b[0m                      meta_block_size\u001b[38;5;241m=\u001b[39mmeta_block_size,\n\u001b[1;32m    558\u001b[0m                      \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m    559\u001b[0m     fcpl \u001b[38;5;241m=\u001b[39m make_fcpl(track_order\u001b[38;5;241m=\u001b[39mtrack_order, fs_strategy\u001b[38;5;241m=\u001b[39mfs_strategy,\n\u001b[1;32m    560\u001b[0m                      fs_persist\u001b[38;5;241m=\u001b[39mfs_persist, fs_threshold\u001b[38;5;241m=\u001b[39mfs_threshold,\n\u001b[1;32m    561\u001b[0m                      fs_page_size\u001b[38;5;241m=\u001b[39mfs_page_size)\n\u001b[0;32m--> 562\u001b[0m     fid \u001b[38;5;241m=\u001b[39m \u001b[43mmake_fid\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43muserblock_size\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfapl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfcpl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mswmr\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mswmr\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    564\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(libver, \u001b[38;5;28mtuple\u001b[39m):\n\u001b[1;32m    565\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_libver \u001b[38;5;241m=\u001b[39m libver\n",
      "File \u001b[0;32m~/miniconda3/envs/py-bio/lib/python3.10/site-packages/h5py/_hl/files.py:235\u001b[0m, in \u001b[0;36mmake_fid\u001b[0;34m(name, mode, userblock_size, fapl, fcpl, swmr)\u001b[0m\n\u001b[1;32m    233\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m swmr \u001b[38;5;129;01mand\u001b[39;00m swmr_support:\n\u001b[1;32m    234\u001b[0m         flags \u001b[38;5;241m|\u001b[39m\u001b[38;5;241m=\u001b[39m h5f\u001b[38;5;241m.\u001b[39mACC_SWMR_READ\n\u001b[0;32m--> 235\u001b[0m     fid \u001b[38;5;241m=\u001b[39m \u001b[43mh5f\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfapl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfapl\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    236\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m mode \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mr+\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m    237\u001b[0m     fid \u001b[38;5;241m=\u001b[39m h5f\u001b[38;5;241m.\u001b[39mopen(name, h5f\u001b[38;5;241m.\u001b[39mACC_RDWR, fapl\u001b[38;5;241m=\u001b[39mfapl)\n",
      "File \u001b[0;32mh5py/_objects.pyx:54\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/_objects.pyx:55\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/h5f.pyx:102\u001b[0m, in \u001b[0;36mh5py.h5f.open\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] Unable to synchronously open file (unable to open file: name = '/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/Stricture-50452C__res0pt5_dist1_npc20.h5ad', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0)"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Load Spatial Data\n",
    "suff = str(f\"res{re.sub('[.]', 'pt', str(resolution))}_dist\"\n",
    "           f\"{re.sub('[.]', 'pt', str(min_dist))}_npc{n_comps}\")  # file end\n",
    "selves, paths_he, file_mks, out_files = [], [], [], []\n",
    "for x in metadata.index.values:\n",
    "    self = cr.Spatial(metadata.loc[x][col_path], library_id=x, **kws_init)\n",
    "    for i in metadata:  # add metadata for subject\n",
    "        self.rna.obs.loc[:, i] = str(metadata.loc[x][i])  # add metadata\n",
    "    selves += [self]\n",
    "    paths_he += [os.path.join(metadata.loc[x][\n",
    "        col_path], \"aux_outputs/image_he.ome.tif\")]  # H&E paths\n",
    "        # out_files += [os.path.join(out_dir, f\"{x}.zarr\")]\n",
    "    if out_dir is not None:\n",
    "        out_files += [os.path.join(out_dir, f\"{x}__{suff}.zarr\")]\n",
    "    file_mks += [os.path.join(out_dir, f\"{x}__{suff}_markers.csv\")]\n",
    "\n",
    "# Reload Processed & Clustered Data (Optionally)\n",
    "if reload is True:\n",
    "    for i, s in enumerate(selves):\n",
    "        mks = file_mks[i] if os.path.exists(file_mks[i]) else None\n",
    "        s.update_from_h5ad(file=out_files[i], file_path_markers=mks)\n",
    "        print(s.adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if reload is True:\n",
    "    for i, s in enumerate(selves):\n",
    "        s.plot_spatial(color=\"leiden\")\n",
    "        # s.plot_spatial(color=col_cell_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotation & Sub-Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in selves:\n",
    "    _ = s.cluster(**kws_cluster)\n",
    "    _ = s.annotate_clusters(file_ann, col_cell_type=\"leiden\",\n",
    "                            col_annotation=col_cell_type,\n",
    "                            col_assignment=col_assignment)  # annotate & write\n",
    "    s.plot_spatial(color=col_cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cat_epi = [\"BEST2+ Goblet Cell\", \"BEST4+ Epithelial\", \"EC Cells (NPW+)\",\n",
    "#            \"EC Cells (TAC1+)\", \"Enterocyte\", \"Epithelial\", \"Goblet Cell\",\n",
    "#            \"I Cells (CCK+)\", \"Paneth\", \"D Cells (SST+)\",\n",
    "#            \"Stem Cells\", \"N Cells (NTS+)\", \"K Cells (GIP+)\",\n",
    "#            # \"Colonocyte\",\n",
    "#            \"L Cells (PYY+)\", \"Tuft\", \"Microfold Cell\"]\n",
    "# cat_epi = list(np.array()[np.where([\"epithelial\" in x.lower()\n",
    "#                                     for x in marker_genes_dict])[0]])\n",
    "# file_ann_epi = pd.concat([pd.Series(marker_genes_dict[x]) for x in cat_epi],\n",
    "#                          keys=cat_epi, names=[\"Type\"]).to_frame(\n",
    "#                              \"Gene\").reset_index(0).set_index(\"Gene\")\n",
    "file_ann_epi = file_ann\n",
    "# key_cell_type, col_ann = [\"7\"], \"Subclustering_Epi\"\n",
    "key_cell_type, col_ann = None, \"leiden_sub\"\n",
    "for s in selves:\n",
    "    ann = s.subcluster(restrict_to=(col_cell_type, key_cell_type), copy=False,\n",
    "                       key_added=\"leiden_sub\", **kws_subcluster)\n",
    "    s.annotate_clusters(kws_annotation={\"model\": file_ann_epi,\n",
    "                                        \"col_assignment\": \"Bin\"},\n",
    "                        col_annotation=[col_ann, col_cell_type])\n",
    "    s.plot_spatial(color=col_ann)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in selves:\n",
    "    s.write(out_files[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workspace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "<<< INITIALIZING SPATIAL CLASS OBJECT >>>\n",
      "\n",
      "\n",
      "\n",
      "<<< INITIALIZING OMICS CLASS OBJECT >>>\n",
      "\n",
      "Unused keyword arguments: {'library_id': 'Stricture-50452C', 'visium': False}.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\u001b[34mINFO    \u001b[0m reading                                                                                                   \n",
      "         \u001b[35m/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/CHO-001/output-XETG00189__0010700__\u001b[0m\n",
      "         \u001b[35m50452A-TUQ97N-EA__20240126__205019/\u001b[0m\u001b[95mcell_feature_matrix.h5\u001b[0m                                                 \n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "dsfasfda AnnData object with n_obs × n_vars = 333825 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Cell Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "Observations: 333825\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Gene Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "Empty DataFrame\n",
      "Columns: [index, gene_ids, feature_types, genome]\n",
      "Index: []\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " SpatialData object with:\n",
      "├── Images\n",
      "│     ├── 'morphology_focus': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "│     └── 'morphology_mip': MultiscaleSpatialImage[cyx] (1, 57808, 45580), (1, 28904, 22790), (1, 14452, 11395), (1, 7226, 5697), (1, 3613, 2848)\n",
      "├── Points\n",
      "│     └── 'transcripts': DataFrame with shape: (<Delayed>, 10) (3D points)\n",
      "├── Shapes\n",
      "│     ├── 'cell_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "│     ├── 'cell_circles': GeoDataFrame shape: (333825, 2) (2D shapes)\n",
      "│     └── 'nucleus_boundaries': GeoDataFrame shape: (333825, 1) (2D shapes)\n",
      "└── Table\n",
      "      └── AnnData object with n_obs × n_vars = 333825 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts': AnnData (333825, 469)\n",
      "with coordinate systems:\n",
      "▸ 'global', with elements:\n",
      "        morphology_focus (Images), morphology_mip (Images), transcripts (Points), cell_boundaries (Shapes), cell_circles (Shapes), nucleus_boundaries (Shapes)\n",
      "None\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 333825 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts' \n",
      "\n",
      "                      gene_ids    feature_types   genome\n",
      "gene_symbols                                           \n",
      "ABCA7         ENSG00000064687  Gene Expression  Unknown\n",
      "ACTA2         ENSG00000107796  Gene Expression  Unknown\n",
      "ACTG2         ENSG00000163017  Gene Expression  Unknown\n",
      "ADAM15        ENSG00000143537  Gene Expression  Unknown\n",
      "ADAM28        ENSG00000042980  Gene Expression  Unknown \n",
      "\n",
      "       cell_id  transcript_counts  control_probe_counts  \\\n",
      "0  aaaaibki-1                168                     0   \n",
      "1  aaaaiiji-1                166                     0   \n",
      "2  aaaajape-1                135                     0   \n",
      "3  aaaamhbm-1                 99                     0   \n",
      "4  aaabdoak-1                 61                     0   \n",
      "\n",
      "   control_codeword_counts  unassigned_codeword_counts  \\\n",
      "0                        0                           0   \n",
      "1                        0                           0   \n",
      "2                        0                           0   \n",
      "3                        0                           0   \n",
      "4                        0                           0   \n",
      "\n",
      "   deprecated_codeword_counts  total_counts   cell_area  nucleus_area  \\\n",
      "0                           0           168  328.150481     13.908126   \n",
      "1                           0           166  107.607348     18.649532   \n",
      "2                           0           135  110.226410     10.250469   \n",
      "3                           0            99   46.736720     16.482032   \n",
      "4                           0            61   22.216876      6.954063   \n",
      "\n",
      "         region            Sample  \n",
      "0  cell_circles  Stricture-50452C  \n",
      "1  cell_circles  Stricture-50452C  \n",
      "2  cell_circles  Stricture-50452C  \n",
      "3  cell_circles  Stricture-50452C  \n",
      "4  cell_circles  Stricture-50452C  \n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"Annotation\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Slide\"\n",
      "col_subject=\"Patient\"\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "col_segment=\"region\"\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n"
     ]
    }
   ],
   "source": [
    "self = cr.Spatial(metadata.iloc[0][col_path], library_id=x, **kws_init)\n",
    "sdata = self.adata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>Diagnosis</th>\n",
       "      <th>Fraction</th>\n",
       "      <th>batch</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>doublet_scores</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>category</th>\n",
       "      <th>Age_group</th>\n",
       "      <th>author_cell_type</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>hca_data_portal_donor_uuid</th>\n",
       "      <th>hca_data_portal_cellsuspension_uuid</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>assay</th>\n",
       "      <th>disease</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>development_stage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcodes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAACT-1-4918STDY7333456</th>\n",
       "      <td>T036</td>\n",
       "      <td>Pediatric healthy</td>\n",
       "      <td>SC-EPCAMP</td>\n",
       "      <td>4918STDY7333456</td>\n",
       "      <td>968</td>\n",
       "      <td>968</td>\n",
       "      <td>1912.7936</td>\n",
       "      <td>46.228200</td>\n",
       "      <td>2.416790</td>\n",
       "      <td>0.016442</td>\n",
       "      <td>False</td>\n",
       "      <td>B cells</td>\n",
       "      <td>Pediatric</td>\n",
       "      <td>DZ GC cell</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0000844</td>\n",
       "      <td>HsapDv:0000098</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>9b2a3014-c6db-4426-8676-81b5ce2d93b2</td>\n",
       "      <td>f0e61dad-fc4f-48df-bf6c-8c53bb3adde2</td>\n",
       "      <td>germinal center B cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCGATATA-1-4918STDY7333456</th>\n",
       "      <td>T036</td>\n",
       "      <td>Pediatric healthy</td>\n",
       "      <td>SC-EPCAMP</td>\n",
       "      <td>4918STDY7333456</td>\n",
       "      <td>693</td>\n",
       "      <td>693</td>\n",
       "      <td>2152.0894</td>\n",
       "      <td>0.122579</td>\n",
       "      <td>0.005696</td>\n",
       "      <td>0.027389</td>\n",
       "      <td>False</td>\n",
       "      <td>B cells</td>\n",
       "      <td>Pediatric</td>\n",
       "      <td>Cycling B cell</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0000236</td>\n",
       "      <td>HsapDv:0000098</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>9b2a3014-c6db-4426-8676-81b5ce2d93b2</td>\n",
       "      <td>f0e61dad-fc4f-48df-bf6c-8c53bb3adde2</td>\n",
       "      <td>B cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGAGTTTA-1-4918STDY7333456</th>\n",
       "      <td>T036</td>\n",
       "      <td>Pediatric healthy</td>\n",
       "      <td>SC-EPCAMP</td>\n",
       "      <td>4918STDY7333456</td>\n",
       "      <td>720</td>\n",
       "      <td>720</td>\n",
       "      <td>1686.8834</td>\n",
       "      <td>80.195816</td>\n",
       "      <td>4.754082</td>\n",
       "      <td>0.014817</td>\n",
       "      <td>False</td>\n",
       "      <td>T cells</td>\n",
       "      <td>Pediatric</td>\n",
       "      <td>gdT</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0000798</td>\n",
       "      <td>HsapDv:0000098</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>9b2a3014-c6db-4426-8676-81b5ce2d93b2</td>\n",
       "      <td>f0e61dad-fc4f-48df-bf6c-8c53bb3adde2</td>\n",
       "      <td>gamma-delta T cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTATCTCG-1-4918STDY7333456</th>\n",
       "      <td>T036</td>\n",
       "      <td>Pediatric healthy</td>\n",
       "      <td>SC-EPCAMP</td>\n",
       "      <td>4918STDY7333456</td>\n",
       "      <td>746</td>\n",
       "      <td>746</td>\n",
       "      <td>2135.1838</td>\n",
       "      <td>104.506760</td>\n",
       "      <td>4.894509</td>\n",
       "      <td>0.039416</td>\n",
       "      <td>False</td>\n",
       "      <td>B cells</td>\n",
       "      <td>Pediatric</td>\n",
       "      <td>Memory B</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0000787</td>\n",
       "      <td>HsapDv:0000098</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>9b2a3014-c6db-4426-8676-81b5ce2d93b2</td>\n",
       "      <td>f0e61dad-fc4f-48df-bf6c-8c53bb3adde2</td>\n",
       "      <td>memory B cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTGACTCT-1-4918STDY7333456</th>\n",
       "      <td>T036</td>\n",
       "      <td>Pediatric healthy</td>\n",
       "      <td>SC-EPCAMP</td>\n",
       "      <td>4918STDY7333456</td>\n",
       "      <td>1227</td>\n",
       "      <td>1227</td>\n",
       "      <td>3103.0503</td>\n",
       "      <td>95.846080</td>\n",
       "      <td>3.088770</td>\n",
       "      <td>0.025763</td>\n",
       "      <td>False</td>\n",
       "      <td>B cells</td>\n",
       "      <td>Pediatric</td>\n",
       "      <td>DZ GC cell</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0000844</td>\n",
       "      <td>HsapDv:0000098</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>9b2a3014-c6db-4426-8676-81b5ce2d93b2</td>\n",
       "      <td>f0e61dad-fc4f-48df-bf6c-8c53bb3adde2</td>\n",
       "      <td>germinal center B cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTATGCTCTTCCTTC-1-WTDAtest7844019</th>\n",
       "      <td>A33 (414C)</td>\n",
       "      <td>Healthy adult</td>\n",
       "      <td>SC-45P</td>\n",
       "      <td>WTDAtest7844019</td>\n",
       "      <td>1015</td>\n",
       "      <td>1015</td>\n",
       "      <td>1993.9489</td>\n",
       "      <td>69.591800</td>\n",
       "      <td>3.490149</td>\n",
       "      <td>0.039728</td>\n",
       "      <td>False</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Enterocyte</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:1000347</td>\n",
       "      <td>HsapDv:0000090</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>enterocyte of colon</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>25-44 year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCCTCTCTCGTTTA-1-WTDAtest7844019</th>\n",
       "      <td>A33 (414C)</td>\n",
       "      <td>Healthy adult</td>\n",
       "      <td>SC-45P</td>\n",
       "      <td>WTDAtest7844019</td>\n",
       "      <td>582</td>\n",
       "      <td>582</td>\n",
       "      <td>971.5292</td>\n",
       "      <td>51.639854</td>\n",
       "      <td>5.315317</td>\n",
       "      <td>0.077900</td>\n",
       "      <td>False</td>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Mature venous EC</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:0002543</td>\n",
       "      <td>HsapDv:0000090</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>vein endothelial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>25-44 year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCGCCACACCGAC-1-WTDAtest7844019</th>\n",
       "      <td>A33 (414C)</td>\n",
       "      <td>Healthy adult</td>\n",
       "      <td>SC-45P</td>\n",
       "      <td>WTDAtest7844019</td>\n",
       "      <td>640</td>\n",
       "      <td>640</td>\n",
       "      <td>1095.2365</td>\n",
       "      <td>276.850860</td>\n",
       "      <td>25.277725</td>\n",
       "      <td>0.048101</td>\n",
       "      <td>False</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Enterocyte</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:1000347</td>\n",
       "      <td>HsapDv:0000090</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>enterocyte of colon</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>25-44 year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCGCTCCCTGACT-1-WTDAtest7844019</th>\n",
       "      <td>A33 (414C)</td>\n",
       "      <td>Healthy adult</td>\n",
       "      <td>SC-45P</td>\n",
       "      <td>WTDAtest7844019</td>\n",
       "      <td>1940</td>\n",
       "      <td>1940</td>\n",
       "      <td>4612.1630</td>\n",
       "      <td>79.811610</td>\n",
       "      <td>1.730459</td>\n",
       "      <td>0.098630</td>\n",
       "      <td>False</td>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Mature arterial EC</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:1000413</td>\n",
       "      <td>HsapDv:0000090</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>endothelial cell of artery</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>25-44 year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCTTGCCGT-1-WTDAtest7844019</th>\n",
       "      <td>A33 (414C)</td>\n",
       "      <td>Healthy adult</td>\n",
       "      <td>SC-45P</td>\n",
       "      <td>WTDAtest7844019</td>\n",
       "      <td>547</td>\n",
       "      <td>547</td>\n",
       "      <td>799.6813</td>\n",
       "      <td>56.212936</td>\n",
       "      <td>7.029418</td>\n",
       "      <td>0.084044</td>\n",
       "      <td>False</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Enterocyte</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>CL:1000347</td>\n",
       "      <td>HsapDv:0000090</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>unknown</td>\n",
       "      <td>False</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>UBERON:0002108</td>\n",
       "      <td>cell</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>enterocyte of colon</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>unknown</td>\n",
       "      <td>25-44 year-old human stage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>265248 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      donor_id          Diagnosis   Fraction  \\\n",
       "barcodes                                                                       \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456        T036  Pediatric healthy  SC-EPCAMP   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456        T036  Pediatric healthy  SC-EPCAMP   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456        T036  Pediatric healthy  SC-EPCAMP   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456        T036  Pediatric healthy  SC-EPCAMP   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456        T036  Pediatric healthy  SC-EPCAMP   \n",
       "...                                        ...                ...        ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019  A33 (414C)      Healthy adult     SC-45P   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019  A33 (414C)      Healthy adult     SC-45P   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019  A33 (414C)      Healthy adult     SC-45P   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019  A33 (414C)      Healthy adult     SC-45P   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019  A33 (414C)      Healthy adult     SC-45P   \n",
       "\n",
       "                                              batch  n_genes  \\\n",
       "barcodes                                                       \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456  4918STDY7333456      968   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456  4918STDY7333456      693   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456  4918STDY7333456      720   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456  4918STDY7333456      746   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456  4918STDY7333456     1227   \n",
       "...                                             ...      ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019  WTDAtest7844019     1015   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019  WTDAtest7844019      582   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019  WTDAtest7844019      640   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019  WTDAtest7844019     1940   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019  WTDAtest7844019      547   \n",
       "\n",
       "                                    n_genes_by_counts  total_counts  \\\n",
       "barcodes                                                              \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456                968     1912.7936   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456                693     2152.0894   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456                720     1686.8834   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456                746     2135.1838   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456               1227     3103.0503   \n",
       "...                                               ...           ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019               1015     1993.9489   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019                582      971.5292   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019                640     1095.2365   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019               1940     4612.1630   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019                547      799.6813   \n",
       "\n",
       "                                    total_counts_mt  pct_counts_mt  \\\n",
       "barcodes                                                             \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456        46.228200       2.416790   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456         0.122579       0.005696   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456        80.195816       4.754082   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456       104.506760       4.894509   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456        95.846080       3.088770   \n",
       "...                                             ...            ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019        69.591800       3.490149   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019        51.639854       5.315317   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019       276.850860      25.277725   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019        79.811610       1.730459   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019        56.212936       7.029418   \n",
       "\n",
       "                                    doublet_scores  predicted_doublets  \\\n",
       "barcodes                                                                 \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456        0.016442               False   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456        0.027389               False   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456        0.014817               False   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456        0.039416               False   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456        0.025763               False   \n",
       "...                                            ...                 ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019        0.039728               False   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019        0.077900               False   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019        0.048101               False   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019        0.098630               False   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019        0.084044               False   \n",
       "\n",
       "                                       category  Age_group  \\\n",
       "barcodes                                                     \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456      B cells  Pediatric   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456      B cells  Pediatric   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456      T cells  Pediatric   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456      B cells  Pediatric   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456      B cells  Pediatric   \n",
       "...                                         ...        ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019   Epithelial      Adult   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019  Endothelial      Adult   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019   Epithelial      Adult   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019  Endothelial      Adult   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019   Epithelial      Adult   \n",
       "\n",
       "                                      author_cell_type assay_ontology_term_id  \\\n",
       "barcodes                                                                        \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456          DZ GC cell            EFO:0009899   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456      Cycling B cell            EFO:0009899   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456                 gdT            EFO:0009899   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456            Memory B            EFO:0009899   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456          DZ GC cell            EFO:0009899   \n",
       "...                                                ...                    ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019          Enterocyte            EFO:0009899   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019    Mature venous EC            EFO:0009899   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019          Enterocyte            EFO:0009899   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019  Mature arterial EC            EFO:0009899   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019          Enterocyte            EFO:0009899   \n",
       "\n",
       "                                   cell_type_ontology_term_id  \\\n",
       "barcodes                                                        \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456                 CL:0000844   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456                 CL:0000236   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456                 CL:0000798   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456                 CL:0000787   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456                 CL:0000844   \n",
       "...                                                       ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019                 CL:1000347   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019                 CL:0002543   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019                 CL:1000347   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019                 CL:1000413   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019                 CL:1000347   \n",
       "\n",
       "                                   development_stage_ontology_term_id  \\\n",
       "barcodes                                                                \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456                     HsapDv:0000098   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456                     HsapDv:0000098   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456                     HsapDv:0000098   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456                     HsapDv:0000098   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456                     HsapDv:0000098   \n",
       "...                                                               ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019                     HsapDv:0000090   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019                     HsapDv:0000090   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019                     HsapDv:0000090   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019                     HsapDv:0000090   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019                     HsapDv:0000090   \n",
       "\n",
       "                                   disease_ontology_term_id  \\\n",
       "barcodes                                                      \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456             PATO:0000461   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456             PATO:0000461   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456             PATO:0000461   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456             PATO:0000461   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456             PATO:0000461   \n",
       "...                                                     ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019             PATO:0000461   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019             PATO:0000461   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019             PATO:0000461   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019             PATO:0000461   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019             PATO:0000461   \n",
       "\n",
       "                                   self_reported_ethnicity_ontology_term_id  \\\n",
       "barcodes                                                                      \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456                                  unknown   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456                                  unknown   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456                                  unknown   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456                                  unknown   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456                                  unknown   \n",
       "...                                                                     ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019                                  unknown   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019                                  unknown   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019                                  unknown   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019                                  unknown   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019                                  unknown   \n",
       "\n",
       "                                    is_primary_data organism_ontology_term_id  \\\n",
       "barcodes                                                                        \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456            False            NCBITaxon:9606   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456            False            NCBITaxon:9606   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456            False            NCBITaxon:9606   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456            False            NCBITaxon:9606   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456            False            NCBITaxon:9606   \n",
       "...                                             ...                       ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019            False            NCBITaxon:9606   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019            False            NCBITaxon:9606   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019            False            NCBITaxon:9606   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019            False            NCBITaxon:9606   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019            False            NCBITaxon:9606   \n",
       "\n",
       "                                   sex_ontology_term_id  \\\n",
       "barcodes                                                  \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456         PATO:0000384   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456         PATO:0000384   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456         PATO:0000384   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456         PATO:0000384   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456         PATO:0000384   \n",
       "...                                                 ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019         PATO:0000384   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019         PATO:0000384   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019         PATO:0000384   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019         PATO:0000384   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019         PATO:0000384   \n",
       "\n",
       "                                   tissue_ontology_term_id suspension_type  \\\n",
       "barcodes                                                                     \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456          UBERON:0002108            cell   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456          UBERON:0002108            cell   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456          UBERON:0002108            cell   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456          UBERON:0002108            cell   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456          UBERON:0002108            cell   \n",
       "...                                                    ...             ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019          UBERON:0002108            cell   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019          UBERON:0002108            cell   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019          UBERON:0002108            cell   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019          UBERON:0002108            cell   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019          UBERON:0002108            cell   \n",
       "\n",
       "                                              hca_data_portal_donor_uuid  \\\n",
       "barcodes                                                                   \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456  9b2a3014-c6db-4426-8676-81b5ce2d93b2   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456  9b2a3014-c6db-4426-8676-81b5ce2d93b2   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456  9b2a3014-c6db-4426-8676-81b5ce2d93b2   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456  9b2a3014-c6db-4426-8676-81b5ce2d93b2   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456  9b2a3014-c6db-4426-8676-81b5ce2d93b2   \n",
       "...                                                                  ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019                                  None   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019                                  None   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019                                  None   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019                                  None   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019                                  None   \n",
       "\n",
       "                                     hca_data_portal_cellsuspension_uuid  \\\n",
       "barcodes                                                                   \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456  f0e61dad-fc4f-48df-bf6c-8c53bb3adde2   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456  f0e61dad-fc4f-48df-bf6c-8c53bb3adde2   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456  f0e61dad-fc4f-48df-bf6c-8c53bb3adde2   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456  f0e61dad-fc4f-48df-bf6c-8c53bb3adde2   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456  f0e61dad-fc4f-48df-bf6c-8c53bb3adde2   \n",
       "...                                                                  ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019                                  None   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019                                  None   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019                                  None   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019                                  None   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019                                  None   \n",
       "\n",
       "                                                     cell_type      assay  \\\n",
       "barcodes                                                                    \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456      germinal center B cell  10x 3' v2   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456                      B cell  10x 3' v2   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456          gamma-delta T cell  10x 3' v2   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456               memory B cell  10x 3' v2   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456      germinal center B cell  10x 3' v2   \n",
       "...                                                        ...        ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019         enterocyte of colon  10x 3' v2   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019       vein endothelial cell  10x 3' v2   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019         enterocyte of colon  10x 3' v2   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019  endothelial cell of artery  10x 3' v2   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019         enterocyte of colon  10x 3' v2   \n",
       "\n",
       "                                   disease      organism   sex  \\\n",
       "barcodes                                                         \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456  normal  Homo sapiens  male   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456  normal  Homo sapiens  male   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456  normal  Homo sapiens  male   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456  normal  Homo sapiens  male   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456  normal  Homo sapiens  male   \n",
       "...                                    ...           ...   ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019  normal  Homo sapiens  male   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019  normal  Homo sapiens  male   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019  normal  Homo sapiens  male   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019  normal  Homo sapiens  male   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019  normal  Homo sapiens  male   \n",
       "\n",
       "                                             tissue self_reported_ethnicity  \\\n",
       "barcodes                                                                      \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456  small intestine                 unknown   \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456  small intestine                 unknown   \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456  small intestine                 unknown   \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456  small intestine                 unknown   \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456  small intestine                 unknown   \n",
       "...                                             ...                     ...   \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019  small intestine                 unknown   \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019  small intestine                 unknown   \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019  small intestine                 unknown   \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019  small intestine                 unknown   \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019  small intestine                 unknown   \n",
       "\n",
       "                                             development_stage  \n",
       "barcodes                                                        \n",
       "AAACCTGAGAACAACT-1-4918STDY7333456      4-year-old human stage  \n",
       "AAACCTGAGCGATATA-1-4918STDY7333456      4-year-old human stage  \n",
       "AAACCTGAGGAGTTTA-1-4918STDY7333456      4-year-old human stage  \n",
       "AAACCTGAGTATCTCG-1-4918STDY7333456      4-year-old human stage  \n",
       "AAACCTGAGTGACTCT-1-4918STDY7333456      4-year-old human stage  \n",
       "...                                                        ...  \n",
       "TTTATGCTCTTCCTTC-1-WTDAtest7844019  25-44 year-old human stage  \n",
       "TTTCCTCTCTCGTTTA-1-WTDAtest7844019  25-44 year-old human stage  \n",
       "TTTGCGCCACACCGAC-1-WTDAtest7844019  25-44 year-old human stage  \n",
       "TTTGCGCTCCCTGACT-1-WTDAtest7844019  25-44 year-old human stage  \n",
       "TTTGGTTTCTTGCCGT-1-WTDAtest7844019  25-44 year-old human stage  \n",
       "\n",
       "[265248 rows x 34 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sc.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata = selves[0].adata\n",
    "# adata_sc = sc.read(os.path.join(\"/mnt/cho_lab/disk2/elizabeth/data\",\n",
    "#                                 \"2023-05-12_CombinedCD-v2_new.h5ad\"))\n",
    "# adata_sc = adata_sc[adata_sc.obs.Region == \"Ileum\"]\n",
    "# sdata_new, sdata, adata_sc, ad_map, df_compare = cr.pp.integrate_spatial(\n",
    "#     sdata, adata_sc, col_cell_type=\"ClusterAnnotation\")\n",
    "# adata_sc = sc.read(os.path.join(\"/mnt/cho_lab/disk2/elizabeth/data\",\n",
    "#                                 \"elmentaite.h5ad\"))\n",
    "# adata_sc = adata_sc[adata_sc.obs.tissue == \"small intestine\"]\n",
    "sdata_new, sdata, adata_sc, ad_map, df_compare = cr.pp.integrate_spatial(\n",
    "    sdata, adata_sc, col_cell_type=\"cell_type\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key_cell_type, col_ann = None, \"leiden_sub\"\n",
    "for s in selves:\n",
    "    s.plot_spatial(color=col_ann)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s.annotate_clusters(file_ann, col_cell_type=\"leiden\", col_annotation=\"Annotation\")\n",
    "s.annotate_clusters(file_ann, col_cell_type=\"leiden_subcluster\", col_annotation=\"Annotation_Subcluster\")\n",
    "s.plot_spatial(color=\"Annotation_Subcluster\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dff = pd.concat([pd.Series(marker_genes_dict[x]).to_frame(\"Marker\")\n",
    "                 for x in marker_genes_dict], keys=marker_genes_dict)\n",
    "dff[dff.Marker.isin(self.rna.var_names)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adf = pd.read_excel(file_ann)\n",
    "adf\n",
    "\n",
    "\n",
    "i = 0\n",
    "self = selves[i]\n",
    "self.add_image(paths_he[i], name=\"H_E\", file_align=paths_he_align[i])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py-bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
