{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #808000; text-decoration-color: #808000; font-weight: bold\">ryp2 is not installed. Install with </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">pip install rpy2 </span><span style=\"color: #808000; text-decoration-color: #808000; font-weight: bold\">to run tools with R support.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;33mryp2 is not installed. Install with \u001b[0m\u001b[1;32mpip install rpy2 \u001b[0m\u001b[1;33mto run tools with R support.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-05-30 14:33:41.095464: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /home/elizabeth/elizabeth/miniconda3/envs/py-bio/lib/python3.10/site-packages/cv2/../../lib64:/usr/local/root/lib:/usr/local/root/lib:/usr/local/root/bindings/pyroot:/usr/local/cuda-9.0/extras/CUPTI/lib64:/usr/local/cuda-9.0/lib64:/usr/local/root/lib:/usr/local/root/lib:/usr/local/root/bindings/pyroot:/usr/local/cuda-9.0/extras/CUPTI/lib64:/usr/local/cuda-9.0/lib64\n",
      "2024-05-30 14:33:41.958533: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /home/elizabeth/elizabeth/miniconda3/envs/py-bio/lib/python3.10/site-packages/cv2/../../lib64:/usr/local/root/lib:/usr/local/root/lib:/usr/local/root/bindings/pyroot:/usr/local/cuda-9.0/extras/CUPTI/lib64:/usr/local/cuda-9.0/lib64:/usr/local/root/lib:/usr/local/root/lib:/usr/local/root/bindings/pyroot:/usr/local/cuda-9.0/extras/CUPTI/lib64:/usr/local/cuda-9.0/lib64\n",
      "2024-05-30 14:33:41.958657: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /home/elizabeth/elizabeth/miniconda3/envs/py-bio/lib/python3.10/site-packages/cv2/../../lib64:/usr/local/root/lib:/usr/local/root/lib:/usr/local/root/bindings/pyroot:/usr/local/cuda-9.0/extras/CUPTI/lib64:/usr/local/cuda-9.0/lib64:/usr/local/root/lib:/usr/local/root/lib:/usr/local/root/bindings/pyroot:/usr/local/cuda-9.0/extras/CUPTI/lib64:/usr/local/cuda-9.0/lib64\n",
      "2024-05-30 14:33:41.958667: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "partially initialized module 'corescpy' has no attribute 'pp' (most likely due to a circular import)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 22\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[1;32m     21\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mnp\u001b[39;00m\n\u001b[0;32m---> 22\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mcorescpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mcr\u001b[39;00m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mcorescpy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01max\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m create_spot_grid\n\u001b[1;32m     25\u001b[0m warnings\u001b[38;5;241m.\u001b[39mfilterwarnings(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/corescpy/corescpy/__init__.py:5\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# __init__.py\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# pylint: disable=unused-import\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01msys\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m utils \u001b[38;5;28;01mas\u001b[39;00m tl\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m processing \u001b[38;5;28;01mas\u001b[39;00m pp\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m analysis \u001b[38;5;28;01mas\u001b[39;00m ax\n",
      "File \u001b[0;32m~/corescpy/corescpy/utils/__init__.py:4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdisplay\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (print_dictionary, make_printable_object,\n\u001b[1;32m      2\u001b[0m                       print_pretty_dictionary, explore_h5_file, print_counts)\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmath\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m is_outlier\n\u001b[0;32m----> 4\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdata_manipulation\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (create_pseudobulk, create_condition_combo,\n\u001b[1;32m      5\u001b[0m                                 merge_pca_subset, write_ome_tif)\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01margument_manipulation\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m to_list, merge\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mresources\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m get_markers_database, get_topp_gene\n",
      "File \u001b[0;32m~/corescpy/corescpy/utils/data_manipulation.py:17\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mnp\u001b[39;00m\n\u001b[0;32m---> 17\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mcorescpy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mprocessing\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m get_layer_dict\n\u001b[1;32m     19\u001b[0m layers \u001b[38;5;241m=\u001b[39m get_layer_dict()\n\u001b[1;32m     22\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcreate_pseudobulk\u001b[39m(adata, col_cell_type, col_sample_id\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m     23\u001b[0m                       layer\u001b[38;5;241m=\u001b[39mlayers[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcounts\u001b[39m\u001b[38;5;124m\"\u001b[39m], mode\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msum\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     24\u001b[0m                       kws_process\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n",
      "File \u001b[0;32m~/corescpy/corescpy/processing/__init__.py:14\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mguide_rna\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m process_guide_rna, process_guide_rna_custom\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mimporting\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (combine_matrix_protospacer, construct_file,\n\u001b[1;32m     12\u001b[0m                         process_multimodal_crispr)\n\u001b[0;32m---> 14\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mspatial_pp\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (describe_tiff, extract_tiff, command_cellpose,\n\u001b[1;32m     15\u001b[0m                          read_spatial, update_spatial_uns,\n\u001b[1;32m     16\u001b[0m                          _get_control_probe_names, impute_spatial,\n\u001b[1;32m     17\u001b[0m                          construct_obs_spatial_imputation, create_spot_grid,\n\u001b[1;32m     18\u001b[0m                          SPATIAL_KEY, SPATIAL_IMAGE_KEY_SEP)\n\u001b[1;32m     20\u001b[0m __all__ \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     21\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcreate_object\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcreate_object_multi\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mget_layer_dict\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprocess_data\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     22\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcheck_normalization\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mz_normalize_by_reference\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mperform_qc\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     29\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSPATIAL_KEY\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSPATIAL_IMAGE_KEY_SEP\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     30\u001b[0m ]\n",
      "File \u001b[0;32m~/corescpy/corescpy/processing/spatial_pp.py:32\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mnp\u001b[39;00m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mcorescpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mcr\u001b[39;00m\n\u001b[0;32m---> 32\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mcorescpy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mvisualization\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m plot_space\n\u001b[1;32m     34\u001b[0m \u001b[38;5;66;03m# Define constant.\u001b[39;00m\n\u001b[1;32m     35\u001b[0m \u001b[38;5;66;03m# z-slices are 3 microns apart\u001b[39;00m\n\u001b[1;32m     36\u001b[0m Z_SLICE_MICRON \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m3\u001b[39m\n",
      "File \u001b[0;32m~/corescpy/corescpy/visualization/__init__.py:13\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcommunication_plots\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m plot_receptor_ligand, plot_cooccurrence\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mscanpy_plots\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m plot_hvgs\n\u001b[0;32m---> 13\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mspatial_plots\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (plot_tiff, plot_spatial, plot_space,\n\u001b[1;32m     14\u001b[0m                             plot_integration_spatial)\n\u001b[1;32m     16\u001b[0m __all__ \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     17\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msquare_grid\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplot_umap\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplot_clustering\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     18\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplot_targeting_efficiency\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplot_mixscape\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplot_spatial\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplot_space\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplot_integration_spatial\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     24\u001b[0m ]\n",
      "File \u001b[0;32m~/corescpy/corescpy/visualization/spatial_plots.py:43\u001b[0m\n\u001b[1;32m     38\u001b[0m         plt\u001b[38;5;241m.\u001b[39maxis(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mScaled\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     39\u001b[0m         plt\u001b[38;5;241m.\u001b[39mshow()\n\u001b[1;32m     42\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mplot_spatial\u001b[39m(adata, color\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mleiden\u001b[39m\u001b[38;5;124m\"\u001b[39m, col_segment\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, figsize\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m20\u001b[39m,\n\u001b[0;32m---> 43\u001b[0m                  spatial_key\u001b[38;5;241m=\u001b[39m\u001b[43mcr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpp\u001b[49m\u001b[38;5;241m.\u001b[39mSPATIAL_KEY, key_image\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m     44\u001b[0m                  library_id\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, col_sample_id\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m     45\u001b[0m                  wspace\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.1\u001b[39m, shape\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, cmap\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m     46\u001b[0m                  title\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, title_offset\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     47\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Plot spatial by clusters, transcripts, batches, etc.\"\"\"\u001b[39;00m\n\u001b[1;32m     48\u001b[0m     ann \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mcopy()\n",
      "\u001b[0;31mAttributeError\u001b[0m: partially initialized module 'corescpy' has no attribute 'pp' (most likely due to a circular import)"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "\n",
    "import os\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sb\n",
    "import traceback\n",
    "import warnings\n",
    "import functools\n",
    "from dask_image.imread import imread\n",
    "import imagecodecs\n",
    "from PIL import Image\n",
    "import dask.array as da\n",
    "import scipy.spatial as spatial\n",
    "import stlearn as st\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import corescpy as cr\n",
    "from corescpy.ax import create_spot_grid\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "pd.options.display.max_colwidth = 1000\n",
    "pd.options.display.max_columns = 100\n",
    "pd.options.display.max_rows = 500\n",
    "\n",
    "SPATIAL_KEY = \"spatial\"\n",
    "\n",
    "col_cell_type = \"leiden_res1pt5_dist0_npc30\"  # high resolution\n",
    "l_r = {\"CSF1\": \"CSF1R\", \"CSF2\": [\"CSF2RA\", \"CSF2RB\"]}\n",
    "plot_lr = [\"CSF1_CSF1R\", \"CSF2_CSF2RA\", \"CSF2_CSF2RB\"]\n",
    "\n",
    "n_spots = 400\n",
    "organism = \"human\"\n",
    "resource = \"connectomeDB2020_lit\"\n",
    "# distance = 200\n",
    "distance = \"auto\"\n",
    "min_med_neighbors = 5\n",
    "min_spots = 2\n",
    "n_pairs = 10000  # CHANGE DEFAULT TO 10000\n",
    "n_top = 50\n",
    "n_jobs = 20\n",
    "layer = \"counts\"\n",
    "cell_prop_cutoff = 0.1\n",
    "n_perms = 10000\n",
    "adj_method = \"fdr_bh\"\n",
    "pval_adj_cutoff = 0.05\n",
    "adj_axis = \"spot\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_subject, col_condition = \"Patient\", \"Condition\"\n",
    "col_inflamed, col_stricture = \"Inflamed\", \"Stricture\"  # in metadata file\n",
    "col_sample_id_o, col_sample_id = \"Sample ID\", \"Sample\"  # in metadata, new\n",
    "key_uninfl, key_infl, key_stric = \"Uninflamed\", \"Inflamed\", \"Stricture\"\n",
    "ddm = \"/mnt/cho_lab\" if os.path.exists(\"/mnt/cho_lab\") else \"/mnt\"  # Spark?\n",
    "ddl = f\"{ddm}/disk2/{os.getlogin()}/data/shared-xenium-library\" if (\n",
    "    \"cho\" in ddm) else os.path.join(ddu, \"shared-xenium-library\")\n",
    "file_mdf = os.path.join(ddl, \"samples.csv\")  # metadata\n",
    "m_d = (pd.read_excel if file_mdf[-4:] == \"xlsx\" else pd.read_csv)(\n",
    "    file_mdf, dtype={\"Slide ID\": str}).rename({\n",
    "        \"Name\": col_subject, \"Inflammation\": col_inflamed}, axis=1)\n",
    "m_d.loc[:, col_condition] = m_d.apply(lambda x: \"Stricture\" if x[\n",
    "    col_stricture].lower() in [\"stricture\", \"yes\"] else x[\n",
    "        col_inflamed].capitalize(), axis=1)  # inflamation/stricture condition\n",
    "m_d.loc[:, col_sample_id] = m_d[[col_condition, col_sample_id_o]].apply(\n",
    "    \"-\".join, axis=1)\n",
    "m_d = m_d.set_index(col_sample_id)\n",
    "print(m_d[[col_subject, col_condition]].reset_index(0)[\n",
    "    col_condition].value_counts())\n",
    "samps_paired = m_d.groupby(\"Patient\").apply(\n",
    "    lambda x: list(x.reset_index()[col_sample_id].sort_values()) if all((\n",
    "        i in list(x.reset_index()[col_condition]) for i in [\n",
    "            key_uninfl, key_infl, key_stric])) else np.nan).dropna(\n",
    "                ).explode()\n",
    "# m_d.reset_index().set_index([col_subject, col_condition]).sort_index()\n",
    "samps_paired"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spatial Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# File Paths & Options\n",
    "libids = [\"50452A\", \"50452B\", \"50006A\", \"50006B\",\n",
    "          \"50217A\", \"50217B\", \"50336B\", \"50336C\"]  # paired (un)inflamed\n",
    "libids = [m_d.reset_index().set_index(\n",
    "    col_sample_id_o).loc[i][col_sample_id] for i in libids]\n",
    "dir_data = os.path.join(ddm, \"bbdata2/outputs/TUQ97N\")\n",
    "out_dir = os.path.join(ddl, \"outputs/TUQ97N/nebraska\")\n",
    "col_cell_type = \"leiden_res1pt5_dist0_npc30\"  # high resolution\n",
    "\n",
    "# Spatial Data\n",
    "files = functools.reduce(lambda i, j: i + j, [[os.path.join(\n",
    "    run, i) for i in os.listdir(os.path.join(\n",
    "        dir_data, run))] for run in os.listdir(dir_data)])\n",
    "file_path = functools.reduce(\n",
    "    lambda x, y: list(x) + list(y), [np.array(files)[\n",
    "        np.where([\"-\".join(i.split(\"-\")[1:]) == os.path.basename(x).split(\n",
    "            \"__\")[2].split(\"-\")[0] for x in files])[0][0]] for i in libids])\n",
    "adatas = dict(zip(libids, [sc.read(os.path.join(\n",
    "    out_dir, f\"{i}.h5ad\")) for i in libids]))\n",
    "for x in adatas:\n",
    "    adatas[x].X = adatas[x].layers[layer].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Object Compatibility\n",
    "\n",
    "We have to re-configure a few object attributes to make it compatible with the expectations of `stlearn`, including legacy Squidpy styles of storing morphology image(s) and `stlearn`'s bespoke columns for coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make Compatible with Expected Hard-Coded Columns/Attributes in stlearn Code\n",
    "# max_coor = np.max(adata.obsm[\"spatial\"])\n",
    "# scale = 2000 / max_coor\n",
    "for x in adatas:\n",
    "    adatas[x].uns[SPATIAL_KEY] = {x: {\"images\": {}}}\n",
    "scale = 1\n",
    "quality = \"hires\"\n",
    "spot_diameter_fullres = 15\n",
    "for x in adatas:\n",
    "    adatas[x].obs.loc[:, \"imagecol\"] = adatas[x].obsm[\"spatial\"][:, 0] * scale\n",
    "    adatas[x].obs.loc[:, \"imagerow\"] = adatas[x].obsm[\"spatial\"][:, 1] * scale\n",
    "    if \"scalefactors\" not in adatas[x].uns[\"spatial\"]:\n",
    "        adatas[x].uns[\"spatial\"][x][\"scalefactors\"] = {}\n",
    "        adatas[x].uns[\"spatial\"][x][\"scalefactors\"][\n",
    "            \"tissue_\" + quality + \"_scalef\"] = scale\n",
    "        adatas[x].uns[\"spatial\"][x][\"scalefactors\"][\n",
    "            \"spot_diameter_fullres\"] = spot_diameter_fullres\n",
    "        adatas[x].uns[\"spatial\"][x][\"use_quality\"] = \"hires\"\n",
    "\n",
    "# Full Resolution? (Not Compatible Yet with stlearn)\n",
    "# image_kws = {}\n",
    "# img_files = {f for f in os.listdir(img_dir) if f.endswith(\"_focus.ome.tif\")}\n",
    "# channel_names = {0: \"DAPI\"} if len(img_files) == 1 else {\n",
    "#     0: \"DAPI\", 1: \"ATP1A1/CD45/E-Cadherin\", 2: \"18S\",\n",
    "#     3: \"AlphaSMA/Vimentin\", 4: \"dummy\"}\n",
    "# image_kws[\"c_coords\"] = list(channel_names.values())\n",
    "# image_path = os.path.join(img_dir, \"morphology_focus_{:04}.ome.tif\".format(\n",
    "#     0) if len(img_files) > 1 else \"morphology_focus.ome.tif\")\n",
    "# img = imread(image_path)\n",
    "# if \"c_coords\" in image_kws and \"dummy\" in image_kws[\"c_coords\"]:\n",
    "#     img = da.concatenate([img, da.zeros_like(img[0:1])], axis=0)\n",
    "# adata.uns[\"spatial\"][library_id][\"images\"] = {\"hires\": sq.im.ImageContainer(\n",
    "#     img, library_id=library_id)}\n",
    "\n",
    "# Stlearn Way of Loading Images (Full Resolution)\n",
    "# st.add.image(adata, library_id=library_id, quality=quality,\n",
    "#              imgpath=image_path, scale=scale)\n",
    "\n",
    "# Stlearn Way of Loading Images (Not Full Resolution)\n",
    "for x in adatas:\n",
    "    max_size = np.max([adatas[x].obs[\"imagecol\"].max(),\n",
    "                       adatas[x].obs[\"imagerow\"].max()])\n",
    "    max_size = int(max_size + 0.1 * max_size)\n",
    "    img = Image.new(\"RGBA\", (max_size, max_size), (255, 255, 255, 255))\n",
    "    adatas[x].uns[\"spatial\"][x][\"images\"] = {\"hires\": np.array(img)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Grids \n",
    "\n",
    "`stlearn` can create Visum-like data from Xenium data by aggregating the singl cells into larger \"spots\" (reducing the resolution) and performing label transfer to annotate the cells according to the dominant cluster/cell type in that spot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "# Re-Load Instead of Running\n",
    "# for x in adatas:\n",
    "#     grids[x] = sc.read(os.path.join(out_dir, f\"{x}_stlearn{suffix}.h5ad\"))\n",
    "#     _ = grids[x].uns.pop(col_cell_type + \"_colors\", None)\n",
    "#     # _ = adatas[x].uns.pop(col_cell_type + \"_colors\", None)\n",
    "\n",
    "# Run Gridding\n",
    "grids = {}\n",
    "for x in adatas:\n",
    "    grids[x], _ = create_spot_grid(adatas[x], col_cell_type, n_spots,\n",
    "                                   cmap=\"jana_40\", n_jobs=n_jobs, title=x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Cell Types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for x in grids[x]:\n",
    "#     groups = list(grids[x].obs[col_cell_type].cat.categories)\n",
    "#     for g in groups:\n",
    "#         fig, axes = plt.subplots(ncols=3, figsize=(20, 8))\n",
    "#         group_props = grids[x].uns[col_cell_type][g].values\n",
    "#         grids[x].obs.loc[:, \"Group\"] = group_props\n",
    "#         st.pl.feat_plot(grids[x], feature=\"Group\", ax=axes[0],\n",
    "#                         show_plot=False, vmax=1, show_color_bar=False)\n",
    "#         st.pl.cluster_plot(grids[x], use_label=col_cell_type,\n",
    "#                            list_clusters=[g], ax=axes[1], show_plot=False)\n",
    "#         st.pl.cluster_plot(grids[x], use_label=col_cell_type,\n",
    "#                            list_clusters=[g], ax=axes[2], show_plot=False)\n",
    "#         axes[0].set_title(f\"Grid {g} Proportions (Maximum = 1)\")\n",
    "#         axes[1].set_title(f\"Grid {g} Maximum Spots\")\n",
    "#         axes[2].set_title(f\"Individual Cell {g}\")\n",
    "#         fig.suptitle(x)\n",
    "#         plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check Number of Neighbors Given Different Distance Parameters\n",
    "\n",
    "Per https://github.com/BiomedicalMachineLearning/stLearn/issues/186#issuecomment-1238036775\n",
    "\n",
    "If distance is set to \"auto\" in the options cell earlier, then this dataframe will be used to set object/sample-specific distance parameters according to the `min_med_neighbors` parameter (i.e., will use the minimum distance that yields at least that specified number)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "neighbors = {}\n",
    "for d in np.arange(0, 301, 10):\n",
    "    nnb = {}\n",
    "    for x in grids:\n",
    "        point_tree = spatial.cKDTree(grids[x].obs[[\"imagerow\", \"imagecol\"]])\n",
    "        nnb[x] = point_tree.query_ball_point(point_tree.data[0], d)\n",
    "    neighbors[d] = pd.Series([len(nnb[i]) for i in nnb], index=pd.Index(\n",
    "        grids.keys(), name=col_sample_id))\n",
    "neighbors = pd.concat(neighbors, names=[\"Distances\"]).unstack(col_sample_id)\n",
    "sb.relplot(neighbors.stack().to_frame(\"Number of Neighbors\"), kind=\"line\",\n",
    "           hue=col_sample_id, x=\"Distances\", y=\"Number of Neighbors\")\n",
    "neighbors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis\n",
    "\n",
    "**Important Input**\n",
    "\n",
    "`adj_axis`: Per `stlearn` documentation: \"adjusting by no. of LRs tested per spot (LR), no. of spots tested per LR (spot), or no adjustment (None)\".\n",
    "\n",
    "`distance`: Per author on `stlearn` GitHub: Distance is the \"pixel length as the radius from each spot center. We will query all the spots within that radius.\" Also see https://github.com/BiomedicalMachineLearning/stLearn/issues/199#issuecomment-1272672934.\n",
    "\n",
    "**Important Output**\n",
    "\n",
    "`grid.uns[\"lr_summary\"]`: Summarizes results for each ligand-receptor pair (number of spots with L-R, number of significant spots, and p-values).\n",
    "\n",
    "`grid.obsm[\"lr_scores\"]`: Ligand-receptor pair scores.\n",
    "\n",
    "`grid.obsm[\"lr_sig_scores\"]`: Rows = spots and columns = ligand-receptor pairs. Per stLearn GitHub issue, \"Positive values in this matrix indicate significance of the LR in the respective spot (non-significant spots are set to zero).\" It has the same index as in `grid.uns[\"lr_summary\"]`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "distances = {}\n",
    "for x in grids:\n",
    "    ddd = min(neighbors[x].index.values[np.where(np.array(neighbors[x].apply(\n",
    "        lambda y: y >= min_med_neighbors)))[0]]\n",
    "              )  # least distance to have >= specified # of median neighbors\n",
    "    distances[x] = ddd if distance == \"auto\" else distance  # auto or fixed?\n",
    "    kws = dict(n_spots=n_spots, organism=organism, resource=resource,\n",
    "               distance=distances[x], min_med_neighbors=min_med_neighbors,\n",
    "               min_spots=min_spots, n_pairs=n_pairs, n_top=n_top,\n",
    "               n_jobs=n_jobs, layer=layer, cell_prop_cutoff=cell_prop_cutoff,\n",
    "               n_perms=n_perms, adj_method=adj_method,\n",
    "               pval_adj_cutoff=pval_adj_cutoff, adj_axis=adj_axis)\n",
    "    for v in kws:\n",
    "        grids[x].obs.loc[:, v] = str(kws[v])  # store keyword argument in .obs\n",
    "    print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\\n\\n\")\n",
    "    grids[x] = analyze_lr_spatial(\n",
    "        grids[x], col_cell_type, distances[x], min_spots, n_pairs,\n",
    "        n_perms=n_perms, n_jobs=n_jobs, cell_prop_cutoff=cell_prop_cutoff,\n",
    "        resource=resource, organism=organism, adj_axis=adj_axis,\n",
    "        pval_adj_cutoff=pval_adj_cutoff, adj_method=adj_method)\n",
    "pd.concat([grids[x].uns[\"lr_summary\"] for x in grids],\n",
    "          keys=grids.keys(), names=[col_sample_id])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    ddd = min(neighbors[x].index.values[np.where(np.array(neighbors[x].apply(\n",
    "        lambda y: y >= min_med_neighbors)))[0]]\n",
    "              )  # least distance to have >= specified # of median neighbors\n",
    "    distances[x] = ddd if distance == \"auto\" else distance  # auto or fixed?\n",
    "    kws = dict(n_spots=n_spots, organism=organism, resource=resource,\n",
    "               distance=distances[x], min_med_neighbors=min_med_neighbors,\n",
    "               min_spots=min_spots, n_pairs=n_pairs, n_top=n_top,\n",
    "               n_jobs=n_jobs, layer=layer, cell_prop_cutoff=cell_prop_cutoff,\n",
    "               n_perms=n_perms, adj_method=adj_method,\n",
    "               pval_adj_cutoff=pval_adj_cutoff, adj_axis=adj_axis)\n",
    "    for v in kws:\n",
    "        grids[x].obs.loc[:, v] = str(kws[v])  # store keyword argument in .obs\n",
    "    print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\\n\\n\")\n",
    "    grids[x] = analyze_lr_spatial(\n",
    "        grids[x], col_cell_type, distances[x], min_spots, n_pairs,\n",
    "        n_perms=n_perms, n_jobs=n_jobs, cell_prop_cutoff=cell_prop_cutoff,\n",
    "        resource=resource, organism=organism, adj_axis=adj_axis,\n",
    "        pval_adj_cutoff=pval_adj_cutoff, adj_method=adj_method)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write Objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "suffix = \"\"\n",
    "# suffix = f\"_{date}\"\n",
    "for x in grids:\n",
    "    grids[x].write_h5ad(os.path.join(out_dir, f\"{x}_stlearn{suffix}.h5ad\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def plot_spatial_lr_qc(grid, col_cell_type, n_top=50):\n",
    "#     \"\"\"Plot QC & results from spatial LR & CCI analysis.\"\"\"\n",
    "#     figs = {}\n",
    "#     try:\n",
    "#         st.pl.lr_diagnostics(grid, figsize=(10, 2.5))  # QC\n",
    "#         figs[\"qc_lr_diagnostics\"] = plt.gcf()\n",
    "#         st.pl.lr_n_spots(grid, n_top=n_top, figsize=(11, 3), max_text=100)\n",
    "#         figs[\"qc_lr_spots\"] = plt.gcf()\n",
    "#         figs[\"qc_lr_spots\"]].suptitle(\"L-R QC\")\n",
    "#     except Exception:\n",
    "#         traceback.print_exc()\n",
    "#         print(err, \"\\n\\L-R QC plot failed\")\n",
    "#     try:\n",
    "#         st.pl.cci_check(grid, col_cell_type, figsize=(16, 5))\n",
    "#         figs[\"qc_cci\"] = plt.gcf()\n",
    "#         figs[\"qc_cci\"].suptitle(\"CCI QC\")\n",
    "#     except Exception:\n",
    "#         traceback.print_exc()\n",
    "#         print(err, \"\\n\\nCCI QC plot failed\")\n",
    "#     return figs\n",
    "\n",
    "\n",
    "# def plot_spatial_lr_results(grid, col_cell_type, figsize=None,\n",
    "#                             stats=None, plot_lr=3, n_top=50, **kwargs):\n",
    "#     \"\"\"Plot QC & results from spatial LR & CCI analysis.\"\"\"\n",
    "#     figs = {}\n",
    "#     f_s = (12, 6) if figsize is None else figsize\n",
    "#     if stats is None:\n",
    "#         stats = [\"lr_scores\", \"p_vals\", \"p_adjs\", \"-log10(p_adjs)\"]\n",
    "#     if plot_lr not in [False, None]:\n",
    "#         plot_lr = [plot_lr] if isinstance(plot_lr, str) else grid.uns[\n",
    "#             \"lr_summary\"].index.values[:int(plot_lr)] if isinstance(plot_lr, (\n",
    "#                 int, float)) else list(plot_lr)  # use top N (plot_lr)\n",
    "#         ncols = 3 if len(plot_lr) > 3 else len(plot_lr)  # L-R facets (row)\n",
    "#         nrows = int(len(plot_lr) / ncols) if ncols > 1 else 1  # L-R columns\n",
    "#     st.pl.lr_summary(grid, n_top=n_top, figsize=f_s)  # plot L-R ranks\n",
    "#     figs[\"lr_ranks\"] = plt.gcf()\n",
    "#     if plot_lr is not False:\n",
    "#         fig, axes = kwargs.pop(\"fig\", None), kwargs.pop(\"axes\", None)\n",
    "#         if fig is None or axes is None:\n",
    "#             fig, axes = plt.subplots(ncols=len(stats), nrows=len(plot_lr),\n",
    "#                                         figsize=f_s)  # subplots: L-R pairs\n",
    "#         for r, lr in enumerate(plot_lr):  # iterate ligand-receptors\n",
    "#             kws = {\"show_color_bar\": False, **kwargs}\n",
    "#             try:\n",
    "#                 for c, stat in enumerate(stats):  # iterate statistics\n",
    "#                     st.pl.lr_result_plot(\n",
    "#                         grid, use_result=stat, use_lr=lr,\n",
    "#                         fig=fig, ax=axes[r, c], **kws)  # plot L-R\n",
    "#                     axes[r, c].set_title(f\"{lr} {stat}\")\n",
    "#             except Exception:\n",
    "#                 traceback.print_exc()\n",
    "#                 print(f\"Results summary plot error for {lr}\")\n",
    "#         figs[\"results\"] = fig, axes\n",
    "#     return figs\n",
    "\n",
    "\n",
    "\n",
    "# def plot_spatial_lr_map(grid, col_cell_type, figsize=None,\n",
    "#                         plot_lr=3, min_total=100, **kwargs):\n",
    "#     \"\"\"Plot QC & results from spatial LR & CCI analysis.\"\"\"\n",
    "#     figs = {}\n",
    "#     f_s = (12, 6) if figsize is None else figsize\n",
    "#     if plot_lr not in [False, None]:\n",
    "#         plot_lr = [plot_lr] if isinstance(plot_lr, str) else grid.uns[\n",
    "#             \"lr_summary\"].index.values[:int(plot_lr)] if isinstance(plot_lr, (\n",
    "#                 int, float)) else list(plot_lr)  # use top N (plot_lr)\n",
    "#         ncols = 3 if len(plot_lr) > 3 else len(plot_lr)  # L-R facets (row)\n",
    "#         nrows = int(len(plot_lr) / ncols) if ncols > 1 else 1  # L-R columns\n",
    "#     if \"lr_colors\" in kwargs:\n",
    "#         kwargs[\"lr_colors\"] = dict(zip(kwargs[\"lr_colors\"], [\n",
    "#             mpl.colors.to_hex(kwargs[\"lr_colors\"][u])\n",
    "#             for u in kwargs[\"lr_colors\"]]))  # convert colors to hex\n",
    "#     st.pl.cci_map(grid, col_cell_type, **kwargs)\n",
    "#     figs[\"map\"] = plt.gcf()\n",
    "#     if plot_lr is not False:\n",
    "#         fig, axes = kwargs.pop(\"fig\", None), kwargs.pop(\"axes\", None)\n",
    "#         if fig is None or axes is None:  # subplots: L-R pairs\n",
    "#             fig, axes = plt.subplots(ncols=ncols, nrows=nrows, figsize=f_s)\n",
    "#         for i, lr in enumerate(plot_lr):  # iterate L-R pairs\n",
    "#             try:\n",
    "#                 st.pl.cci_map(grid, col_cell_type, lr, show=False,\n",
    "#                               ax=axes.flatten()[i], **kwargs)  # heatmap plot\n",
    "#             except Exception:\n",
    "#                 traceback.print_exc()\n",
    "#                 print(f\"CCI map plot error for {lr}\")\n",
    "#         plt.show()\n",
    "#         try:\n",
    "#             st.pl.lr_cci_map(grid, col_cell_type, lrs=plot_lr,\n",
    "#                              min_total=min_total, figsize=f_s)\n",
    "#         except Exception:\n",
    "#             traceback.print_exc()\n",
    "#         figs[\"map_lr\"] = fig\n",
    "#     return figs\n",
    "\n",
    "\n",
    "# def plot_spatial_lr_network(grid, col_cell_type, **kwargs):\n",
    "#     \"\"\"Plot QC & results from spatial LR & CCI analysis.\"\"\"\n",
    "#     kind_all = [\"net\", \"chord\"]\n",
    "#     kind = [kind_all if kind.lower() == \"all\" else kind.lower()] if (\n",
    "#         isinstance(kind, str)) else [k.lower() for k in kind]\n",
    "#     if \"net\" in kind:\n",
    "#         f_s = (12, 6) if figsize is None else figsize\n",
    "#         pos_1 = st.pl.ccinet_plot(grid, col_cell_type, return_pos=True)  # all\n",
    "#         if plot_lr is not False:\n",
    "#             for lr in plot_lr:  # iterate L-R pairs\n",
    "#                 try:\n",
    "#                     st.pl.ccinet_plot(grid, col_cell_type, lr, figsize=f_s,\n",
    "#                                       min_counts=min_counts,\n",
    "#                                       pos=pos_1, **kwargs)\n",
    "#                 except Exception as err:\n",
    "#                     print(f\"Net plot error for {lr}\")\n",
    "#     if \"chord\" in kind:\n",
    "#         for lr in plot_lr:  # iterate L-R pairs\n",
    "#             try:\n",
    "#                 st.pl.lr_chord_plot(grid, col_cell_type, lr, **kwargs)\n",
    "#             except Exception:\n",
    "#                 traceback.print_exc()\n",
    "#                 print(f\"Chord plot error for {lr}\")\n",
    "\n",
    "\n",
    "# def plot_spatial_lr_arrow(grid, col_cell_type, kind=\"results\", figsize=None,\n",
    "#                           sig_spots=False, sig_cci=True, show_arrows=True,\n",
    "#                           stats=None, plot_lr=3, n_top=50, min_total=100, **kwargs):\n",
    "#     \"\"\"Plot QC & results from spatial LR & CCI analysis.\"\"\"\n",
    "#     figs = {}\n",
    "#     f_s = (12, 6) if figsize is None else figsize\n",
    "#     if \"lr_colors\" in kwargs:\n",
    "#         kwargs[\"lr_colors\"] = dict(zip(kwargs[\"lr_colors\"], [\n",
    "#             mpl.colors.to_hex(kwargs[\"lr_colors\"][u])\n",
    "#             for u in kwargs[\"lr_colors\"]]))  # convert colors to hex\n",
    "#     if plot_lr is not False:\n",
    "#         fig, axes = kwargs.pop(\"fig\", None), kwargs.pop(\"axes\", None)\n",
    "#         if fig is None or axes is None:\n",
    "#             fig, axes = plt.subplots(ncols=len(stats), nrows=len(plot_lr),\n",
    "#                                      figsize=f_s)  # subplots: L-R pairs\n",
    "#         for r, lr in enumerate(plot_lr):  # iterate ligand-receptors\n",
    "#             kws = {\"show_color_bar\": False, **kwargs}\n",
    "#             try:\n",
    "#                 for c, stat in enumerate(stats):  # iterate statistics\n",
    "#                     st.pl.lr_result_plot(grid, use_result=stat, ax=axes[r, c],\n",
    "#                                          use_lr=lr, fig=fig, **kws)  # plot\n",
    "#                     axes[r, c].set_title(f\"{lr} {stat}\")\n",
    "#             except Exception:\n",
    "#                 traceback.print_exc()\n",
    "#                 print(f\"Results summary plot error for {lr}\")\n",
    "#         figs[\"results\"] = fig, axes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_spatial_lr(grid, col_cell_type, kind=\"results\", figsize=None,\n",
    "                    sig_spots=False, sig_cci=True, show_arrows=True,\n",
    "                    stats=None, plot_lr=3, n_top=50, min_total=100, **kwargs):\n",
    "    \"\"\"Plot QC & results from spatial LR & CCI analysis.\"\"\"\n",
    "    figs = {}\n",
    "    f_s = (12, 6) if figsize is None else figsize\n",
    "    if stats is None:\n",
    "        stats = [\"lr_scores\", \"p_vals\", \"p_adjs\", \"-log10(p_adjs)\"]\n",
    "    kind_all = [\"qc\", \"results\", \"arrow\", \"map\", \"net\", \"chord\"]\n",
    "    kind = [kind_all if kind.lower() == \"all\" else kind.lower()] if (\n",
    "        isinstance(kind, str)) else [k.lower() for k in kind]\n",
    "    if plot_lr not in [False, None]:\n",
    "        plot_lr = [plot_lr] if isinstance(plot_lr, str) else grid.uns[\n",
    "            \"lr_summary\"].index.values[:int(plot_lr)] if isinstance(plot_lr, (\n",
    "                int, float)) else list(plot_lr)  # use top N (plot_lr)\n",
    "        ncols = 3 if len(plot_lr) > 3 else len(plot_lr)  # L-R facets (row)\n",
    "        nrows = int(len(plot_lr) / ncols) if ncols > 1 else 1  # L-R columns\n",
    "    if \"lr_colors\" in kwargs:\n",
    "        kwargs[\"lr_colors\"] = dict(zip(kwargs[\"lr_colors\"], [\n",
    "            mpl.colors.to_hex(kwargs[\"lr_colors\"][u])\n",
    "            for u in kwargs[\"lr_colors\"]]))  # convert colors to hex\n",
    "\n",
    "    # QC/Diagnostics (Metrics Shouldn't Correlate Much with L-R Expression)\n",
    "    if \"qc\" in kind:\n",
    "        print(f\"\\n\\n{'*' * 40}\\nQC Plots\\n{'*' * 40}\\n\\n\"\n",
    "              \"Expression frequency and L-R/CCI metrics shouldn't correlate much\"\n",
    "              \"; otherwise, increase `n_perms` and/or `n_pairs`.\")\n",
    "        try:\n",
    "            st.pl.lr_diagnostics(grid, figsize=(10, 2.5))  # QC\n",
    "            figs[\"qc_lr_diagnostics\"] = plt.gcf()\n",
    "            st.pl.lr_n_spots(grid, n_top=n_top, figsize=(11, 3), max_text=100)\n",
    "            figs[\"qc_lr_spots\"] = plt.gcf()\n",
    "        except Exception:\n",
    "            traceback.print_exc()\n",
    "            print(err, \"\\n\\L-R QC plot failed\")\n",
    "        try:\n",
    "            st.pl.cci_check(grid, col_cell_type, figsize=(16, 5))\n",
    "            figs[\"qc_cci\"] = plt.gcf()\n",
    "            figs[\"qc_cci\"].suptitle(\"CCI Check: Interactions & cell type \"\n",
    "                                    \"frequency shouldn't correlate much\")\n",
    "        except Exception:\n",
    "            traceback.print_exc()\n",
    "            print(err, \"\\n\\nCCI QC plot failed\")\n",
    "\n",
    "    # Results Summary (Rankings of L-R Pairs by Number of Significant Spots)\n",
    "    if \"results\" in kind:\n",
    "        print(f\"\\n\\n{'*' * 40}\\nResults Plots\\n{'*' * 40}\\n\\n\")\n",
    "        st.pl.lr_summary(grid, n_top=n_top, figsize=f_s)  # plot L-R ranks\n",
    "        figs[\"lr_ranks\"] = plt.gcf()\n",
    "        if plot_lr is not False:\n",
    "            fig, axes = kwargs.pop(\"fig\", None), kwargs.pop(\"axes\", None)\n",
    "            if fig is None or axes is None:\n",
    "                fig, axes = plt.subplots(ncols=len(stats), nrows=len(plot_lr),\n",
    "                                         figsize=f_s)  # subplots: L-R pairs\n",
    "            for r, lr in enumerate(plot_lr):  # iterate ligand-receptors\n",
    "                kws = {\"show_color_bar\": False, **kwargs}\n",
    "                try:\n",
    "                    for c, stat in enumerate(stats):  # iterate statistics\n",
    "                        st.pl.lr_result_plot(\n",
    "                            grid, use_result=stat, use_lr=lr,\n",
    "                            fig=fig, ax=axes[r, c], **kws)  # plot L-R\n",
    "                        axes[r, c].set_title(f\"{lr} {stat}\")\n",
    "                except Exception:\n",
    "                    traceback.print_exc()\n",
    "                    print(f\"Results summary plot error for {lr}\")\n",
    "            figs[\"results\"] = fig, axes\n",
    "\n",
    "    # Color Spot by Mean L-R Expression in Spots Connected by Arrow\n",
    "    if \"arrow\" in kind:\n",
    "        fig, axes = kwargs.pop(\"fig\", None), kwargs.pop(\"axes\", None)\n",
    "        if fig is None or axes is None:\n",
    "            fig, axes = plt.subplots(ncols=ncols, nrows=nrows,\n",
    "                                     figsize=f_s, squeeze=False)\n",
    "        for i, lr in enumerate(plot_lr):  # iterate L-R pairs\n",
    "            axes.flatten()[i].set_title(lr)\n",
    "            kws = dict(arrow_width=2, pt_scale=10, arrow_head_width=4,\n",
    "                       arrow_cmap=\"YlOrRd\", arrow_vmax=1.5, outer_mode=None,\n",
    "                       outer_size_prop=1, fig=fig, ax=axes.flatten()[i],\n",
    "                       sig_spots=sig_spots, sig_cci=sig_cci)\n",
    "            kws.update({**kwargs, \"show_arrows\": False} if (\n",
    "                col_cell_type is None) else {**kwargs})\n",
    "            if \"lr_colors\" in kws:\n",
    "                kws[\"lr_colors\"] = {**kws[\"lr_colors\"]}\n",
    "                for v, k in enumerate([\"l\", \"r\", \"lr\"]):  # key palette by L-R\n",
    "                    lab = lr.split(\"_\")[v] if v < 2 else lr\n",
    "                    kws[\"lr_colors\"][lab] = kws[\"lr_colors\"][k]\n",
    "                    _ = kws[\"lr_colors\"].pop(k)\n",
    "            try:\n",
    "                st.pl.lr_plot(grid, lr, use_label=col_cell_type, **kws)\n",
    "            except Exception:\n",
    "                traceback.print_exc()\n",
    "                print(f\"L-R arrow plot error for {lr}\")\n",
    "        figs[\"arrow\"] = fig, axes\n",
    "        plt.show()\n",
    "\n",
    "    # Net Plot: # of Interactions between Cell Types across All L-R Pairs\n",
    "    if \"net\" in kind:\n",
    "        pos_1 = st.pl.ccinet_plot(grid, col_cell_type, return_pos=True)  # all\n",
    "        if plot_lr is not False:\n",
    "            for lr in plot_lr:  # iterate L-R pairs\n",
    "                try:\n",
    "                    st.pl.ccinet_plot(grid, col_cell_type, lr, min_counts=2,\n",
    "                                      figsize=f_s, pos=pos_1, **kwargs)\n",
    "                except Exception as err:\n",
    "                    print(f\"Net plot error for {lr}\")\n",
    "\n",
    "    # Chord Plot\n",
    "    if \"chord\" in kind:\n",
    "        # st.pl.lr_chord_plot(grid, col_cell_type)  # all\n",
    "        if plot_lr is not False:\n",
    "            for lr in plot_lr:  # just between selected pairs\n",
    "                try:\n",
    "                    st.pl.lr_chord_plot(grid, col_cell_type, lr)\n",
    "                except Exception:\n",
    "                    traceback.print_exc()\n",
    "                    print(f\"Chord plot error for {lr}\")\n",
    "\n",
    "    # Heatmap Equivalent of Net & Chord\n",
    "    if \"map\" in kind:\n",
    "        st.pl.cci_map(grid, col_cell_type, **kwargs)\n",
    "        if plot_lr is not False:\n",
    "            fig, axes = kwargs.pop(\"fig\", None), kwargs.pop(\"axes\", None)\n",
    "            if fig is None or axes is None:\n",
    "                fig, axes = plt.subplots(ncols=ncols, nrows=nrows,\n",
    "                                         figsize=f_s)  # subplots: L-R pairs\n",
    "            for i, lr in enumerate(plot_lr):  # iterate L-R pairs\n",
    "                try:\n",
    "                    figs[\"map\"] = st.pl.cci_map(\n",
    "                        grid, col_cell_type, lr, show=False,\n",
    "                        ax=axes.flatten()[i], **kwargs)  # heatmap plot\n",
    "                except Exception:\n",
    "                    traceback.print_exc()\n",
    "                    print(f\"CCI map plot error for {lr}\")\n",
    "            plt.show()\n",
    "            try:\n",
    "                st.pl.lr_cci_map(grid, col_cell_type, lrs=plot_lr,\n",
    "                                 min_total=min_total, figsize=(20, 4))\n",
    "            except Exception:\n",
    "                traceback.print_exc()\n",
    "    return figs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## QC/Diagnostics\n",
    "\n",
    "**First Set of Plots (For L-R Analysis; Scatter)**\n",
    "\n",
    "`st.pl.lr_diagnostics`: Per `stlearn` documentation...\n",
    "\n",
    "> A key aspect of the LR analysis is to control for LR expression level and frequency when calling significant hotspots.\n",
    ">\n",
    "> Hence, our diagnostic plots should show next to no correlation between the hotspots of the LR pairs and the expression level and frequency of expression.\n",
    ">\n",
    "> The following diagnostics allow us to check and make sure this is the case; if not, could indicate a larger number of permutations is required.\n",
    ">\n",
    ">\n",
    "> * Left plot: Relationship between LR expression level (non-zero spots average median expression of genes in the LR pair) and the ranking of the LR.\n",
    "> \n",
    "> * Right plot: Relationship between LR expression frequency (average proportion of zero spots for each gene in the LR pair) and the ranking of the LR.\n",
    "> \n",
    "> [If] there is a...correlation between the LR expression frequency and number of significant spots, [this indicates] the n_pairs parameter should be set higher to create more accurate background distributions (10,000 pairs was used in the case of the paper version of the above).\n",
    "\n",
    "**Second Set of Plots (For L-R Analysis; Bars)**\n",
    "\n",
    "`st.pl.lr_n_spots`: Per `stlearn` documentation...\n",
    "> The above boxplots show the number of spots with ligand-receptor scores for each LR on the y-axis, with the LR ranking on the x-axis. The bar area is stratified by spots which were significant (green) and non-significant (blue).\n",
    "> \n",
    "> [If] there does appear to be some correlation with more highly frequent LR pairs and LR ranking...the n_pairs parameter above should be set higher.\n",
    "\n",
    "**Third Set of Plots (for CCI Analysis)**\n",
    "\n",
    "Interactions & cell type frequency shouldn't correlate much if well-controlled for."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in grids:\n",
    "    print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\")\n",
    "    plot_spatial_lr(grids[x], col_cell_type, kind=\"qc\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## L-R Results Summary\n",
    "\n",
    "Rankings, p-values, and scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_spots = True\n",
    "# sig_spots = False\n",
    "for x in grids:\n",
    "    print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\")\n",
    "    plot_spatial_lr(grids[x], col_cell_type, kind=\"results\",\n",
    "                    plot_lr=plot_lr, sig_spots=sig_spots)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## L-R Co-Expression\n",
    "\n",
    "`outer_mode` & corresponding color map arguments: Per `stlearn` documentation...\n",
    "> The mode for the larger points when displaying LR expression; can either be ‘binary’ or ‘continuous’ or None. \n",
    "* > ‘Binary’ discretizes each spot as expressing L, R, both, or neither. \n",
    "   - `inner_cmap`: \"Cmap for the inner point\"\n",
    "   - `lr_colors`: \"Specifies the colors of the LRs...{‘l’: color, ‘r’: color, ‘lr’: color, ‘’: color}; the last key-value indicates colour for spots not expressing the ligand or receptor\"\n",
    "* > ‘Continuous’ shows color gradient for levels of LR expression by plotting two points for each spot, the ‘inner’ point is the receptor expression levels, and the ‘outer’ point is the ligand expression level. None plots no ligand/receptor expression.\n",
    "   - `l_cmap`:  \"Cmap for coloring the ligand expression\"\n",
    "   - `r_cmap`: \"Cmap for coloring the receptor expression\"\n",
    "   - `lr_cmap`: \"Cmap for coloring coexpression\"\n",
    "\n",
    "`min_expr`: Per `stlearn`...\"The minimum expr above which LR considered expressed when plotting binary LR expression.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set to None for Just L-R; `col_cell_type` to Show Labels Too\n",
    "cct = None\n",
    "# cct = col_cell_type\n",
    "\n",
    "# Plot\n",
    "for m in [\"binary\", \"continuous\"]:\n",
    "    for s in [True, False]:\n",
    "        for x in grids:\n",
    "            palette = {\"lr_colors\": {\"l\": \"g\", \"r\": \"b\", \"lr\":\n",
    "                \"y\", \"\": \"w\"}} if m == \"binary\" else {}\n",
    "            title = str(f\"{x}: {m.capitalize()} Coexpression\"\n",
    "                        f\" ({'Significant Only' if s else 'All'})\")\n",
    "            plot_spatial_lr(grids[x], cct, kind=\"arrow\", plot_lr=plot_lr,\n",
    "                            outer_mode=m, title=title, sig_spots=s,\n",
    "                            figsize=(35, 15), **palette)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell-Cell Interactions\n",
    "\n",
    "Per `stlearn`...\n",
    "> The number of interactions refers to the number of times a spot with the reciever cell type expressed the ligand and the source cell type expressed the receptor in the same neighbourhood."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CCI Map\n",
    "\n",
    "**First Plot (CCI)**\n",
    "Per `stlearn` documentation:\n",
    "> This is a heatmap equivalent to the network diagrams and chordplots, it has more quantitative benefits.\n",
    "> The # of interactions refers to the number of times a spot with the reciever cell type expressed the ligand and the source cell type expressed the receptor in the same neighbourhood.\n",
    "\n",
    "**Second Plot (LR-CCI)**\n",
    "\n",
    "CCI plots by individual L-R pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sig = False\n",
    "sig = True  # only show significant\n",
    "for x in grids:\n",
    "    print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\")\n",
    "    plot_spatial_lr(grids[x], col_cell_type, kind=\"map\",\n",
    "                    figsize=(12 * len(plot_lr), 10), sig_interactions=sig,\n",
    "                    title=str(f\"{x}: Cell-Cell Interactions (\"\n",
    "                              f\"(Significant{'' if sig else ' and N.S.'})\"),\n",
    "                    plot_lr=plot_lr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CCI Chord\n",
    "\n",
    "Most useful for small numbers of cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in grids:\n",
    "    print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\")\n",
    "    plot_spatial_lr(grids[x], col_cell_type, kind=\"chord\", plot_lr=plot_lr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CCI Arrow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in grids:\n",
    "    print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\")\n",
    "    plot_spatial_lr(grids[x], col_cell_type, kind=\"arrow\",\n",
    "                    figsize=(30, 30), plot_lr=plot_lr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CCI Net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in grids:\n",
    "    print(f\"\\n\\n{'=' * 80}\\n{x}\\n{'=' * 80}\")\n",
    "    plot_spatial_lr(grids[x], col_cell_type, title=f\"{x}: L-R Interactions\",\n",
    "                    kind=\"net\", plot_lr=plot_lr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if plot_lr is not None:\n",
    "    genes = functools.reduce(lambda i, j: list(i) + list(j),\n",
    "                             [i.split(\"_\") for i in plot_lr])\n",
    "    for g in genes:\n",
    "        fig, axes = plt.subplots(ncols=2, figsize=(20, 5))\n",
    "        st.pl.gene_plot(grid, gene_symbols=g, ax=axes[0],\n",
    "                        show_color_bar=False, show_plot=False)\n",
    "        st.pl.gene_plot(adata, gene_symbols=g, ax=axes[1],\n",
    "                        show_color_bar=False, show_plot=False, vmax=80)\n",
    "        axes[0].set_title(f\"Grid {g} Expression\")\n",
    "        axes[1].set_title(f\"Cell {g} Expression\")\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_int_dfs = grid.uns[f\"per_lr_cci_{col_cell_type}\"]\n",
    "lr_dfs = [lr_int_dfs[lr] if lr in lr_int_dfs else np.nan for lr in plot_lr]\n",
    "[x.replace(0, np.nan).stack().dropna().to_frame(plot_lr[i]) if isinstance(\n",
    "    x, pd.DataFrame) else None for i, x in enumerate(lr_dfs)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "spatial-alt",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
