{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "This is the stuff you have to edit; the rest of the sections can run as-is after you've set the needed parameters.\n",
    "\n",
    "---\n",
    "\n",
    "`coord_suffix` must align with \n",
    "* the sub-directory (corresponds to region, e.g., \"mucosa\") under the `dir_coord` directory where the Xenium Explorer-exported selection files are stored, and\n",
    "* the suffixes of the coordinate selection files (see file naming conventions below).\n",
    "  \n",
    "The `AnnData` objects created will have this suffix as well (e.g., `Uninflamed-50452A_mucosa.h5ad`).\n",
    "\n",
    "---\n",
    "\n",
    "Selection files should be named by this convention:\n",
    "`<library_id>_<coord_suffix>.csv`.\n",
    "\n",
    "For example, if `dir_coord` is `.../coordinates/mucosa`, the mucosa selection file for sample 50452A should be under `.../coordinates/mucosa/50452A_mucosa.csv`. \n",
    "\n",
    "More specifically, if the coordinates directory is under `/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/coordinates`, and the selection region is \"mucosa,\"`dir_coord` should be `/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/coordinates/mucosa`, and the full file path for this sample would be `/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/coordinates/mucosa/50452A_mucosa.csv`.\n",
    "\n",
    "---\n",
    "\n",
    "**As with any other file naming schema, suffixes/directory names should not any special characters other than underscores (`_`) (no periods, dashes, spaces, etc.).**\n",
    "\n",
    "N.B. In the above explanation, `library_id` refers to library/original sample ID without condition (e.g., \"50452A\", not \"Uninflamed-50452A\" like in other places). Remember that `coord_suffix` should also be the name of the parent directory of the coordinate file. I include this information in both the directory and file name to prevent mix-ups should files be moved or placed in the wrong folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #808000; text-decoration-color: #808000; font-weight: bold\">ryp2 is not installed. Install with </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">pip install rpy2 </span><span style=\"color: #808000; text-decoration-color: #808000; font-weight: bold\">to run tools with R support.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;33mryp2 is not installed. Install with \u001b[0m\u001b[1;32mpip install rpy2 \u001b[0m\u001b[1;33mto run tools with R support.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading data from `https://omnipathdb.org/queries/enzsub?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/interactions?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/complexes?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/annotations?format=json`\n",
      "Downloading data from `https://omnipathdb.org/queries/intercell?format=json`\n",
      "Downloading data from `https://omnipathdb.org/about?format=text`\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "\n",
    "import os\n",
    "import re\n",
    "import math\n",
    "import functools\n",
    "import traceback\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import corescpy as cr\n",
    "\n",
    "# Main\n",
    "write_object = True  # change to True when you're ready to save objects\n",
    "overwrite = False  # overwrite if already exists?\n",
    "coord_suffix = \"smc_circular\"  # sub-directory & file/object suffix\n",
    "\n",
    "# Process Options\n",
    "panel_id = \"TUQ97N\"  # Xenium panel ID\n",
    "show_cols = [cr.pp.COL_SUBJECT]  # only change if not TUQ97N\n",
    "cso = cr.pp.COL_SAMPLE_ID_O  # original library ID; change if not TUQ97N\n",
    "col_sample = cr.pp.COL_SAMPLE_ID  # object sample ID; change if not TUQ97N\n",
    "# libs = [  # sample IDs from patients for whom we have all conditions\n",
    "#     \"50452A\", \"50452B\", \"50452C\",  # old segmentation\n",
    "#     \"50006A\", \"50006B\", \"50006C\",  # rest are new segmentation\n",
    "#     \"50217A\", \"50217B\", \"50217C\",\n",
    "#     \"50336B\", \"50336C\", \"50336A\",\n",
    "#     \"50403A2\", \"50403B\", \"50403C1\"\n",
    "# ]  # excludes low-quality sample/condition replicates 50403A1 & 50403C2\n",
    "libs = [\"50006A\", \"50006B\", \"50217A\", \"50217B\", \"50336B\", \"50336C\",\n",
    "        \"50403B\", \"50403A2\"]  # just inflamed/uninflamed (no strictures)\n",
    "# libs = None  # to run all available samples\n",
    "input_suffix = \"\"  # in case want to crop objects with some suffix\n",
    "# due to creation of a subsidiary object, e.g., for\n",
    "# \"Stricture-50452C_downsampled.h5ad\"\n",
    "# input_suffix would be \"_downsampled\". For \"main\" objects, input_suffix=\"\"\n",
    "plot = True  # could slow process down if large samples/cropped area\n",
    "\n",
    "# Files & Directories\n",
    "direc = \"/mnt/cho_lab/bbdata2/\"  # mounted NFS with data\n",
    "dir_entry = \"/mnt/cho_lab/disk2\"  # Spark writeable data directory\n",
    "mdf = str(\"/mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/\"\n",
    "          \"samples.csv\")  # metadata file path (for now; will soon be on NFS)\n",
    "dir_writeable = os.path.join(\n",
    "    dir_entry, f\"elizabeth/data/shared-xenium-library\")  # where objects are\n",
    "out_dir = os.path.join(\n",
    "    dir_writeable, f\"outputs/{panel_id}/nebraska\")  # object output directory\n",
    "\n",
    "#  Your Folders\n",
    "out_new = os.path.join(\n",
    "    dir_entry,\n",
    "    f\"{os.getlogin()}/data/shared-xenium-library/outputs/{panel_id}/nebraska\")\n",
    "dir_coord = os.path.join(\n",
    "    out_new, \"coordinates\", coord_suffix)  # coordinates (also maybe NFS soon)\n",
    "out_sub = os.path.join(out_new, \"objects_cropped\",\n",
    "                       coord_suffix)  # where cropped objects will go\n",
    "# Your Folders\n",
    "\n",
    "# Constants (Shouldn't Need Edits Unless Extreme Process Changes)\n",
    "dir_data = os.path.join(direc, f\"outputs/{panel_id}\")\n",
    "col_obj = cr.pp.COL_OBJECT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find Files & Load Metadata\n",
    "\n",
    "Allows us to find the object IDs (e.g., for TUQ97N, object IDs are in the format <condition (Inflamed/Uninflamed/Stricture)><block_id>) corresponding to the sample IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Inflamed-50006A</th>\n",
       "      <td>50006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uninflamed-50006B</th>\n",
       "      <td>50006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inflamed-50217A</th>\n",
       "      <td>50217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uninflamed-50217B</th>\n",
       "      <td>50217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inflamed-50336B</th>\n",
       "      <td>50336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uninflamed-50336C</th>\n",
       "      <td>50336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inflamed-50403B</th>\n",
       "      <td>50403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uninflamed-50403A2</th>\n",
       "      <td>50403</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Name\n",
       "Sample                   \n",
       "Inflamed-50006A     50006\n",
       "Uninflamed-50006B   50006\n",
       "Inflamed-50217A     50217\n",
       "Uninflamed-50217B   50217\n",
       "Inflamed-50336B     50336\n",
       "Uninflamed-50336C   50336\n",
       "Inflamed-50403B     50403\n",
       "Uninflamed-50403A2  50403"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = functools.reduce(lambda i, j: i + j, [[os.path.join(\n",
    "    run, i) for i in os.listdir(os.path.join(\n",
    "        dir_data, run))] for run in os.listdir(dir_data)])  # all data paths\n",
    "os.makedirs(out_dir, exist_ok=True)  # make output directory if needed\n",
    "os.makedirs(out_sub, exist_ok=True)  # make sub-directory to write new object\n",
    "metadata = cr.pp.get_metadata_cho(direc, mdf, panel_id=panel_id, samples=libs)\n",
    "metadata[show_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subset Data by Coordinate Files & Write Cropped Objects\n",
    "\n",
    "Subset the data by coordinates (`corescpy` can use Xenium Explorer-exported manual selection files to get those coordinates) and then write the cropped objects to `out_dir/<coord_suffix>`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "****************************************\n",
      "50006A\n",
      "****************************************\n",
      "\n",
      "\n",
      "*** Subset /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/objects_cropped/smc_circular/Inflamed-50006A_smc_circular.h5ad already exists\n",
      "\n",
      "\n",
      "****************************************\n",
      "50006B\n",
      "****************************************\n",
      "\n",
      "\n",
      "*** Subset /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/objects_cropped/smc_circular/Uninflamed-50006B_smc_circular.h5ad already exists\n",
      "\n",
      "\n",
      "****************************************\n",
      "50217A\n",
      "****************************************\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Data: /mnt/cho_lab/bbdata2/outputs/TUQ97N/CHO-008/output-XETG00189__0015521__50217A-TUQ97N-EA__20240415__212906\n",
      "Object: /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/Inflamed-50217A.h5ad\n",
      "Coordinates: /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/coordinates/smc_circular/50217A_smc_circular.csv\n",
      "Ouput: /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/objects_cropped/smc_circular/Inflamed-50217A_smc_circular.h5ad\n",
      "\n",
      "\n",
      "<<< INITIALIZING SPATIAL CLASS OBJECT >>>\n",
      "\n",
      "\u001b[34mINFO    \u001b[0m reading                                                                                                   \n",
      "         \u001b[35m/mnt/cho_lab/bbdata2/outputs/TUQ97N/CHO-008/output-XETG00189__0015521__50217A-TUQ97N-EA__20240415__212906/\u001b[0m\n",
      "         \u001b[95mcell_feature_matrix.h5\u001b[0m                                                                                    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "\n",
      "\n",
      "================================================================================\n",
      "Counts: Initial\n",
      "================================================================================\n",
      "\n",
      "\n",
      "\tObservations: 1038830\n",
      "\n",
      "\tGenes: 469\n",
      "\n",
      "\n",
      "\n",
      "================================================================================\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " AnnData object with n_obs × n_vars = 1038830 × 469\n",
      "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'z_level', 'nucleus_count', 'cell_labels', 'Sample'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "    uns: 'spatialdata_attrs', 'spatial', 'original_ix'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'counts' \n",
      "\n",
      "                      gene_ids    feature_types   genome\n",
      "gene_symbols                                           \n",
      "ABCA7         ENSG00000064687  Gene Expression  Unknown\n",
      "ACTA2         ENSG00000107796  Gene Expression  Unknown\n",
      "ACTG2         ENSG00000163017  Gene Expression  Unknown\n",
      "ADAM15        ENSG00000143537  Gene Expression  Unknown\n",
      "ADAM28        ENSG00000042980  Gene Expression  Unknown \n",
      "\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "col_gene_symbols=\"gene_symbols\"\n",
      "col_cell_type=\"leiden\"\n",
      "col_sample_id=\"Sample\"\n",
      "col_batch=\"Sample\"\n",
      "col_subject=None\n",
      "col_condition=None\n",
      "col_num_umis=None\n",
      "col_segment=\"region\"\n",
      "\n",
      "\n",
      "\n",
      "key_control=None\n",
      "key_treatment=None\n",
      "\n",
      "\n",
      "<<< RESTORING SINGLE-SAMPLE FROM h5ad >>>\n",
      "\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "WARNING: Unable to find `y`, `x` or `z` dimension in `('c', 'y', 'x')`. Renaming to `('channels', 'y', 'x', 'z')`\n",
      "Traceback (most recent call last):\n",
      "  File \"<timed exec>\", line 21, in <module>\n",
      "  File \"/home/elizabeth/elizabeth/corescpy/corescpy/class_spatial.py\", line 264, in crop\n",
      "    sdata_crop = self.adata.query.polygon(coords, **kws)  # crop\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/spatialdata/_core/spatialdata.py\", line 2192, in polygon\n",
      "    return polygon_query(  # type: ignore[return-value]\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/functools.py\", line 889, in wrapper\n",
      "    return dispatch(args[0].__class__)(*args, **kw)\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/spatialdata/_core/query/spatial_query.py\", line 823, in _\n",
      "    queried_elements = _dict_query_dispatcher(\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/spatialdata/_core/query/spatial_query.py\", line 422, in _dict_query_dispatcher\n",
      "    result = query_function(element, **kwargs)\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/functools.py\", line 889, in wrapper\n",
      "    return dispatch(args[0].__class__)(*args, **kw)\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/spatialdata/_core/query/spatial_query.py\", line 902, in _\n",
      "    buffered = to_polygons(element) if ShapesModel.RADIUS_KEY in element.columns else element\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/functools.py\", line 889, in wrapper\n",
      "    return dispatch(args[0].__class__)(*args, **kw)\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/spatialdata/_core/operations/vectorize.py\", line 267, in _\n",
      "    buffered_df[\"geometry\"] = buffered_df.apply(\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/geopandas/geodataframe.py\", line 1834, in apply\n",
      "    result = super().apply(\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/pandas/core/frame.py\", line 10374, in apply\n",
      "    return op.apply().__finalize__(self, method=\"apply\")\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/pandas/core/apply.py\", line 916, in apply\n",
      "    return self.apply_standard()\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/pandas/core/apply.py\", line 1063, in apply_standard\n",
      "    results, res_index = self.apply_series_generator()\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/pandas/core/apply.py\", line 1081, in apply_series_generator\n",
      "    results[i] = self.func(v, *self.args, **self.kwargs)\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/spatialdata/_core/operations/vectorize.py\", line 268, in <lambda>\n",
      "    lambda row: row.geometry.buffer(row[ShapesModel.RADIUS_KEY], resolution=buffer_resolution), axis=1\n",
      "  File \"/home/elizabeth/elizabeth/miniconda3/envs/corescpy/lib/python3.10/site-packages/shapely/geometry/base.py\", line 543, in buffer\n",
      "    raise ValueError(\"buffer distance must be finite\")\n",
      "ValueError: buffer distance must be finite\n",
      " Cropping \n",
      "\n",
      "50217A failed!\n",
      "\n",
      "\n",
      "****************************************\n",
      "50217B\n",
      "****************************************\n",
      "\n",
      "\n",
      "*** Subset /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/objects_cropped/smc_circular/Uninflamed-50217B_smc_circular.h5ad already exists\n",
      "\n",
      "\n",
      "****************************************\n",
      "50336B\n",
      "****************************************\n",
      "\n",
      "\n",
      "*** Subset /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/objects_cropped/smc_circular/Inflamed-50336B_smc_circular.h5ad already exists\n",
      "\n",
      "\n",
      "****************************************\n",
      "50336C\n",
      "****************************************\n",
      "\n",
      "\n",
      "*** Subset /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/objects_cropped/smc_circular/Uninflamed-50336C_smc_circular.h5ad already exists\n",
      "\n",
      "\n",
      "****************************************\n",
      "50403B\n",
      "****************************************\n",
      "\n",
      "\n",
      "*** Subset /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/objects_cropped/smc_circular/Inflamed-50403B_smc_circular.h5ad already exists\n",
      "\n",
      "\n",
      "****************************************\n",
      "50403A2\n",
      "****************************************\n",
      "\n",
      "\n",
      "*** Subset /mnt/cho_lab/disk2/elizabeth/data/shared-xenium-library/outputs/TUQ97N/nebraska/objects_cropped/smc_circular/Uninflamed-50403A2_smc_circular.h5ad already exists\n",
      "CPU times: user 1h 6min 20s, sys: 39min 8s, total: 1h 45min 28s\n",
      "Wall time: 1h 48s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "for s in libs:  # iterate samples\n",
    "    print(f\"\\n\\n{'*' * 40}\\n{s}\\n{'*' * 40}\\n\\n\")\n",
    "    fff = os.path.join(dir_data, np.array(files)[np.where([\n",
    "        s == os.path.basename(x).split(\"__\")[2].split(\"-\")[0]\n",
    "        for x in files])[0][0]])  # sample's Xenium data directory path\n",
    "    lib = metadata.reset_index().set_index(cso).loc[s][col_sample]\n",
    "    file_obj_proc = os.path.join(out_dir, f\"{lib}{input_suffix}.h5ad\")\n",
    "    file_coord = os.path.join(dir_coord, s + f\"_{coord_suffix}.csv\")\n",
    "    file_obj_crop = os.path.join(out_sub, f\"{lib}_{coord_suffix}.h5ad\")\n",
    "    if overwrite is False and os.path.exists(file_obj_crop):\n",
    "        print(f\"*** Subset {file_obj_crop} already exists\")\n",
    "        continue\n",
    "    if not os.path.exists(file_coord):\n",
    "        print(f\"*** Coordinate file {file_coord} doesn't exist\")\n",
    "        continue\n",
    "    print(f\"\\n\\nData: {fff}\\nObject: {file_obj_proc}\"\n",
    "          f\"\\nCoordinates: {file_coord}\\nOuput: {file_obj_crop}\")\n",
    "    self = cr.Spatial(fff, library_id=lib)  # load original data\n",
    "    self.update_from_h5ad(file_obj_proc)  # update with processed object\n",
    "    try:\n",
    "        sdata = self.crop(file_coord)  # crop data to coordinates\n",
    "        if plot:\n",
    "            try:\n",
    "                sdata.pl.render_labels(\"cell_labels\").pl.show()  # plot crop\n",
    "            except Exception:\n",
    "                pass\n",
    "        self.adata = sdata\n",
    "        if write_object is True:\n",
    "            self.write(file_obj_crop)  # write cropped\n",
    "    except Exception:\n",
    "        print(traceback.format_exc(), f\"Cropping \\n\\n{s} failed!\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
